{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b92a1ed5-e3b5-493f-ad95-1898c100da13",
   "metadata": {},
   "source": [
    "```bash\n",
    "{\n",
    "conda activate\n",
    "cd ~/link/res_publish/run\n",
    "jupyter nbconvert info_model_gene.ipynb --to python\n",
    "\n",
    "conda activate publish\n",
    "nohup python info_model_gene.py > nohup_info_model_gene &\n",
    "sleep 30 && rm info_model_gene.py\n",
    "conda activate\n",
    "clear\n",
    "}\n",
    "jobs\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "afea8487-7cdd-4f1f-901d-1adb1e3aadb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "p_root = Path('~/link/res_publish').expanduser()\n",
    "None if str(p_root) in sys.path else sys.path.append(str(p_root))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8ee1846c-c62d-4465-9d10-4a2b4fe88756",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-08-19 02:05:17.713959: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-------------------------func_help-------------------------\n",
      "> parameter\n",
      "    p_root\t[name] res_publish\n",
      "        p_run, p_plot, p_res, p_cache, p_pdf\n",
      "    p_df_varmap\n",
      "    map_sp_reverse\n",
      "    rng\n",
      "> run\n",
      "    run_cross_species_models\n",
      "    h5ad_to_mtx\n",
      "    load_adata\n",
      "    get_path_varmap\n",
      "    find_path_from_para\n",
      "    load_normalized_adata\n",
      "\n",
      "> res\n",
      "    get_test_result_df\n",
      "    get_res_obs\n",
      "    get_adata_umap\n",
      "    show_umap\n",
      "\n",
      "> plot\n",
      "    get_color_map\n",
      "    show_color_map\n",
      "    show_color\n",
      "    plot_umap\n",
      "    savefig\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from func import *\n",
    "\n",
    "import utils as ut\n",
    "from utils.general import *\n",
    "pl = ut.pl\n",
    "\n",
    "p_res_model_gene  = p_res.joinpath('model_gene')\n",
    "p_res_model_gene.mkdir(exist_ok=True,parents=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5e5321a0-cd5c-413d-bab4-b72958fd5427",
   "metadata": {},
   "outputs": [],
   "source": [
    "Tensor = csMAHN.pp.Tensor\n",
    "torch = csMAHN.pp.torch\n",
    "with Block(\"run_csMAHN_info_model_gene\"):\n",
    "    def run_csMAHN_info_model_gene(\n",
    "        path_adata1,\n",
    "        path_adata2,\n",
    "        key_class1,\n",
    "        key_class2,\n",
    "        sp1,\n",
    "        sp2,\n",
    "        tissue_name,\n",
    "        path_varmap,\n",
    "        aligned=False,\n",
    "        resdir_tag=\".\",\n",
    "        resdir=Path('.'),\n",
    "        limite_func=lambda adata1, adata2: (adata1, adata2),\n",
    "        **kvargs\n",
    "    ):\n",
    "        \"\"\"\n",
    "        version = 0.0.9\n",
    "        kvargs:\n",
    "            n_epochs:\n",
    "                default,[100, 200, 300]\n",
    "                stages,即res_0,res_1，res_2 的 epochs\n",
    "                累加制，res_0,res_1，res_2,实际epochs分别为100,300,600\n",
    "                故最终epochs为stages之和\n",
    "                stages = kvargs.setdefault(\"n_epochs\",[100, 200, 300])\n",
    "    \n",
    "            is_1v1: bool\n",
    "                default,False\n",
    "            n_hvgs:\n",
    "                default,2000\n",
    "            n_degs:\n",
    "                default,50\n",
    "        \"\"\"\n",
    "        homo_method = 'biomart'\n",
    "        n_hvgs = kvargs.setdefault('n_hvgs', 2000)\n",
    "        n_degs = kvargs.setdefault('n_degs', 50)\n",
    "        seed = 123\n",
    "        stages = kvargs.setdefault(\n",
    "            'n_epochs', [\n",
    "                100, 200, 200])  # [200, 200, 200]\n",
    "        nfeats = kvargs.setdefault('nfeats', 64)  # 64  # embedding size #128\n",
    "        hidden = kvargs.setdefault('hidden', 64)  # 64  # 128\n",
    "        input_drop = 0.2\n",
    "        att_drop = 0.2\n",
    "        residual = True\n",
    "    \n",
    "        threshold = 0.9  # 0.8\n",
    "        lr = 0.01  # lr = 0.01\n",
    "        weight_decay = 0.001\n",
    "        patience = 100\n",
    "        enhance_gama = 10\n",
    "        simi_gama = 0.1\n",
    "    \n",
    "        dsnames = (\n",
    "            '{}_{}'.format(\n",
    "                tissue_name, sp1), '{}_{}'.format(\n",
    "                tissue_name, sp2))\n",
    "        assert key_class1 == key_class2, \"key_class is not equal\"\n",
    "        key_class = key_class1\n",
    "    \n",
    "        # make file to save\n",
    "        resdir_tag = \"{}_{}-corss-{};{}\".format(tissue_name, sp1, sp2, resdir_tag) if len(\n",
    "            resdir_tag) > 0 else \"{}_{}-corss-{}\".format(tissue_name, sp1, sp2)\n",
    "        # curdir = os.path.join()\n",
    "        resdir = Path(resdir).joinpath(resdir_tag)\n",
    "        model_dir = resdir.joinpath('model_')\n",
    "        figdir = resdir.joinpath('figs')\n",
    "        \n",
    "        resdir.mkdir(exist_ok=True, parents=True)\n",
    "        # [_.mkdir(exist_ok=True, parents=True)\n",
    "        #  for _ in [resdir, model_dir, figdir]]\n",
    "        # [resdir.joinpath('res_{}'.format(i)).mkdir(\n",
    "        #     exist_ok=True, parents=True) for i in range(len(stages))]\n",
    "    \n",
    "        checkpt_file = model_dir.joinpath(\"mutistages\")\n",
    "    \n",
    "        # is finish\n",
    "        # p_finish = Path(resdir).joinpath(\"finish\")\n",
    "        # if p_finish.exists():\n",
    "        if resdir.joinpath('info_mode_gene.json').exists() and \n",
    "        np.setdiff1d(\n",
    "            \"reference_hvgs,reference_degs,reference_higs,query_hvgs,query_degs,query_higs,reference_gene_nodes,query_gene_nodes,ortholog_one2one,ortholog_one2many,ortholog_many2many\".split(','),\n",
    "            list(json.loads(resdir.joinpath('info_mode_gene.json').read_text()).keys())\n",
    "        ).size == 0 :\n",
    "            print(\n",
    "                \"[has finish]{} {}\".format(\n",
    "                    time.strftime('%y%m%d-%H%M', time.localtime()), resdir.name\n",
    "                ))\n",
    "            return\n",
    "        else:\n",
    "            pass\n",
    "        print(\n",
    "            \"[start]{} {}\".format(\n",
    "                time.strftime('%y%m%d-%H%M', time.localtime()),\n",
    "                resdir.name\n",
    "            ))\n",
    "    \n",
    "        finish_content = [\"[strat] {}\".format(time.time())]\n",
    "        print('[path_varmap] {}'.format(path_varmap))\n",
    "        adata_raw1 = load_adata(path_adata1)\n",
    "        adata_raw2 = load_adata(path_adata2)\n",
    "        if key_class not in adata_raw2.obs.columns:\n",
    "            adata_raw2.obs[key_class] = ''\n",
    "        adata_raw1.obs[key_class] = adata_raw1.obs[key_class].astype(str)\n",
    "        adata_raw2.obs[key_class] = adata_raw2.obs[key_class].astype(str)\n",
    "        # limite 进一步对adata进行限制，默认不操作直接返回\n",
    "        adata_raw1, adata_raw2 = limite_func(\n",
    "            adata_raw1, adata_raw2\n",
    "        )\n",
    "        # group_counts_unalign.csv\n",
    "        pd.concat([adata_raw1.obs[key_class].value_counts(),\n",
    "                   adata_raw2.obs[key_class].value_counts(),],\n",
    "                  axis=1, keys=dsnames,)\n",
    "        # .to_csv(\n",
    "        #     resdir.joinpath(\"group_counts_unalign.csv\"), index=True\n",
    "        # )\n",
    "        # 仅保留公共细胞类群\n",
    "        if aligned:\n",
    "            adata_raw1, adata_raw2 = csMAHN.pp.aligned_type(\n",
    "                [adata_raw1, adata_raw2], key_class\n",
    "            )\n",
    "    \n",
    "        # group_counts.csv\n",
    "        temp = pd.concat([adata_raw1.obs[key_class].value_counts(),\n",
    "                          adata_raw2.obs[key_class].value_counts(),],\n",
    "                         axis=1, keys=dsnames)\n",
    "        print(temp)\n",
    "        # temp.to_csv(resdir.joinpath(\"group_counts.csv\"), index=True)\n",
    "        # adata_raw1.obs.to_csv(resdir.joinpath(\"obs_ref.csv\"), index=True)\n",
    "        # adata_raw2.obs.to_csv(resdir.joinpath(\"obs_que.csv\"), index=True)\n",
    "    \n",
    "        # homo = pd.read_csv(path_varmap)\n",
    "        homo = pd.read_csv(path_varmap, usecols=range(3))\n",
    "        homo.columns = [\"gn_ref\", \"gn_que\", \"homology_type\"]\n",
    "        if kvargs.setdefault(\"is_1v1\", False):\n",
    "            homo = get_1v1_matches(homo)\n",
    "            homology_parameter = get_homology_parameters(\n",
    "                adata_raw1, adata_raw2, homo)\n",
    "            print(\"\"\"\n",
    "    [homology one2one]find {homology_one2one_find} genes\n",
    "    [homology one2one]use {homology_one2one_use} genes\"\"\".format(\n",
    "                **homology_parameter))\n",
    "            kvargs.update(homology_parameter)\n",
    "    \n",
    "        kvargs.update({'path_adata1': str(path_adata1),\n",
    "                       'path_adata2': str(path_adata2),\n",
    "                       'key_class1': key_class1,\n",
    "                       'key_class2': key_class2,\n",
    "                       'sp1': sp1,\n",
    "                       'sp2': sp2,\n",
    "                       'tissue_name': tissue_name,\n",
    "                       'path_varmap': str(path_varmap),\n",
    "                       'aligned': aligned,\n",
    "                       'resdir_tag': resdir_tag,\n",
    "                       'resdir': str(resdir),\n",
    "                      'n_hvgs': n_hvgs,\n",
    "                       'n_degs': n_degs,\n",
    "                       'nfeats': nfeats,\n",
    "                       'hidden': hidden\n",
    "                       })\n",
    "        resdir.joinpath(\"kvargs.json\").write_text(dumps(kvargs))\n",
    "        print(\n",
    "            \"\"\"Task: refernece:{} {} cells x {} gene -> query:{} {} cells x {} gene in {}\"\"\".format(\n",
    "                dsnames[0],\n",
    "                adata_raw1.shape[0],\n",
    "                adata_raw1.shape[1],\n",
    "                dsnames[1],\n",
    "                adata_raw2.shape[0],\n",
    "                adata_raw2.shape[1],\n",
    "                tissue_name))\n",
    "    \n",
    "        start = time.time()\n",
    "        finish_content.append(\"[finish before run] {}\".format(time.time()))\n",
    "        # knn时间较长\n",
    "        print(\"\\n[process_for_graph]\\n\".center(100, '-'))\n",
    "        adatas, features_genes, nodes_genes, scnets, one2one, n2n = csMAHN.pp.process_for_graph(\n",
    "            [adata_raw1, adata_raw2], homo, key_class, 'leiden', n_hvgs=n_hvgs, n_degs=n_degs,\n",
    "            resdir=resdir)\n",
    "        g, inter_net, one2one_gene_nodes_net, cell_label, n_classes, list_idx = csMAHN.pp.make_graph(\n",
    "            adatas, aligned, key_class, features_genes, nodes_genes, scnets,\n",
    "            one2one, n2n, has_mnn=True, seed=seed,resdir=resdir)\n",
    "        end = time.time()\n",
    "        # 包括预处理时间\n",
    "        print('Times preprocess for graph:{:.2f}'.format(end - start))\n",
    "        print('\\n[end][run_csMAHN_info_model_gene]\\n'.center(100,'-'))\n",
    "\n",
    "with Block(\"[redefine] csMAHN.pp.get_adj_gene\"):\n",
    "    def _func(nodes_genes, homo,**kvarg):\n",
    "        \"\"\"\n",
    "        Outputs a gene aparse adjacent matrix\n",
    "        :param homo: genes homology relation (select one2one or many2many to control)\n",
    "        \"\"\"\n",
    "        reference_gene_nodes = nodes_genes[0]\n",
    "        query_gene_nodes = nodes_genes[1]\n",
    "        adj = csMAHN.pp.match_bigraph(reference_gene_nodes, query_gene_nodes, homo)\n",
    "        print(\"--------------homo edges---------------\")\n",
    "        print(adj.iloc[:, 2].value_counts())\n",
    "        \n",
    "        # [redefine][add]\n",
    "        with Block(\"info_mode_gene.json\",context=dict(\n",
    "            p_out = kvarg['resdir'].joinpath('info_mode_gene.json'),\n",
    "            data = {}\n",
    "        )) as context:\n",
    "            if context.p_out.exists():\n",
    "                context.data.update(json.loads(context.p_out.read_text()))\n",
    "            context.data.update(\n",
    "                adj.iloc[:, 2].value_counts().to_frame()\\\n",
    "                    .reset_index(names='value')\\\n",
    "                    .pipe(ut.df.to_dict,'value','count'))\n",
    "            context.p_out.write_text(json.dumps(context.data))\n",
    "    \n",
    "        gn = np.append(reference_gene_nodes, query_gene_nodes)\n",
    "        gnind_1 = pd.DataFrame(data=np.arange(reference_gene_nodes.size)[None, :], columns=reference_gene_nodes)\n",
    "        gnind_2 = pd.DataFrame(data=np.arange(gn.size)[None, :], columns=gn)\n",
    "        gnind_2 = gnind_2.iloc[:, len(reference_gene_nodes):]\n",
    "        a = np.array(adj.iloc[:, 0]).tolist()\n",
    "        b = np.array(adj.iloc[:, 1]).tolist()\n",
    "        reference_indic = pd.Index(gnind_1[a].values.flatten())\n",
    "        query_indic = pd.Index(gnind_2[b].values.flatten())\n",
    "        indice1 = np.vstack((reference_indic, query_indic))\n",
    "        indice2 = np.vstack((query_indic, reference_indic))\n",
    "        gene_adj = np.concatenate((indice1, indice2), axis=1)\n",
    "    \n",
    "        return torch.LongTensor(gene_adj)\n",
    "\n",
    "    csMAHN.pp.get_adj_gene = _func\n",
    "    del _func\n",
    "\n",
    "with Block(\"[redefine] csMAHN.pp.select_gene_nodes\"):\n",
    "    def _func(reference_all_gene, query_all_gene, higs_reference, higs_query, homo,**kvarg):\n",
    "        # 物种1的higs与物种2的higs在物种1的同源 联合\n",
    "        query_homo = homo[homo.iloc[:, 1].isin(higs_query)].iloc[:, 0]\n",
    "        reference_gene_nodes = np.union1d(query_homo, higs_reference)\n",
    "    \n",
    "        reference_homo = homo[homo.iloc[:, 0].isin(higs_reference)].iloc[:, 1]\n",
    "        query_gene_nodes = np.union1d(reference_homo, higs_query)\n",
    "    \n",
    "        # biomart和数据集基因名的一致性筛选\n",
    "        reference_gene_nodes = np.intersect1d(reference_all_gene, reference_gene_nodes)\n",
    "        query_gene_nodes = np.intersect1d(query_all_gene, query_gene_nodes)\n",
    "        print(\"--------------gene nodes info---------------\")\n",
    "        print(\"num of reference_gene_node is {0}\".format(len(reference_gene_nodes)))\n",
    "        print(\"num of query_gene_node is {0}\".format(len(query_gene_nodes)))\n",
    "        \n",
    "        # [redefine][add]\n",
    "        with Block(\"info_mode_gene.json\",context=dict(\n",
    "            p_out = kvarg['resdir'].joinpath('info_mode_gene.json'),\n",
    "            data = {}\n",
    "        )) as context:\n",
    "            if context.p_out.exists():\n",
    "                context.data.update(json.loads(context.p_out.read_text()))\n",
    "            context.data.update(\n",
    "                reference_gene_nodes=len(reference_gene_nodes),\n",
    "                query_gene_nodes=len(query_gene_nodes)\n",
    "                )\n",
    "            context.p_out.write_text(json.dumps(context.data))\n",
    "\n",
    "        return reference_gene_nodes, query_gene_nodes\n",
    "    \n",
    "    csMAHN.pp.select_gene_nodes = _func\n",
    "    del _func\n",
    "    \n",
    "with Block(\"[redefine] csMAHN.pp.process_for_graph\"):\n",
    "    def _func(adatas,\n",
    "                  homo,\n",
    "                  key_class,\n",
    "                  key_clust,\n",
    "                  relation='many',\n",
    "                  add_no1v1=True,\n",
    "                  n_pcs=30,\n",
    "                  n_hvgs=2000,\n",
    "                  n_degs=50,\n",
    "                  use_scnets=True,\n",
    "                  n_neighbors_scnet=5,\n",
    "                  n_neighbors_clust=20,\n",
    "                  reso=0.4,\n",
    "                  degs_method=\"wilcoxon\",\n",
    "                  **kvarg\n",
    "                  ):\n",
    "        \n",
    "        print('*'*75,'\\n[redefind] csMAHN.pp.process_for_graph\\n{}\\n'.format(kvarg),'*'*75)\n",
    "        # process homo relation\n",
    "        one2one, n2n = csMAHN.pp.biomart_process(homo)\n",
    "        \n",
    "        # process for adata\n",
    "        params_preproc = dict(\n",
    "            copy=True,\n",
    "            target_sum=None,\n",
    "            n_hvgs=n_hvgs,\n",
    "            n_pcs=n_pcs,\n",
    "            n_neighbors=n_neighbors_scnet,\n",
    "        )\n",
    "        time1 = time.time()\n",
    "        adata1 = csMAHN.pp.preprocess_for_adata(adatas[0], **params_preproc)\n",
    "        adata2 = csMAHN.pp.preprocess_for_adata(adatas[1], **params_preproc)\n",
    "        time2 = time.time()\n",
    "        print(f'the time2 of processing adatas is {time2-time1}')\n",
    "        if use_scnets:\n",
    "            scnets = [csMAHN.pp.get_intra_net_from_adata(adata1),\n",
    "                      csMAHN.pp.get_intra_net_from_adata(adata2)]\n",
    "            scnets[0].data[:] = 1\n",
    "            scnets[1].data[:] = 1\n",
    "        else:\n",
    "            scnets = None\n",
    "        \n",
    "        reference_hvgs, query_hvgs = csMAHN.pp.get_hvgs(adata1), csMAHN.pp.get_hvgs(adata2)\n",
    "        \n",
    "        time1 = time.time()\n",
    "        # 或许需要重新knn\n",
    "        clust_lbs2 = csMAHN.pp.get_leiden_labels(adata2,\n",
    "                                       n_neighbors=n_neighbors_clust,\n",
    "                                       reso=reso,\n",
    "                                       neighbors_key='clust',\n",
    "                                       key_added='leiden',\n",
    "                                       copy=False)\n",
    "        time2 = time.time()\n",
    "        print(f'the time of leiden is {time2 - time1}')\n",
    "        \n",
    "        \n",
    "        adatas[1].obs[key_clust] = clust_lbs2\n",
    "        \n",
    "        reference_degs = csMAHN.pp.get_degs(adata1, groupby=key_class, n_degs=n_degs, method=degs_method)\n",
    "        query_degs = csMAHN.pp.get_degs(adata2, groupby=key_clust, n_degs=n_degs, method=degs_method)\n",
    "        time3 = time.time()\n",
    "        print(f'the time of degs is {time3 - time2}')\n",
    "        reference_higs = np.union1d(reference_hvgs, reference_degs)\n",
    "        query_higs = np.union1d(query_hvgs, query_degs)\n",
    "        _adatas = [adata1, adata2]\n",
    "        \n",
    "        print(\"--------------hvgs, degs info---------------\")\n",
    "        print(\"num of reference_hvgs,reference_degs,reference_higs are {0},{1},{2}\".format(len(reference_hvgs),\n",
    "                                                                                           len(reference_degs),\n",
    "                                                                                           len(reference_higs)))\n",
    "        print(\"num of query_hvgs,query_degs,query_higs are {0},{1},{2}\".format(len(query_hvgs), len(query_degs),\n",
    "                                                                                   len(query_higs)))\n",
    "        # [redefine][add]\n",
    "        with Block(\"info_mode_gene.json\",context=dict(\n",
    "            p_out = kvarg['resdir'].joinpath('info_mode_gene.json'),\n",
    "            data = {}\n",
    "        )) as context:\n",
    "\n",
    "\n",
    "            if context.p_out.exists():\n",
    "                context.data.update(json.loads(context.p_out.read_text()))\n",
    "            context.data.update(\n",
    "                reference_hvgs=len(reference_hvgs),\n",
    "                reference_degs=len(reference_degs),\n",
    "                reference_higs=len(reference_higs),\n",
    "                query_hvgs=len(query_hvgs),\n",
    "                query_degs=len(query_degs),\n",
    "                query_higs=len(query_higs),\n",
    "                )\n",
    "            context.p_out.write_text(json.dumps(context.data))\n",
    "        \n",
    "        # nodes gene\n",
    "        # 使用one2one,n2n选择加入的同源关系类型\n",
    "        gene_relation = n2n if relation == 'many' else one2one\n",
    "        reference_all_gene, query_all_gene = adata1.raw.var.index.tolist(), adata2.raw.var.index.tolist()\n",
    "        reference_gene_nodes, query_gene_nodes = csMAHN.pp.select_gene_nodes(reference_all_gene, query_all_gene,\n",
    "                                                                   reference_higs, query_higs, gene_relation,\n",
    "                                                                             **kvarg)\n",
    "        nodes_genes = [reference_gene_nodes, query_gene_nodes]\n",
    "        # feature gene\n",
    "        features_genes = csMAHN.pp.get_feature_genes(reference_all_gene, query_all_gene, reference_degs, query_degs, one2one , add_no1v1=add_no1v1, n2n=n2n)\n",
    "        \n",
    "        return _adatas, features_genes, nodes_genes, scnets, one2one, n2n\n",
    "\n",
    "    csMAHN.pp.process_for_graph = _func\n",
    "    del _func\n",
    "\n",
    "with Block(\"[redefine] csMAHN.pp.make_graph\"):\n",
    "    def _func(adatas,\n",
    "                   aligned,\n",
    "                   key_class,\n",
    "                   features_genes,\n",
    "                   nodes_genes,\n",
    "                   scnets,\n",
    "                   one2one, n2n,\n",
    "                   gene_embedding=None,\n",
    "                   has_mnn=False,\n",
    "                   graph_type='dgl',\n",
    "                   seed=123,**kvarg):\n",
    "        \"\"\"\n",
    "        Load cell expression matrix ,cell label, homo information to camputa cell hvgs, degs\n",
    "        adata should be normalized\n",
    "        \"\"\"\n",
    "    \n",
    "        # normalized and log\n",
    "        norm_counts = csMAHN.pp.get_counts_from_adatas(adatas)\n",
    "        # scale\n",
    "    \n",
    "        # def _zscore(adata):\n",
    "        #     \"\"\"\n",
    "        #     Adata has been normalized\n",
    "        #     \"\"\"\n",
    "        #     _adata = adata.copy()\n",
    "        #     sc.pp.scale(_adata)\n",
    "        #     return pd.DataFrame(_adata.X, columns=_adata.var.index)\n",
    "        #\n",
    "        # scale_counts = [_zscore(adata) for adata in adatas]\n",
    "        reference_labels, query_labels = csMAHN.pp.get_labels_from_adatas(adatas, key_class)\n",
    "    \n",
    "        cell_feature = csMAHN.pp.get_feature_counts(norm_counts, features_genes)\n",
    "        cell_label = csMAHN.pp.get_labelEncoder(reference_labels, query_labels)\n",
    "        if aligned:\n",
    "            n_classes = max(cell_label) + 1\n",
    "        else:\n",
    "            n_classes = len(np.unique(adatas[0].obs[key_class]))\n",
    "        train_idx, val_idx, pred_idx = csMAHN.pp.get_idx_cross_classes(reference_labels, query_labels)\n",
    "        list_idx = [train_idx, val_idx, pred_idx]\n",
    "    \n",
    "        def _get_nums_gene(data):\n",
    "            return len(data)\n",
    "    \n",
    "        def _get_nums_feature_from_cell(data: Tensor):\n",
    "            return data.shape[1]\n",
    "    \n",
    "        all_gene_nums = _get_nums_gene(nodes_genes[0]) + _get_nums_gene(nodes_genes[1])\n",
    "        cell_feature_nums = _get_nums_feature_from_cell(cell_feature)\n",
    "        gene_feature = torch.Tensor(csMAHN.pp.get_gene_feature(all_gene_nums, cell_feature_nums, gene_embedding))\n",
    "    \n",
    "        cell_gene_adj, gene_cell_adj = csMAHN.pp.get_adj_cell_gene(norm_counts[0], norm_counts[1], nodes_genes[0], nodes_genes[1])\n",
    "        gene_adj = csMAHN.pp.get_adj_gene(nodes_genes, n2n,**kvarg)\n",
    "        one2one_gene_nodes_net = csMAHN.pp.get_adj_gene(nodes_genes, one2one,**kvarg)\n",
    "    \n",
    "        inter_net = csMAHN.pp.get_pair_from_adatas(norm_counts, features_genes, N1=10, N2=10, N=5, n_jobs=32)\n",
    "        acc_inter_net = (cell_label[inter_net[0]] == cell_label[inter_net[1]]).sum() / len(cell_label[inter_net[0]])\n",
    "        print('Inter-net pairs ACC: {0}'.format(acc_inter_net))\n",
    "        cell_adj = csMAHN.pp.get_adj_cell(scnets, inter_net) if has_mnn else get_adj_cell(scnets)\n",
    "    \n",
    "        # cell_adj = get_adj_cell\n",
    "        \n",
    "        # print graph info\n",
    "        print(\"-------------------nodes info-------------------\")\n",
    "        print(f'the num of cell feats is {cell_feature_nums}')\n",
    "        print(f'the num of cell nodes is {cell_feature.shape[0]}')\n",
    "        print(f'the num of gene nodes is {all_gene_nums}')\n",
    "        dct = dict(\n",
    "            cell_feature=cell_feature,\n",
    "            cell_label=cell_label,\n",
    "            list_idx=list_idx,\n",
    "            gene_feature=gene_feature,\n",
    "    \n",
    "            adj_gg=gene_adj,\n",
    "            adj_cc=cell_adj,\n",
    "            adj_gc=gene_cell_adj,\n",
    "            adj_cg=cell_gene_adj,\n",
    "        )\n",
    "        # if graph_type == 'pyg':\n",
    "        #     g = get_pyg(dct)\n",
    "        # elif graph_type == 'dgl':\n",
    "        #     g = get_dgl(dct)\n",
    "        g = csMAHN.pp.get_dgl(dct)\n",
    "        return g, inter_net, one2one_gene_nodes_net, cell_label, n_classes, list_idx\n",
    "\n",
    "    csMAHN.pp.make_graph = _func\n",
    "    del _func"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ca12d71e-066e-4dc7-aed5-31d68a4ee80b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>sp_ref</th>\n",
       "      <th>path_ref</th>\n",
       "      <th>name_ref</th>\n",
       "      <th>sp_simple_ref</th>\n",
       "      <th>sp_que</th>\n",
       "      <th>path_que</th>\n",
       "      <th>name_que</th>\n",
       "      <th>sp_simple_que</th>\n",
       "      <th>key_cell_type</th>\n",
       "      <th>mask</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Adrenal-Gland;h_adr;m_adr</th>\n",
       "      <td>Adrenal-Gland</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_adr</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_adr</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bone-Marrow;h_bon;m_bon</th>\n",
       "      <td>Bone-Marrow</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_bon</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_bon</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  tissue sp_ref  \\\n",
       "Adrenal-Gland;h_adr;m_adr  Adrenal-Gland  human   \n",
       "Bone-Marrow;h_bon;m_bon      Bone-Marrow  human   \n",
       "\n",
       "                                                                    path_ref  \\\n",
       "Adrenal-Gland;h_adr;m_adr  /public/workspace/licanchengup/link/res_publis...   \n",
       "Bone-Marrow;h_bon;m_bon    /public/workspace/licanchengup/link/res_publis...   \n",
       "\n",
       "                          name_ref sp_simple_ref sp_que  \\\n",
       "Adrenal-Gland;h_adr;m_adr    h_adr             h  mouse   \n",
       "Bone-Marrow;h_bon;m_bon      h_bon             h  mouse   \n",
       "\n",
       "                                                                    path_que  \\\n",
       "Adrenal-Gland;h_adr;m_adr  /public/workspace/licanchengup/link/res_publis...   \n",
       "Bone-Marrow;h_bon;m_bon    /public/workspace/licanchengup/link/res_publis...   \n",
       "\n",
       "                          name_que sp_simple_que key_cell_type     mask  \n",
       "Adrenal-Gland;h_adr;m_adr    m_adr             m            CL  HCL_MCA  \n",
       "Bone-Marrow;h_bon;m_bon      m_bon             m            CL  HCL_MCA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "(44, 11)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_para = pd.concat([\n",
    "     pd.concat([pd.read_csv(p_cache.joinpath(\n",
    "        'parameter_healthy_{}.csv'.format(i))).assign(mask=i)\n",
    "        for i in 'HCL_MCA,Retina,Retina-parameters'.split(',')]),\n",
    "    pd.concat([pd.read_csv(p_cache.joinpath(\n",
    "        'parameter_{}.csv'.format(i)\n",
    "    )).assign(mask=i) for i in 'LC,RA,Mtb,LN,COVID'.split(',')])\n",
    "])\n",
    "\n",
    "df_para['path_ref'] = df_para['path_ref'].apply(lambda x:p_cache.joinpath(x))\n",
    "df_para['path_que'] = df_para['path_que'].apply(lambda x:p_cache.joinpath(x))\n",
    "df_para['item'] = ut.df.apply_merge_field(df_para,'{tissue};{name_ref};{name_que}')\n",
    "df_para = ut.df.reindex_with_unique_col(df_para,'item',drop=True)\n",
    "ut.df.show(df_para)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5f3125da-e4a0-4557-9385-2d1fa571376d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mask\n",
       "HCL_MCA              18\n",
       "Retina-parameters     6\n",
       "LN                    5\n",
       "Mtb                   5\n",
       "LC                    4\n",
       "Retina                3\n",
       "COVID                 2\n",
       "RA                    1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_para['mask'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55f3d9cb-63af-4bc1-a264-b7f68b2c4029",
   "metadata": {},
   "source": [
    "# HCL_MCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "72ba1cb6-615c-4d1e-aca5-fdfe10c6d4ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_para_item['path_ref'].apply(lambda x:x.joinpath('matrix.mtx').exists()).all()\n",
    "df_para_item['path_que'].apply(lambda x:x.joinpath('matrix.mtx').exists()).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d3e80847-18f9-4427-9653-db33c33e63ac",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[runstart] HCL_MCA\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish]240819-0218 Adrenal-Gland_h-corss-m;h_adr-map-m_adr;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish]240819-0218 Bone-Marrow_h-corss-m;h_bon-map-m_bon;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0218 Brain_h-corss-m;h_bra-map-m_bra;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                  reference    query\n",
      "CL                                  \n",
      "neural cell         15645.0  26621.0\n",
      "oligodendrocyte      7546.0  11170.0\n",
      "macrophage           6633.0  12633.0\n",
      "astrocyte            6143.0      NaN\n",
      "stem cell            4596.0   7878.0\n",
      "fibroblast           1382.0      NaN\n",
      "granule cell          686.0      NaN\n",
      "endothelial cell      386.0   2264.0\n",
      "T cell                  NaN  13425.0\n",
      "ciliated cell           NaN   6732.0\n",
      "neutrophil              NaN   2135.0\n",
      "----new----\n",
      "                  reference    query\n",
      "CL                                  \n",
      "neural cell         15645.0  26621.0\n",
      "oligodendrocyte      7546.0  11170.0\n",
      "macrophage           6633.0  12633.0\n",
      "stem cell            4596.0   7878.0\n",
      "endothelial cell      386.0   2264.0\n",
      "                  Brain_h  Brain_m\n",
      "CL                                \n",
      "neural cell         15645    26621\n",
      "oligodendrocyte      7546    11170\n",
      "macrophage           6633    12633\n",
      "stem cell            4596     7878\n",
      "endothelial cell      386     2264\n",
      "Task: refernece:Brain_h 34806 cells x 11016 gene -> query:Brain_m 60566 cells x 11100 gene in Brain\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Brain_h-corss-m;h_bra-map-m_bra;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 17.095465183258057\n",
      "Leiden results:\n",
      "leiden\n",
      "0     10573\n",
      "1     10021\n",
      "2      7048\n",
      "3      5422\n",
      "4      5196\n",
      "5      3356\n",
      "6      3153\n",
      "7      2773\n",
      "8      2151\n",
      "9      1766\n",
      "10     1616\n",
      "11     1458\n",
      "12     1316\n",
      "13     1202\n",
      "14      771\n",
      "15      696\n",
      "16      689\n",
      "17      503\n",
      "18      385\n",
      "19      354\n",
      "20      117\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 18.82611918449402\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 158.77962279319763\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,242,2084\n",
      "num of query_hvgs,query_degs,query_higs are 2000,628,2313\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 2865\n",
      "num of query_gene_node is 3112\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2249\n",
      "ortholog_one2many      112\n",
      "ortholog_many2many      34\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2249\n",
      "Name: count, dtype: int64\n",
      "knn time is 192.33667135238647 s\n",
      "mnn time is 0.083251953125 s\n",
      "the time of compute mnn is  194.8939187526703 s\n",
      "Inter-net pairs ACC: 0.8554301833568406\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 517\n",
      "the num of cell nodes is 95372\n",
      "the num of gene nodes is 5977\n",
      "Times preprocess for graph:533.74\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0228 Heart_h-corss-m;h_hea-map-m_hea;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                     reference  query\n",
      "CL                                   \n",
      "cardiac muscle cell     5928.0   2890\n",
      "mesenchymal cell        3129.0  50417\n",
      "myeloid cell            1120.0  11505\n",
      "epithelial cell          606.0   3865\n",
      "T cell                     NaN   6003\n",
      "----new----\n",
      "                     reference  query\n",
      "CL                                   \n",
      "cardiac muscle cell     5928.0   2890\n",
      "mesenchymal cell        3129.0  50417\n",
      "myeloid cell            1120.0  11505\n",
      "epithelial cell          606.0   3865\n",
      "                     Heart_h  Heart_m\n",
      "CL                                   \n",
      "cardiac muscle cell     5928     2890\n",
      "mesenchymal cell        3129    50417\n",
      "myeloid cell            1120    11505\n",
      "epithelial cell          606     3865\n",
      "Task: refernece:Heart_h 10783 cells x 11069 gene -> query:Heart_m 68677 cells x 12528 gene in Heart\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Heart_h-corss-m;h_hea-map-m_hea;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 10.486703634262085\n",
      "Leiden results:\n",
      "leiden\n",
      "0     14414\n",
      "1      9119\n",
      "2      7755\n",
      "3      7351\n",
      "4      5407\n",
      "5      4403\n",
      "6      3833\n",
      "7      2589\n",
      "8      2543\n",
      "9      2517\n",
      "10     2069\n",
      "11     1684\n",
      "12     1243\n",
      "13     1231\n",
      "14     1063\n",
      "15      665\n",
      "16      355\n",
      "17      305\n",
      "18      131\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 21.714441537857056\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 72.27946710586548\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,180,2085\n",
      "num of query_hvgs,query_degs,query_higs are 2000,538,2217\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 2912\n",
      "num of query_gene_node is 3292\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2302\n",
      "ortholog_one2many      165\n",
      "ortholog_many2many      65\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2302\n",
      "Name: count, dtype: int64\n",
      "knn time is 59.06409764289856 s\n",
      "mnn time is 0.085113525390625 s\n",
      "the time of compute mnn is  62.03797674179077 s\n",
      "Inter-net pairs ACC: 0.7425983977708116\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 428\n",
      "the num of cell nodes is 79460\n",
      "the num of gene nodes is 6204\n",
      "Times preprocess for graph:252.97\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0232 Intestine_h-corss-m;h_int-map-m_int;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                      reference    query\n",
      "CL                                      \n",
      "enterocyte              36272.0  67108.0\n",
      "B cell                  12604.0  12316.0\n",
      "macrophage               8243.0   4208.0\n",
      "stromal cell             6429.0      NaN\n",
      "fibroblast               4843.0      NaN\n",
      "endothelial cell         1227.0      NaN\n",
      "enteroendocrine cell        NaN   1749.0\n",
      "epithelial cell             NaN    672.0\n",
      "tuft cell of colon          NaN    510.0\n",
      "----new----\n",
      "            reference    query\n",
      "CL                            \n",
      "enterocyte    36272.0  67108.0\n",
      "B cell        12604.0  12316.0\n",
      "macrophage     8243.0   4208.0\n",
      "            Intestine_h  Intestine_m\n",
      "CL                                  \n",
      "enterocyte        36272        67108\n",
      "B cell            12604        12316\n",
      "macrophage         8243         4208\n",
      "Task: refernece:Intestine_h 57119 cells x 10208 gene -> query:Intestine_m 83632 cells x 10665 gene in Intestine\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Intestine_h-corss-m;h_int-map-m_int;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 23.77963137626648\n",
      "Leiden results:\n",
      "leiden\n",
      "0     16472\n",
      "1     11197\n",
      "2     11038\n",
      "3      5430\n",
      "4      5406\n",
      "5      5190\n",
      "6      4528\n",
      "7      4411\n",
      "8      4127\n",
      "9      3075\n",
      "10     2793\n",
      "11     2709\n",
      "12     2509\n",
      "13     2177\n",
      "14      922\n",
      "15      888\n",
      "16      760\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 40.31179070472717\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 114.65888285636902\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,142,2077\n",
      "num of query_hvgs,query_degs,query_higs are 2000,516,2228\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 2995\n",
      "num of query_gene_node is 3228\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2253\n",
      "ortholog_one2many      152\n",
      "ortholog_many2many      46\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2253\n",
      "Name: count, dtype: int64\n",
      "knn time is 472.9138283729553 s\n",
      "mnn time is 0.137282133102417 s\n",
      "the time of compute mnn is  476.55330896377563 s\n",
      "Inter-net pairs ACC: 0.8910256410256411\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 381\n",
      "the num of cell nodes is 140751\n",
      "the num of gene nodes is 6223\n",
      "Times preprocess for graph:978.19\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0249 Kidney_h-corss-m;h_kid-map-m_kid;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                                                    reference    query\n",
      "CL                                                                    \n",
      "mesenchymal cell                                      23683.0  19649.0\n",
      "kidney tubule cell                                     7482.0  17060.0\n",
      "endothelial cell                                       5147.0   3716.0\n",
      "pancreatic ductal cell                                 4321.0   4754.0\n",
      "myeloid cell                                           2748.0   4431.0\n",
      "podocyte                                               1255.0      NaN\n",
      "epithelial cell                                         495.0    265.0\n",
      "kidney loop of Henle ascending limb epithelial ...        NaN  25916.0\n",
      "----new----\n",
      "                        reference    query\n",
      "CL                                        \n",
      "mesenchymal cell          23683.0  19649.0\n",
      "kidney tubule cell         7482.0  17060.0\n",
      "endothelial cell           5147.0   3716.0\n",
      "pancreatic ductal cell     4321.0   4754.0\n",
      "myeloid cell               2748.0   4431.0\n",
      "epithelial cell             495.0    265.0\n",
      "                        Kidney_h  Kidney_m\n",
      "CL                                        \n",
      "mesenchymal cell           23683     19649\n",
      "kidney tubule cell          7482     17060\n",
      "endothelial cell            5147      3716\n",
      "pancreatic ductal cell      4321      4754\n",
      "myeloid cell                2748      4431\n",
      "epithelial cell              495       265\n",
      "Task: refernece:Kidney_h 43876 cells x 13592 gene -> query:Kidney_m 49875 cells x 12846 gene in Kidney\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Kidney_h-corss-m;h_kid-map-m_kid;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 19.807610273361206\n",
      "Leiden results:\n",
      "leiden\n",
      "0     5744\n",
      "1     5080\n",
      "2     4872\n",
      "3     4732\n",
      "4     4542\n",
      "5     3928\n",
      "6     3755\n",
      "7     3275\n",
      "8     2904\n",
      "9     2267\n",
      "10    1904\n",
      "11    1793\n",
      "12    1780\n",
      "13    1567\n",
      "14     676\n",
      "15     511\n",
      "16     290\n",
      "17     255\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 15.803159236907959\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 91.08294034004211\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,259,2105\n",
      "num of query_hvgs,query_degs,query_higs are 2000,520,2196\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 2997\n",
      "num of query_gene_node is 2788\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2040\n",
      "ortholog_one2many      119\n",
      "ortholog_many2many      59\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2040\n",
      "Name: count, dtype: int64\n",
      "knn time is 165.26057028770447 s\n",
      "mnn time is 0.08535408973693848 s\n",
      "the time of compute mnn is  168.9568247795105 s\n",
      "Inter-net pairs ACC: 0.6558009864893846\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 459\n",
      "the num of cell nodes is 93751\n",
      "the num of gene nodes is 5785\n",
      "Times preprocess for graph:426.83\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0256 Liver_h-corss-m;h_liv-map-m_liv;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                  reference    query\n",
      "CL                                  \n",
      "erythrocyte         17319.0   7612.0\n",
      "mast cell            4628.0      NaN\n",
      "hepatocyte           3400.0   5268.0\n",
      "macrophage           2036.0   9342.0\n",
      "neutrophil            981.0  12548.0\n",
      "epithelial cell       425.0      NaN\n",
      "stem cell              15.0      NaN\n",
      "endothelial cell        NaN  20082.0\n",
      "B cell                  NaN   9060.0\n",
      "dendritic cell          NaN   8401.0\n",
      "fibroblast              NaN   1514.0\n",
      "----new----\n",
      "             reference    query\n",
      "CL                             \n",
      "erythrocyte    17319.0   7612.0\n",
      "hepatocyte      3400.0   5268.0\n",
      "macrophage      2036.0   9342.0\n",
      "neutrophil       981.0  12548.0\n",
      "             Liver_h  Liver_m\n",
      "CL                           \n",
      "erythrocyte    17319     7612\n",
      "hepatocyte      3400     5268\n",
      "macrophage      2036     9342\n",
      "neutrophil       981    12548\n",
      "Task: refernece:Liver_h 23736 cells x 10120 gene -> query:Liver_m 34770 cells x 11437 gene in Liver\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Liver_h-corss-m;h_liv-map-m_liv;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 15.4103684425354\n",
      "Leiden results:\n",
      "leiden\n",
      "0     5114\n",
      "1     4600\n",
      "2     4578\n",
      "3     3254\n",
      "4     2672\n",
      "5     1869\n",
      "6     1706\n",
      "7     1559\n",
      "8     1444\n",
      "9     1368\n",
      "10     986\n",
      "11     909\n",
      "12     828\n",
      "13     783\n",
      "14     648\n",
      "15     589\n",
      "16     538\n",
      "17     465\n",
      "18     236\n",
      "19     233\n",
      "20     229\n",
      "21     162\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 12.26138973236084\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 52.75525903701782\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,178,2074\n",
      "num of query_hvgs,query_degs,query_higs are 2000,510,2196\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 2939\n",
      "num of query_gene_node is 3136\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2214\n",
      "ortholog_one2many      164\n",
      "ortholog_many2many      92\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2214\n",
      "Name: count, dtype: int64\n",
      "knn time is 59.49287724494934 s\n",
      "mnn time is 0.060976505279541016 s\n",
      "the time of compute mnn is  61.99391055107117 s\n",
      "Inter-net pairs ACC: 0.8598692810457517\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 394\n",
      "the num of cell nodes is 58506\n",
      "the num of gene nodes is 6075\n",
      "Times preprocess for graph:188.88\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0259 Lung_h-corss-m;h_lun-map-m_lun;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                                            reference    query\n",
      "CL                                                            \n",
      "fibroblast                                       7573   8233.0\n",
      "alveolar capillary type 2 endothelial cell       6473   2972.0\n",
      "macrophage                                       5266   5659.0\n",
      "endothelial cell                                 3981      NaN\n",
      "mesenchymal cell                                 3293   8764.0\n",
      "neutrophil                                       2787  23726.0\n",
      "alveolar capillary type 1 endothelial cell       2192  15032.0\n",
      "T cell                                           1960      NaN\n",
      "----new----\n",
      "                                            reference    query\n",
      "CL                                                            \n",
      "fibroblast                                       7573   8233.0\n",
      "alveolar capillary type 2 endothelial cell       6473   2972.0\n",
      "macrophage                                       5266   5659.0\n",
      "mesenchymal cell                                 3293   8764.0\n",
      "neutrophil                                       2787  23726.0\n",
      "alveolar capillary type 1 endothelial cell       2192  15032.0\n",
      "                                            Lung_h  Lung_m\n",
      "CL                                                        \n",
      "fibroblast                                    7573    8233\n",
      "alveolar capillary type 2 endothelial cell    6473    2972\n",
      "macrophage                                    5266    5659\n",
      "mesenchymal cell                              3293    8764\n",
      "neutrophil                                    2787   23726\n",
      "alveolar capillary type 1 endothelial cell    2192   15032\n",
      "Task: refernece:Lung_h 27584 cells x 13874 gene -> query:Lung_m 64386 cells x 11102 gene in Lung\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Lung_h-corss-m;h_lun-map-m_lun;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 12.369564533233643\n",
      "Leiden results:\n",
      "leiden\n",
      "0     10097\n",
      "1      7084\n",
      "2      6203\n",
      "3      5742\n",
      "4      4647\n",
      "5      4354\n",
      "6      3525\n",
      "7      3107\n",
      "8      3008\n",
      "9      2924\n",
      "10     2856\n",
      "11     2317\n",
      "12     2059\n",
      "13     1916\n",
      "14     1352\n",
      "15     1346\n",
      "16     1341\n",
      "17      330\n",
      "18      128\n",
      "19       50\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 25.8231143951416\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 81.39916467666626\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,228,2114\n",
      "num of query_hvgs,query_degs,query_higs are 2000,602,2237\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3444\n",
      "num of query_gene_node is 3025\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2490\n",
      "ortholog_one2many      136\n",
      "ortholog_many2many      42\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2490\n",
      "Name: count, dtype: int64\n",
      "knn time is 146.6163272857666 s\n",
      "mnn time is 0.10057449340820312 s\n",
      "the time of compute mnn is  149.81316757202148 s\n",
      "Inter-net pairs ACC: 0.4270571827057183\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 495\n",
      "the num of cell nodes is 91970\n",
      "the num of gene nodes is 6469\n",
      "Times preprocess for graph:398.13\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start]240819-0306 Spleen_h-corss-m;h_spl-map-m_spl;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                  reference    query\n",
      "CL                                  \n",
      "B cell              10987.0   5481.0\n",
      "T cell               1698.0  11732.0\n",
      "neutrophil           1681.0   1935.0\n",
      "endothelial cell      748.0      NaN\n",
      "macrophage            578.0   3618.0\n",
      "erythrocyte           114.0   6963.0\n",
      "plasma cell             NaN   2791.0\n",
      "----new----\n",
      "             reference    query\n",
      "CL                             \n",
      "B cell         10987.0   5481.0\n",
      "T cell          1698.0  11732.0\n",
      "neutrophil      1681.0   1935.0\n",
      "macrophage       578.0   3618.0\n",
      "erythrocyte      114.0   6963.0\n",
      "             Spleen_h  Spleen_m\n",
      "CL                             \n",
      "B cell          10987      5481\n",
      "T cell           1698     11732\n",
      "neutrophil       1681      1935\n",
      "macrophage        578      3618\n",
      "erythrocyte       114      6963\n",
      "Task: refernece:Spleen_h 15058 cells x 17582 gene -> query:Spleen_m 29729 cells x 10866 gene in Spleen\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Spleen_h-corss-m;h_spl-map-m_spl;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 7.155946731567383\n",
      "Leiden results:\n",
      "leiden\n",
      "0     8202\n",
      "1     3858\n",
      "2     3666\n",
      "3     3621\n",
      "4     1742\n",
      "5     1564\n",
      "6     1425\n",
      "7     1199\n",
      "8      829\n",
      "9      781\n",
      "10     769\n",
      "11     742\n",
      "12     501\n",
      "13     371\n",
      "14     349\n",
      "15     110\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 8.421804428100586\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 37.18666362762451\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,221,2144\n",
      "num of query_hvgs,query_degs,query_higs are 2000,456,2189\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3180\n",
      "num of query_gene_node is 2484\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      1502\n",
      "ortholog_one2many      185\n",
      "ortholog_many2many     136\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    1502\n",
      "Name: count, dtype: int64\n",
      "knn time is 50.78779053688049 s\n",
      "mnn time is 0.05328011512756348 s\n",
      "the time of compute mnn is  52.11101007461548 s\n",
      "Inter-net pairs ACC: 0.6114141838019304\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 393\n",
      "the num of cell nodes is 44787\n",
      "the num of gene nodes is 5664\n",
      "Times preprocess for graph:137.28\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[start]240819-0309 Adrenal-Gland_m-corss-h;m_adr-map-h_adr;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                                reference    query\n",
      "CL                                                \n",
      "cortical cell of adrenal gland     6547.0  21097.0\n",
      "macrophage                         2080.0      NaN\n",
      "endothelial cell                   1876.0   4421.0\n",
      "dendritic cell                      714.0      NaN\n",
      "fibroblast                          533.0   4467.0\n",
      "chromaffin cell                      65.0    996.0\n",
      "neutrophil                            NaN   8094.0\n",
      "epithelial cell                       NaN   2928.0\n",
      "T cell                                NaN   1473.0\n",
      "----new----\n",
      "                                reference    query\n",
      "CL                                                \n",
      "cortical cell of adrenal gland     6547.0  21097.0\n",
      "endothelial cell                   1876.0   4421.0\n",
      "fibroblast                          533.0   4467.0\n",
      "chromaffin cell                      65.0    996.0\n",
      "                                Adrenal-Gland_m  Adrenal-Gland_h\n",
      "CL                                                              \n",
      "cortical cell of adrenal gland             6547            21097\n",
      "endothelial cell                           1876             4421\n",
      "fibroblast                                  533             4467\n",
      "chromaffin cell                              65              996\n",
      "Task: refernece:Adrenal-Gland_m 9021 cells x 17150 gene -> query:Adrenal-Gland_h 30981 cells x 12154 gene in Adrenal-Gland\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Adrenal-Gland_m-corss-h;m_adr-map-h_adr;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 8.075146675109863\n",
      "Leiden results:\n",
      "leiden\n",
      "0     5617\n",
      "1     4398\n",
      "2     3926\n",
      "3     3885\n",
      "4     2885\n",
      "5     2088\n",
      "6     1957\n",
      "7     1612\n",
      "8     1512\n",
      "9     1307\n",
      "10     592\n",
      "11     575\n",
      "12     540\n",
      "13      87\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 10.507326364517212\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 53.845629930496216\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,194,2055\n",
      "num of query_hvgs,query_degs,query_higs are 2000,394,2149\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3256\n",
      "num of query_gene_node is 2456\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      1915\n",
      "ortholog_one2many      112\n",
      "ortholog_many2many      81\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    1915\n",
      "Name: count, dtype: int64\n",
      "knn time is 35.23018026351929 s\n",
      "mnn time is 0.07409143447875977 s\n",
      "the time of compute mnn is  36.276020765304565 s\n",
      "Inter-net pairs ACC: 0.8871449925261584\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 357\n",
      "the num of cell nodes is 40002\n",
      "the num of gene nodes is 5712\n",
      "Times preprocess for graph:148.47\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[start]240819-0311 Bone-Marrow_m-corss-h;m_bon-map-h_bon;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                           reference    query\n",
      "CL                                           \n",
      "neutrophil                   23804.0  16685.0\n",
      "eosinophil                    8032.0      NaN\n",
      "erythrocyte                   5326.0   4813.0\n",
      "fibroblast                    2931.0      NaN\n",
      "macrophage                    2111.0   3394.0\n",
      "basophil                       693.0      NaN\n",
      "hematopoietic stem cell          NaN  20807.0\n",
      "common myeloid progenitor        NaN  20447.0\n",
      "B cell                           NaN   1237.0\n",
      "----new----\n",
      "             reference    query\n",
      "CL                             \n",
      "neutrophil     23804.0  16685.0\n",
      "erythrocyte     5326.0   4813.0\n",
      "macrophage      2111.0   3394.0\n",
      "             Bone-Marrow_m  Bone-Marrow_h\n",
      "CL                                       \n",
      "neutrophil           23804          16685\n",
      "erythrocyte           5326           4813\n",
      "macrophage            2111           3394\n",
      "Task: refernece:Bone-Marrow_m 31241 cells x 13315 gene -> query:Bone-Marrow_h 24892 cells x 10494 gene in Bone-Marrow\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Bone-Marrow_m-corss-h;m_bon-map-h_bon;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 10.514947414398193\n",
      "Leiden results:\n",
      "leiden\n",
      "0     5032\n",
      "1     3745\n",
      "2     3571\n",
      "3     3198\n",
      "4     2821\n",
      "5     2001\n",
      "6     1940\n",
      "7     1562\n",
      "8      755\n",
      "9      222\n",
      "10      45\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 7.840217590332031\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 50.23713421821594\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,150,2086\n",
      "num of query_hvgs,query_degs,query_higs are 2000,303,2198\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3378\n",
      "num of query_gene_node is 2853\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2249\n",
      "ortholog_one2many      130\n",
      "ortholog_many2many      86\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2249\n",
      "Name: count, dtype: int64\n",
      "knn time is 86.29372119903564 s\n",
      "mnn time is 0.04094362258911133 s\n",
      "the time of compute mnn is  87.50010299682617 s\n",
      "Inter-net pairs ACC: 0.9032393773664282\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 256\n",
      "the num of cell nodes is 56133\n",
      "the num of gene nodes is 6231\n",
      "Times preprocess for graph:211.09\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[start]240819-0315 Brain_m-corss-h;m_bra-map-h_bra;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                  reference    query\n",
      "CL                                  \n",
      "neural cell         26621.0  15645.0\n",
      "T cell              13425.0      NaN\n",
      "macrophage          12633.0   6633.0\n",
      "oligodendrocyte     11170.0   7546.0\n",
      "stem cell            7878.0   4596.0\n",
      "ciliated cell        6732.0      NaN\n",
      "endothelial cell     2264.0    386.0\n",
      "neutrophil           2135.0      NaN\n",
      "astrocyte               NaN   6143.0\n",
      "fibroblast              NaN   1382.0\n",
      "granule cell            NaN    686.0\n",
      "----new----\n",
      "                  reference    query\n",
      "CL                                  \n",
      "neural cell         26621.0  15645.0\n",
      "macrophage          12633.0   6633.0\n",
      "oligodendrocyte     11170.0   7546.0\n",
      "stem cell            7878.0   4596.0\n",
      "endothelial cell     2264.0    386.0\n",
      "                  Brain_m  Brain_h\n",
      "CL                                \n",
      "neural cell         26621    15645\n",
      "macrophage          12633     6633\n",
      "oligodendrocyte     11170     7546\n",
      "stem cell            7878     4596\n",
      "endothelial cell     2264      386\n",
      "Task: refernece:Brain_m 60566 cells x 11100 gene -> query:Brain_h 34806 cells x 11016 gene in Brain\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Brain_m-corss-h;m_bra-map-h_bra;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 15.892364978790283\n",
      "Leiden results:\n",
      "leiden\n",
      "0     7452\n",
      "1     6513\n",
      "2     4877\n",
      "3     4035\n",
      "4     3903\n",
      "5     1917\n",
      "6     1493\n",
      "7     1040\n",
      "8      999\n",
      "9      814\n",
      "10     800\n",
      "11     360\n",
      "12     333\n",
      "13     155\n",
      "14     115\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 9.285213947296143\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 117.32652378082275\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,244,2108\n",
      "num of query_hvgs,query_degs,query_higs are 2000,511,2198\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3031\n",
      "num of query_gene_node is 2860\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2265\n",
      "ortholog_one2many       82\n",
      "ortholog_many2many      19\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2265\n",
      "Name: count, dtype: int64\n",
      "knn time is 214.11131811141968 s\n",
      "mnn time is 0.06461286544799805 s\n",
      "the time of compute mnn is  217.33705878257751 s\n",
      "Inter-net pairs ACC: 0.8638466622604097\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 490\n",
      "the num of cell nodes is 95372\n",
      "the num of gene nodes is 5891\n",
      "Times preprocess for graph:497.45\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[start]240819-0323 Heart_m-corss-h;m_hea-map-h_hea;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                     reference   query\n",
      "CL                                    \n",
      "mesenchymal cell         50417  3129.0\n",
      "myeloid cell             11505  1120.0\n",
      "T cell                    6003     NaN\n",
      "epithelial cell           3865   606.0\n",
      "cardiac muscle cell       2890  5928.0\n",
      "----new----\n",
      "                     reference   query\n",
      "CL                                    \n",
      "mesenchymal cell         50417  3129.0\n",
      "myeloid cell             11505  1120.0\n",
      "epithelial cell           3865   606.0\n",
      "cardiac muscle cell       2890  5928.0\n",
      "                     Heart_m  Heart_h\n",
      "CL                                   \n",
      "mesenchymal cell       50417     3129\n",
      "myeloid cell           11505     1120\n",
      "epithelial cell         3865      606\n",
      "cardiac muscle cell     2890     5928\n",
      "Task: refernece:Heart_m 68677 cells x 12528 gene -> query:Heart_h 10783 cells x 11069 gene in Heart\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Heart_m-corss-h;m_hea-map-h_hea;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 11.072198152542114\n",
      "Leiden results:\n",
      "leiden\n",
      "0     1858\n",
      "1     1753\n",
      "2     1243\n",
      "3     1124\n",
      "4      948\n",
      "5      910\n",
      "6      732\n",
      "7      687\n",
      "8      595\n",
      "9      446\n",
      "10     410\n",
      "11      77\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 2.840965509414673\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 70.73217988014221\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,199,2084\n",
      "num of query_hvgs,query_degs,query_higs are 2000,407,2213\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3274\n",
      "num of query_gene_node is 2935\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2368\n",
      "ortholog_one2many       99\n",
      "ortholog_many2many      51\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2368\n",
      "Name: count, dtype: int64\n",
      "knn time is 79.87338709831238 s\n",
      "mnn time is 0.038923025131225586 s\n",
      "the time of compute mnn is  83.80200695991516 s\n",
      "Inter-net pairs ACC: 0.7166310950582827\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 365\n",
      "the num of cell nodes is 79460\n",
      "the num of gene nodes is 6209\n",
      "Times preprocess for graph:271.24\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[start]240819-0328 Intestine_m-corss-h;m_int-map-h_int;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                      reference    query\n",
      "CL                                      \n",
      "enterocyte              67108.0  36272.0\n",
      "B cell                  12316.0  12604.0\n",
      "macrophage               4208.0   8243.0\n",
      "enteroendocrine cell     1749.0      NaN\n",
      "epithelial cell           672.0      NaN\n",
      "tuft cell of colon        510.0      NaN\n",
      "stromal cell                NaN   6429.0\n",
      "fibroblast                  NaN   4843.0\n",
      "endothelial cell            NaN   1227.0\n",
      "----new----\n",
      "            reference    query\n",
      "CL                            \n",
      "enterocyte    67108.0  36272.0\n",
      "B cell        12316.0  12604.0\n",
      "macrophage     4208.0   8243.0\n",
      "            Intestine_m  Intestine_h\n",
      "CL                                  \n",
      "enterocyte        67108        36272\n",
      "B cell            12316        12604\n",
      "macrophage         4208         8243\n",
      "Task: refernece:Intestine_m 83632 cells x 10665 gene -> query:Intestine_h 57119 cells x 10208 gene in Intestine\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Intestine_m-corss-h;m_int-map-h_int;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 22.639941453933716\n",
      "Leiden results:\n",
      "leiden\n",
      "0     15016\n",
      "1      9113\n",
      "2      6267\n",
      "3      5787\n",
      "4      5692\n",
      "5      4060\n",
      "6      3179\n",
      "7      2455\n",
      "8      2321\n",
      "9      1637\n",
      "10     1227\n",
      "11      365\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 18.847252368927002\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time of degs is 111.4465262889862\n",
      "--------------hvgs, degs info---------------\n",
      "num of reference_hvgs,reference_degs,reference_higs are 2000,149,2088\n",
      "num of query_hvgs,query_degs,query_higs are 2000,355,2171\n",
      "--------------gene nodes info---------------\n",
      "num of reference_gene_node is 3148\n",
      "num of query_gene_node is 2965\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one      2268\n",
      "ortholog_one2many       94\n",
      "ortholog_many2many      27\n",
      "Name: count, dtype: int64\n",
      "--------------homo edges---------------\n",
      "homology_type\n",
      "ortholog_one2one    2268\n",
      "Name: count, dtype: int64\n",
      "knn time is 479.2949640750885 s\n",
      "mnn time is 0.11287164688110352 s\n",
      "the time of compute mnn is  482.82108759880066 s\n",
      "Inter-net pairs ACC: 0.8883681295270558\n",
      "-------------------nodes info-------------------\n",
      "the num of cell feats is 283\n",
      "the num of cell nodes is 140751\n",
      "the num of gene nodes is 6113\n",
      "Times preprocess for graph:950.98\n",
      "--------------------------------\n",
      "[end][run_csMAHN_info_model_gene]\n",
      "---------------------------------\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[start]240819-0344 Kidney_m-corss-h;m_kid-map-h_kid;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                                                    reference    query\n",
      "CL                                                                    \n",
      "kidney loop of Henle ascending limb epithelial ...    25916.0      NaN\n",
      "mesenchymal cell                                      19649.0  23683.0\n",
      "kidney tubule cell                                    17060.0   7482.0\n",
      "pancreatic ductal cell                                 4754.0   4321.0\n",
      "myeloid cell                                           4431.0   2748.0\n",
      "endothelial cell                                       3716.0   5147.0\n",
      "epithelial cell                                         265.0    495.0\n",
      "podocyte                                                  NaN   1255.0\n",
      "----new----\n",
      "                        reference    query\n",
      "CL                                        \n",
      "mesenchymal cell          19649.0  23683.0\n",
      "kidney tubule cell        17060.0   7482.0\n",
      "pancreatic ductal cell     4754.0   4321.0\n",
      "myeloid cell               4431.0   2748.0\n",
      "endothelial cell           3716.0   5147.0\n",
      "epithelial cell             265.0    495.0\n",
      "                        Kidney_m  Kidney_h\n",
      "CL                                        \n",
      "mesenchymal cell           19649     23683\n",
      "kidney tubule cell         17060      7482\n",
      "pancreatic ductal cell      4754      4321\n",
      "myeloid cell                4431      2748\n",
      "endothelial cell            3716      5147\n",
      "epithelial cell              265       495\n",
      "Task: refernece:Kidney_m 49875 cells x 12846 gene -> query:Kidney_h 43876 cells x 13592 gene in Kidney\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "*************************************************************************** \n",
      "[redefind] csMAHN.pp.process_for_graph\n",
      "{'resdir': PosixPath('/public/workspace/licanchengup/link/res_publish/res/model_gene/Kidney_m-corss-h;m_kid-map-h_kid;epochs=500;is_1v1=False')}\n",
      " ***************************************************************************\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 12.11421537399292\n",
      "Leiden results:\n",
      "leiden\n",
      "0     9310\n",
      "1     6135\n",
      "2     4559\n",
      "3     3999\n",
      "4     3549\n",
      "5     2924\n",
      "6     2878\n",
      "7     2473\n",
      "8     2417\n",
      "9     2288\n",
      "10    1754\n",
      "11     548\n",
      "12     374\n",
      "13     312\n",
      "14     211\n",
      "15     145\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 14.987849473953247\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_227902/3021580929.py:319: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adatas[1].obs[key_clust] = clust_lbs2\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "item = 'HCL_MCA'\n",
    "df_para_item = df_para.query(\"mask == '{}'\".format(item))\n",
    "print('\\n[runstart] {}\\n'.format(item))\n",
    "\n",
    "aligned=True\n",
    "is_1v1 = False\n",
    "n_epochs = [100, 200, 200]\n",
    "for i, row in df_para_item.iterrows():\n",
    "    path_varmap = get_path_varmap(\n",
    "            map_sp[row['sp_simple_ref']], map_sp[row['sp_simple_que']], model='csMAHN')\n",
    "    print(path_varmap)\n",
    "\n",
    "    run_csMAHN_info_model_gene(\n",
    "        path_adata1=row['path_ref'],\n",
    "        path_adata2=row['path_que'],\n",
    "        key_class1=row['key_cell_type'],\n",
    "        key_class2=row['key_cell_type'],\n",
    "        sp1=row['sp_simple_ref'],\n",
    "        sp2=row['sp_simple_que'],\n",
    "        path_varmap=path_varmap,\n",
    "        tissue_name=row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            row['name_ref'],\n",
    "            row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res_model_gene,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(item))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba082b9c-da89-4181-9e6a-8e3c23ffbe47",
   "metadata": {},
   "source": [
    "# Retina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11b90137-dcea-4736-8f3c-4dfc148b748f",
   "metadata": {},
   "outputs": [],
   "source": [
    "item = 'Retina'\n",
    "df_para_item = df_para.query(\"mask == '{}'\".format(item))\n",
    "print('\\n[runstart] {}\\n'.format(item))\n",
    "\n",
    "aligned=True\n",
    "is_1v1 = False\n",
    "n_epochs = [100, 200, 200]\n",
    "for is_1v1 in [False,True] :\n",
    "    for i, row in df_para_item.iterrows():\n",
    "        path_varmap = get_path_varmap(\n",
    "                map_sp[row['sp_simple_ref']], map_sp[row['sp_simple_que']], model='csMAHN')\n",
    "        print(path_varmap)\n",
    "    \n",
    "        run_csMAHN_info_model_gene(\n",
    "            path_adata1=row['path_ref'],\n",
    "            path_adata2=row['path_que'],\n",
    "            key_class1=row['key_cell_type'],\n",
    "            key_class2=row['key_cell_type'],\n",
    "            sp1=row['sp_simple_ref'],\n",
    "            sp2=row['sp_simple_que'],\n",
    "            path_varmap=path_varmap,\n",
    "            tissue_name=row['tissue'],\n",
    "            aligned=aligned,\n",
    "            resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "                row['name_ref'],\n",
    "                row['name_que'],\n",
    "                sum(n_epochs),\n",
    "                is_1v1),\n",
    "            resdir=p_res_model_gene,\n",
    "            models='csMAHN,came'.split(','),\n",
    "            **{'n_epochs': n_epochs,\n",
    "               'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(item))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e8f742e-b40d-41ad-b22f-f77930dda1c7",
   "metadata": {},
   "source": [
    "# LC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f427c63b-f82d-4622-91ab-741473f84cec",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "item = 'LC'\n",
    "df_para_item = df_para.query(\"tissue == '{}'\".format(item))\n",
    "# df_para_item = df_para.loc[['LC;LChDCs;LCmDCs'],:]\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "print('\\n[run start] {}\\n'.format(item))\n",
    "display(df_para_item)\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "for i, row in df_para_item.iterrows():\n",
    "    path_varmap = get_path_varmap(\n",
    "            map_sp[row['sp_simple_ref']], map_sp[row['sp_simple_que']], model='csMAHN')\n",
    "    print(path_varmap)\n",
    "\n",
    "    run_csMAHN_info_model_gene(\n",
    "        path_adata1=row['path_ref'],\n",
    "        path_adata2=row['path_que'],\n",
    "        key_class1=row['key_cell_type'],\n",
    "        key_class2=row['key_cell_type'],\n",
    "        sp1=row['sp_simple_ref'],\n",
    "        sp2=row['sp_simple_que'],\n",
    "        path_varmap=path_varmap,\n",
    "        tissue_name=row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            row['name_ref'],\n",
    "            row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res_model_gene,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(item))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6597aa4c-e597-4bfb-9091-e61b9dbc1c76",
   "metadata": {},
   "source": [
    "# RA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d57b16c-b79a-4d1f-915e-b41fad262319",
   "metadata": {},
   "outputs": [],
   "source": [
    "item = 'RA'\n",
    "df_para_item = df_para.query(\"tissue == '{}'\".format(item))\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "print('\\n[run start] {}\\n'.format(item))\n",
    "display(df_para_item)\n",
    "\n",
    "n_epochs = [100, 100, 100]\n",
    "for i, row in df_para_item.iterrows():\n",
    "    path_varmap = get_path_varmap(\n",
    "            map_sp[row['sp_simple_ref']], map_sp[row['sp_simple_que']], model='csMAHN')\n",
    "    print(path_varmap)\n",
    "\n",
    "    run_csMAHN_info_model_gene(\n",
    "        path_adata1=row['path_ref'],\n",
    "        path_adata2=row['path_que'],\n",
    "        key_class1=row['key_cell_type'],\n",
    "        key_class2=row['key_cell_type'],\n",
    "        sp1=row['sp_simple_ref'],\n",
    "        sp2=row['sp_simple_que'],\n",
    "        path_varmap=path_varmap,\n",
    "        tissue_name=row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            row['name_ref'],\n",
    "            row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res_model_gene,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(item))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e059c2e-ab75-4e1f-858f-b87c869d9815",
   "metadata": {},
   "source": [
    "# LN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53e56223-f2d4-4aee-915c-20bc2af3e160",
   "metadata": {},
   "outputs": [],
   "source": [
    "item = 'LN'\n",
    "df_para_item = df_para.query(\"tissue == '{}'\".format(item))\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "print('\\n[run start] {}\\n'.format(item))\n",
    "display(df_para_item)\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "for i, row in df_para_item.iterrows():\n",
    "    path_varmap = get_path_varmap(\n",
    "            map_sp[row['sp_simple_ref']], map_sp[row['sp_simple_que']], model='csMAHN')\n",
    "    print(path_varmap)\n",
    "\n",
    "    run_csMAHN_info_model_gene(\n",
    "        path_adata1=row['path_ref'],\n",
    "        path_adata2=row['path_que'],\n",
    "        key_class1=row['key_cell_type'],\n",
    "        key_class2=row['key_cell_type'],\n",
    "        sp1=row['sp_simple_ref'],\n",
    "        sp2=row['sp_simple_que'],\n",
    "        path_varmap=path_varmap,\n",
    "        tissue_name=row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            row['name_ref'],\n",
    "            row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res_model_gene,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(item))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "publish",
   "language": "python",
   "name": "publish"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
