{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "191736c9-dc80-4fcb-b03d-ca1cea5cda79",
   "metadata": {},
   "source": [
    "# run_Seurat\n",
    "\n",
    "最后更新时间2024年4月1日\n",
    "\n",
    "```bash\n",
    "\n",
    "conda activate\n",
    "cd ~/link/res_publish/run\n",
    "\n",
    "jupyter nbconvert run_Seurat.ipynb --to python\n",
    "mv run_Seurat.py run_Seurat.r\n",
    "conda activate publish\n",
    "nohup Rscript run_Seurat.r > nohup_run_Seurat.txt &\n",
    "conda activate\n",
    "jobs\n",
    "\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "22e9a1b0-c9de-4c67-a5de-7a2fee98d390",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "── \u001b[1mAttaching core tidyverse packages\u001b[22m ──────────────────────── tidyverse 2.0.0 ──\n",
      "\u001b[32m✔\u001b[39m \u001b[34mdplyr    \u001b[39m 1.1.4     \u001b[32m✔\u001b[39m \u001b[34mreadr    \u001b[39m 2.1.5\n",
      "\u001b[32m✔\u001b[39m \u001b[34mforcats  \u001b[39m 1.0.0     \u001b[32m✔\u001b[39m \u001b[34mstringr  \u001b[39m 1.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2  \u001b[39m 3.5.0     \u001b[32m✔\u001b[39m \u001b[34mtibble   \u001b[39m 3.2.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mlubridate\u001b[39m 1.9.3     \u001b[32m✔\u001b[39m \u001b[34mtidyr    \u001b[39m 1.3.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr    \u001b[39m 1.0.2     \n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m    masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[36mℹ\u001b[39m Use the conflicted package (\u001b[3m\u001b[34m<http://conflicted.r-lib.org/>\u001b[39m\u001b[23m) to force all conflicts to become errors\n",
      "Loading required package: SeuratObject\n",
      "\n",
      "Loading required package: sp\n",
      "\n",
      "‘SeuratObject’ was built under R 4.3.2 but the current version is\n",
      "4.3.3; it is recomended that you reinstall ‘SeuratObject’ as the ABI\n",
      "for R may have changed\n",
      "\n",
      "‘SeuratObject’ was built with package ‘Matrix’ 1.6.3 but the current\n",
      "version is 1.6.5; it is recomended that you reinstall ‘SeuratObject’ as\n",
      "the ABI for ‘Matrix’ may have changed\n",
      "\n",
      "\n",
      "Attaching package: ‘SeuratObject’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "Loading required package: Rcpp\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "> function----------------------------------------\n",
      "serurat_to_mtx\n",
      "\n",
      "> Map_Seurat function-----------------------------\n",
      "Map_Seurat_normalize\n",
      "Map_Seurat_cluster\n",
      "Map_Seurat_mapquery\n",
      "precess_after_Seurat\n",
      "run_Seurat\t\t\t[simply]\n",
      "\n",
      "> other-------------------------------------------\n",
      "get_path_varmap\n"
     ]
    }
   ],
   "source": [
    "source(file.path('~/link/res_publish','func_r_map_seurat.r'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1993bd16-cea5-4529-b5fa-5f5f14721e75",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m3\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (10): tissue, sp_ref, path_ref, name_ref, sp_simple_ref, sp_que, path_qu...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m22\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (10): tissue, sp_ref, path_ref, name_ref, sp_simple_ref, sp_que, path_qu...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m4\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (10): tissue, sp_ref, path_ref, name_ref, sp_simple_ref, sp_que, path_qu...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m3\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (10): tissue, sp_ref, path_ref, name_ref, sp_simple_ref, sp_que, path_qu...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m1\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m10\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (10): path_ref, name_ref, sp_simple_ref, sp_ref, tissue, path_que, name_...\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 33 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>tissue</th><th scope=col>sp_ref</th><th scope=col>path_ref</th><th scope=col>name_ref</th><th scope=col>sp_simple_ref</th><th scope=col>sp_que</th><th scope=col>path_que</th><th scope=col>name_que</th><th scope=col>sp_simple_que</th><th scope=col>key_cell_type</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>retina       </td><td>zebrafish</td><td>~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult     </td><td>zebrafish_Adult</td><td>z</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/Retina/Macosko                 </td><td>Macosko      </td><td>m </td><td>cell_type     </td></tr>\n",
       "\t<tr><td>retina       </td><td>zebrafish</td><td>~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult     </td><td>zebrafish_Adult</td><td>z</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/Retina/Menon_seqwell           </td><td>Menon_seqwell</td><td>h </td><td>cell_type     </td></tr>\n",
       "\t<tr><td>retina       </td><td>zebrafish</td><td>~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult     </td><td>zebrafish_Adult</td><td>z</td><td>chicken</td><td>~/link/res_publish/run/cache/healthy/Retina/chicken_P10             </td><td>chicken_P10  </td><td>c </td><td>cell_type     </td></tr>\n",
       "\t<tr><td>Adrenal-Gland</td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Adrenal-Gland          </td><td>h_adr          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Adrenal-Gland              </td><td>m_adr        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Bone-Marrow  </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Bone-Marrow            </td><td>h_bon          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Bone-Marrow                </td><td>m_bon        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Brain        </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Brain                  </td><td>h_bra          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Brain                      </td><td>m_bra        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Heart        </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Heart                  </td><td>h_hea          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Heart                      </td><td>m_hea        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Intestine    </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Intestine              </td><td>h_int          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Intestine                  </td><td>m_int        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Kidney       </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Kidney                 </td><td>h_kid          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Kidney                     </td><td>m_kid        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Liver        </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Liver                  </td><td>h_liv          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Liver                      </td><td>m_liv        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Lung         </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Lung                   </td><td>h_lun          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Lung                       </td><td>m_lun        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Pancreas     </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Pancreas               </td><td>h_pan          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Pancreas                   </td><td>m_pan        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Spleen       </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Spleen                 </td><td>h_spl          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Spleen                     </td><td>m_spl        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Stomach      </td><td>human    </td><td>~/link/res_publish/run/cache/healthy/HCL_Stomach                </td><td>h_sto          </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/healthy/MCA_Stomach                    </td><td>m_sto        </td><td>m </td><td>CL            </td></tr>\n",
       "\t<tr><td>Adrenal-Gland</td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Adrenal-Gland          </td><td>m_adr          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Adrenal-Gland              </td><td>h_adr        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Bone-Marrow  </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Bone-Marrow            </td><td>m_bon          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Bone-Marrow                </td><td>h_bon        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Brain        </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Brain                  </td><td>m_bra          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Brain                      </td><td>h_bra        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Heart        </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Heart                  </td><td>m_hea          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Heart                      </td><td>h_hea        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Intestine    </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Intestine              </td><td>m_int          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Intestine                  </td><td>h_int        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Kidney       </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Kidney                 </td><td>m_kid          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Kidney                     </td><td>h_kid        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Liver        </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Liver                  </td><td>m_liv          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Liver                      </td><td>h_liv        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Lung         </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Lung                   </td><td>m_lun          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Lung                       </td><td>h_lun        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Pancreas     </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Pancreas               </td><td>m_pan          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Pancreas                   </td><td>h_pan        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Spleen       </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Spleen                 </td><td>m_spl          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Spleen                     </td><td>h_spl        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>Stomach      </td><td>mouse    </td><td>~/link/res_publish/run/cache/healthy/MCA_Stomach                </td><td>m_sto          </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/healthy/HCL_Stomach                    </td><td>h_sto        </td><td>h </td><td>CL            </td></tr>\n",
       "\t<tr><td>LC           </td><td>human    </td><td>~/link/res_publish/run/cache/disease/LC_h_DendriticCells        </td><td>LChDCs         </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/disease/LC_m_DendriticCells            </td><td>LCmDCs       </td><td>m </td><td>sub_cell_type </td></tr>\n",
       "\t<tr><td>LC           </td><td>human    </td><td>~/link/res_publish/run/cache/disease/LC_h_Neutrophils           </td><td>LChNeu         </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/disease/LC_m_Neutrophils               </td><td>LCmNeu       </td><td>m </td><td>sub_cell_type </td></tr>\n",
       "\t<tr><td>LC           </td><td>human    </td><td>~/link/res_publish/run/cache/disease/LC_h_allImm                </td><td>LChall         </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/disease/LC_m_allImm                    </td><td>LCmall       </td><td>m </td><td>cell_type     </td></tr>\n",
       "\t<tr><td>LC           </td><td>human    </td><td>~/link/res_publish/run/cache/disease/LC_h_Monocyte              </td><td>LChMono        </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/disease/LC_m_Monocyte                  </td><td>LCmMono      </td><td>m </td><td>sub_cell_type </td></tr>\n",
       "\t<tr><td>Mtb          </td><td>mouse    </td><td>~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232 </td><td>Mtb-m-macr     </td><td>m</td><td>human  </td><td>~/link/res_publish/run/cache/disease/Mtb_h_macrophage_GSE167232     </td><td>Mtb-h-macr   </td><td>h </td><td>macr_cell_type</td></tr>\n",
       "\t<tr><td>Mtb          </td><td>mouse    </td><td>~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232 </td><td>Mtb-m-macr     </td><td>m</td><td>macaque</td><td>~/link/res_publish/run/cache/disease/Mtb_ma_macrophage_delCD8_SCP642</td><td>Mtb-ma-macr  </td><td>ma</td><td>macr_cell_type</td></tr>\n",
       "\t<tr><td>Mtb          </td><td>mouse    </td><td>~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232 </td><td>Mtb-m-macr     </td><td>m</td><td>macaque</td><td>~/link/res_publish/run/cache/disease/COVID_ma_macrophage_GSE217483  </td><td>COVID-ma-macr</td><td>ma</td><td>macr_cell_type</td></tr>\n",
       "\t<tr><td>RA           </td><td>human    </td><td>~/link/res_publish/run/cache/disease/RA_h_fibroblast_Amp-Phase-1</td><td>RAhfib         </td><td>h</td><td>mouse  </td><td>~/link/res_publish/run/cache/disease/RA_m_fibroblast_GSE145286      </td><td>RAmfib       </td><td>m </td><td>sub_cell_type </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 33 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " tissue & sp\\_ref & path\\_ref & name\\_ref & sp\\_simple\\_ref & sp\\_que & path\\_que & name\\_que & sp\\_simple\\_que & key\\_cell\\_type\\\\\n",
       " <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t retina        & zebrafish & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/Retina/zebrafish\\_Adult      & zebrafish\\_Adult & z & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/Retina/Macosko                  & Macosko       & m  & cell\\_type     \\\\\n",
       "\t retina        & zebrafish & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/Retina/zebrafish\\_Adult      & zebrafish\\_Adult & z & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/Retina/Menon\\_seqwell            & Menon\\_seqwell & h  & cell\\_type     \\\\\n",
       "\t retina        & zebrafish & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/Retina/zebrafish\\_Adult      & zebrafish\\_Adult & z & chicken & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/Retina/chicken\\_P10              & chicken\\_P10   & c  & cell\\_type     \\\\\n",
       "\t Adrenal-Gland & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Adrenal-Gland           & h\\_adr           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Adrenal-Gland               & m\\_adr         & m  & CL            \\\\\n",
       "\t Bone-Marrow   & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Bone-Marrow             & h\\_bon           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Bone-Marrow                 & m\\_bon         & m  & CL            \\\\\n",
       "\t Brain         & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Brain                   & h\\_bra           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Brain                       & m\\_bra         & m  & CL            \\\\\n",
       "\t Heart         & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Heart                   & h\\_hea           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Heart                       & m\\_hea         & m  & CL            \\\\\n",
       "\t Intestine     & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Intestine               & h\\_int           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Intestine                   & m\\_int         & m  & CL            \\\\\n",
       "\t Kidney        & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Kidney                  & h\\_kid           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Kidney                      & m\\_kid         & m  & CL            \\\\\n",
       "\t Liver         & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Liver                   & h\\_liv           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Liver                       & m\\_liv         & m  & CL            \\\\\n",
       "\t Lung          & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Lung                    & h\\_lun           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Lung                        & m\\_lun         & m  & CL            \\\\\n",
       "\t Pancreas      & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Pancreas                & h\\_pan           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Pancreas                    & m\\_pan         & m  & CL            \\\\\n",
       "\t Spleen        & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Spleen                  & h\\_spl           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Spleen                      & m\\_spl         & m  & CL            \\\\\n",
       "\t Stomach       & human     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Stomach                 & h\\_sto           & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Stomach                     & m\\_sto         & m  & CL            \\\\\n",
       "\t Adrenal-Gland & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Adrenal-Gland           & m\\_adr           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Adrenal-Gland               & h\\_adr         & h  & CL            \\\\\n",
       "\t Bone-Marrow   & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Bone-Marrow             & m\\_bon           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Bone-Marrow                 & h\\_bon         & h  & CL            \\\\\n",
       "\t Brain         & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Brain                   & m\\_bra           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Brain                       & h\\_bra         & h  & CL            \\\\\n",
       "\t Heart         & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Heart                   & m\\_hea           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Heart                       & h\\_hea         & h  & CL            \\\\\n",
       "\t Intestine     & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Intestine               & m\\_int           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Intestine                   & h\\_int         & h  & CL            \\\\\n",
       "\t Kidney        & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Kidney                  & m\\_kid           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Kidney                      & h\\_kid         & h  & CL            \\\\\n",
       "\t Liver         & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Liver                   & m\\_liv           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Liver                       & h\\_liv         & h  & CL            \\\\\n",
       "\t Lung          & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Lung                    & m\\_lun           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Lung                        & h\\_lun         & h  & CL            \\\\\n",
       "\t Pancreas      & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Pancreas                & m\\_pan           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Pancreas                    & h\\_pan         & h  & CL            \\\\\n",
       "\t Spleen        & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Spleen                  & m\\_spl           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Spleen                      & h\\_spl         & h  & CL            \\\\\n",
       "\t Stomach       & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/MCA\\_Stomach                 & m\\_sto           & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/healthy/HCL\\_Stomach                     & h\\_sto         & h  & CL            \\\\\n",
       "\t LC            & human     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_h\\_DendriticCells         & LChDCs          & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_m\\_DendriticCells             & LCmDCs        & m  & sub\\_cell\\_type \\\\\n",
       "\t LC            & human     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_h\\_Neutrophils            & LChNeu          & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_m\\_Neutrophils                & LCmNeu        & m  & sub\\_cell\\_type \\\\\n",
       "\t LC            & human     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_h\\_allImm                 & LChall          & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_m\\_allImm                     & LCmall        & m  & cell\\_type     \\\\\n",
       "\t LC            & human     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_h\\_Monocyte               & LChMono         & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/disease/LC\\_m\\_Monocyte                   & LCmMono       & m  & sub\\_cell\\_type \\\\\n",
       "\t Mtb           & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/Mtb\\_m\\_macrophage\\_GSE167232  & Mtb-m-macr      & m & human   & \\textasciitilde{}/link/res\\_publish/run/cache/disease/Mtb\\_h\\_macrophage\\_GSE167232      & Mtb-h-macr    & h  & macr\\_cell\\_type\\\\\n",
       "\t Mtb           & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/Mtb\\_m\\_macrophage\\_GSE167232  & Mtb-m-macr      & m & macaque & \\textasciitilde{}/link/res\\_publish/run/cache/disease/Mtb\\_ma\\_macrophage\\_delCD8\\_SCP642 & Mtb-ma-macr   & ma & macr\\_cell\\_type\\\\\n",
       "\t Mtb           & mouse     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/Mtb\\_m\\_macrophage\\_GSE167232  & Mtb-m-macr      & m & macaque & \\textasciitilde{}/link/res\\_publish/run/cache/disease/COVID\\_ma\\_macrophage\\_GSE217483   & COVID-ma-macr & ma & macr\\_cell\\_type\\\\\n",
       "\t RA            & human     & \\textasciitilde{}/link/res\\_publish/run/cache/disease/RA\\_h\\_fibroblast\\_Amp-Phase-1 & RAhfib          & h & mouse   & \\textasciitilde{}/link/res\\_publish/run/cache/disease/RA\\_m\\_fibroblast\\_GSE145286       & RAmfib        & m  & sub\\_cell\\_type \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 33 × 10\n",
       "\n",
       "| tissue &lt;chr&gt; | sp_ref &lt;chr&gt; | path_ref &lt;chr&gt; | name_ref &lt;chr&gt; | sp_simple_ref &lt;chr&gt; | sp_que &lt;chr&gt; | path_que &lt;chr&gt; | name_que &lt;chr&gt; | sp_simple_que &lt;chr&gt; | key_cell_type &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| retina        | zebrafish | ~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult      | zebrafish_Adult | z | mouse   | ~/link/res_publish/run/cache/healthy/Retina/Macosko                  | Macosko       | m  | cell_type      |\n",
       "| retina        | zebrafish | ~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult      | zebrafish_Adult | z | human   | ~/link/res_publish/run/cache/healthy/Retina/Menon_seqwell            | Menon_seqwell | h  | cell_type      |\n",
       "| retina        | zebrafish | ~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult      | zebrafish_Adult | z | chicken | ~/link/res_publish/run/cache/healthy/Retina/chicken_P10              | chicken_P10   | c  | cell_type      |\n",
       "| Adrenal-Gland | human     | ~/link/res_publish/run/cache/healthy/HCL_Adrenal-Gland           | h_adr           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Adrenal-Gland               | m_adr         | m  | CL             |\n",
       "| Bone-Marrow   | human     | ~/link/res_publish/run/cache/healthy/HCL_Bone-Marrow             | h_bon           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Bone-Marrow                 | m_bon         | m  | CL             |\n",
       "| Brain         | human     | ~/link/res_publish/run/cache/healthy/HCL_Brain                   | h_bra           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Brain                       | m_bra         | m  | CL             |\n",
       "| Heart         | human     | ~/link/res_publish/run/cache/healthy/HCL_Heart                   | h_hea           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Heart                       | m_hea         | m  | CL             |\n",
       "| Intestine     | human     | ~/link/res_publish/run/cache/healthy/HCL_Intestine               | h_int           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Intestine                   | m_int         | m  | CL             |\n",
       "| Kidney        | human     | ~/link/res_publish/run/cache/healthy/HCL_Kidney                  | h_kid           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Kidney                      | m_kid         | m  | CL             |\n",
       "| Liver         | human     | ~/link/res_publish/run/cache/healthy/HCL_Liver                   | h_liv           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Liver                       | m_liv         | m  | CL             |\n",
       "| Lung          | human     | ~/link/res_publish/run/cache/healthy/HCL_Lung                    | h_lun           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Lung                        | m_lun         | m  | CL             |\n",
       "| Pancreas      | human     | ~/link/res_publish/run/cache/healthy/HCL_Pancreas                | h_pan           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Pancreas                    | m_pan         | m  | CL             |\n",
       "| Spleen        | human     | ~/link/res_publish/run/cache/healthy/HCL_Spleen                  | h_spl           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Spleen                      | m_spl         | m  | CL             |\n",
       "| Stomach       | human     | ~/link/res_publish/run/cache/healthy/HCL_Stomach                 | h_sto           | h | mouse   | ~/link/res_publish/run/cache/healthy/MCA_Stomach                     | m_sto         | m  | CL             |\n",
       "| Adrenal-Gland | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Adrenal-Gland           | m_adr           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Adrenal-Gland               | h_adr         | h  | CL             |\n",
       "| Bone-Marrow   | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Bone-Marrow             | m_bon           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Bone-Marrow                 | h_bon         | h  | CL             |\n",
       "| Brain         | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Brain                   | m_bra           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Brain                       | h_bra         | h  | CL             |\n",
       "| Heart         | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Heart                   | m_hea           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Heart                       | h_hea         | h  | CL             |\n",
       "| Intestine     | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Intestine               | m_int           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Intestine                   | h_int         | h  | CL             |\n",
       "| Kidney        | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Kidney                  | m_kid           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Kidney                      | h_kid         | h  | CL             |\n",
       "| Liver         | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Liver                   | m_liv           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Liver                       | h_liv         | h  | CL             |\n",
       "| Lung          | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Lung                    | m_lun           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Lung                        | h_lun         | h  | CL             |\n",
       "| Pancreas      | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Pancreas                | m_pan           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Pancreas                    | h_pan         | h  | CL             |\n",
       "| Spleen        | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Spleen                  | m_spl           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Spleen                      | h_spl         | h  | CL             |\n",
       "| Stomach       | mouse     | ~/link/res_publish/run/cache/healthy/MCA_Stomach                 | m_sto           | m | human   | ~/link/res_publish/run/cache/healthy/HCL_Stomach                     | h_sto         | h  | CL             |\n",
       "| LC            | human     | ~/link/res_publish/run/cache/disease/LC_h_DendriticCells         | LChDCs          | h | mouse   | ~/link/res_publish/run/cache/disease/LC_m_DendriticCells             | LCmDCs        | m  | sub_cell_type  |\n",
       "| LC            | human     | ~/link/res_publish/run/cache/disease/LC_h_Neutrophils            | LChNeu          | h | mouse   | ~/link/res_publish/run/cache/disease/LC_m_Neutrophils                | LCmNeu        | m  | sub_cell_type  |\n",
       "| LC            | human     | ~/link/res_publish/run/cache/disease/LC_h_allImm                 | LChall          | h | mouse   | ~/link/res_publish/run/cache/disease/LC_m_allImm                     | LCmall        | m  | cell_type      |\n",
       "| LC            | human     | ~/link/res_publish/run/cache/disease/LC_h_Monocyte               | LChMono         | h | mouse   | ~/link/res_publish/run/cache/disease/LC_m_Monocyte                   | LCmMono       | m  | sub_cell_type  |\n",
       "| Mtb           | mouse     | ~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232  | Mtb-m-macr      | m | human   | ~/link/res_publish/run/cache/disease/Mtb_h_macrophage_GSE167232      | Mtb-h-macr    | h  | macr_cell_type |\n",
       "| Mtb           | mouse     | ~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232  | Mtb-m-macr      | m | macaque | ~/link/res_publish/run/cache/disease/Mtb_ma_macrophage_delCD8_SCP642 | Mtb-ma-macr   | ma | macr_cell_type |\n",
       "| Mtb           | mouse     | ~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232  | Mtb-m-macr      | m | macaque | ~/link/res_publish/run/cache/disease/COVID_ma_macrophage_GSE217483   | COVID-ma-macr | ma | macr_cell_type |\n",
       "| RA            | human     | ~/link/res_publish/run/cache/disease/RA_h_fibroblast_Amp-Phase-1 | RAhfib          | h | mouse   | ~/link/res_publish/run/cache/disease/RA_m_fibroblast_GSE145286       | RAmfib        | m  | sub_cell_type  |\n",
       "\n"
      ],
      "text/plain": [
       "   tissue        sp_ref   \n",
       "1  retina        zebrafish\n",
       "2  retina        zebrafish\n",
       "3  retina        zebrafish\n",
       "4  Adrenal-Gland human    \n",
       "5  Bone-Marrow   human    \n",
       "6  Brain         human    \n",
       "7  Heart         human    \n",
       "8  Intestine     human    \n",
       "9  Kidney        human    \n",
       "10 Liver         human    \n",
       "11 Lung          human    \n",
       "12 Pancreas      human    \n",
       "13 Spleen        human    \n",
       "14 Stomach       human    \n",
       "15 Adrenal-Gland mouse    \n",
       "16 Bone-Marrow   mouse    \n",
       "17 Brain         mouse    \n",
       "18 Heart         mouse    \n",
       "19 Intestine     mouse    \n",
       "20 Kidney        mouse    \n",
       "21 Liver         mouse    \n",
       "22 Lung          mouse    \n",
       "23 Pancreas      mouse    \n",
       "24 Spleen        mouse    \n",
       "25 Stomach       mouse    \n",
       "26 LC            human    \n",
       "27 LC            human    \n",
       "28 LC            human    \n",
       "29 LC            human    \n",
       "30 Mtb           mouse    \n",
       "31 Mtb           mouse    \n",
       "32 Mtb           mouse    \n",
       "33 RA            human    \n",
       "   path_ref                                                        \n",
       "1  ~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult     \n",
       "2  ~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult     \n",
       "3  ~/link/res_publish/run/cache/healthy/Retina/zebrafish_Adult     \n",
       "4  ~/link/res_publish/run/cache/healthy/HCL_Adrenal-Gland          \n",
       "5  ~/link/res_publish/run/cache/healthy/HCL_Bone-Marrow            \n",
       "6  ~/link/res_publish/run/cache/healthy/HCL_Brain                  \n",
       "7  ~/link/res_publish/run/cache/healthy/HCL_Heart                  \n",
       "8  ~/link/res_publish/run/cache/healthy/HCL_Intestine              \n",
       "9  ~/link/res_publish/run/cache/healthy/HCL_Kidney                 \n",
       "10 ~/link/res_publish/run/cache/healthy/HCL_Liver                  \n",
       "11 ~/link/res_publish/run/cache/healthy/HCL_Lung                   \n",
       "12 ~/link/res_publish/run/cache/healthy/HCL_Pancreas               \n",
       "13 ~/link/res_publish/run/cache/healthy/HCL_Spleen                 \n",
       "14 ~/link/res_publish/run/cache/healthy/HCL_Stomach                \n",
       "15 ~/link/res_publish/run/cache/healthy/MCA_Adrenal-Gland          \n",
       "16 ~/link/res_publish/run/cache/healthy/MCA_Bone-Marrow            \n",
       "17 ~/link/res_publish/run/cache/healthy/MCA_Brain                  \n",
       "18 ~/link/res_publish/run/cache/healthy/MCA_Heart                  \n",
       "19 ~/link/res_publish/run/cache/healthy/MCA_Intestine              \n",
       "20 ~/link/res_publish/run/cache/healthy/MCA_Kidney                 \n",
       "21 ~/link/res_publish/run/cache/healthy/MCA_Liver                  \n",
       "22 ~/link/res_publish/run/cache/healthy/MCA_Lung                   \n",
       "23 ~/link/res_publish/run/cache/healthy/MCA_Pancreas               \n",
       "24 ~/link/res_publish/run/cache/healthy/MCA_Spleen                 \n",
       "25 ~/link/res_publish/run/cache/healthy/MCA_Stomach                \n",
       "26 ~/link/res_publish/run/cache/disease/LC_h_DendriticCells        \n",
       "27 ~/link/res_publish/run/cache/disease/LC_h_Neutrophils           \n",
       "28 ~/link/res_publish/run/cache/disease/LC_h_allImm                \n",
       "29 ~/link/res_publish/run/cache/disease/LC_h_Monocyte              \n",
       "30 ~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232 \n",
       "31 ~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232 \n",
       "32 ~/link/res_publish/run/cache/disease/Mtb_m_macrophage_GSE167232 \n",
       "33 ~/link/res_publish/run/cache/disease/RA_h_fibroblast_Amp-Phase-1\n",
       "   name_ref        sp_simple_ref sp_que \n",
       "1  zebrafish_Adult z             mouse  \n",
       "2  zebrafish_Adult z             human  \n",
       "3  zebrafish_Adult z             chicken\n",
       "4  h_adr           h             mouse  \n",
       "5  h_bon           h             mouse  \n",
       "6  h_bra           h             mouse  \n",
       "7  h_hea           h             mouse  \n",
       "8  h_int           h             mouse  \n",
       "9  h_kid           h             mouse  \n",
       "10 h_liv           h             mouse  \n",
       "11 h_lun           h             mouse  \n",
       "12 h_pan           h             mouse  \n",
       "13 h_spl           h             mouse  \n",
       "14 h_sto           h             mouse  \n",
       "15 m_adr           m             human  \n",
       "16 m_bon           m             human  \n",
       "17 m_bra           m             human  \n",
       "18 m_hea           m             human  \n",
       "19 m_int           m             human  \n",
       "20 m_kid           m             human  \n",
       "21 m_liv           m             human  \n",
       "22 m_lun           m             human  \n",
       "23 m_pan           m             human  \n",
       "24 m_spl           m             human  \n",
       "25 m_sto           m             human  \n",
       "26 LChDCs          h             mouse  \n",
       "27 LChNeu          h             mouse  \n",
       "28 LChall          h             mouse  \n",
       "29 LChMono         h             mouse  \n",
       "30 Mtb-m-macr      m             human  \n",
       "31 Mtb-m-macr      m             macaque\n",
       "32 Mtb-m-macr      m             macaque\n",
       "33 RAhfib          h             mouse  \n",
       "   path_que                                                            \n",
       "1  ~/link/res_publish/run/cache/healthy/Retina/Macosko                 \n",
       "2  ~/link/res_publish/run/cache/healthy/Retina/Menon_seqwell           \n",
       "3  ~/link/res_publish/run/cache/healthy/Retina/chicken_P10             \n",
       "4  ~/link/res_publish/run/cache/healthy/MCA_Adrenal-Gland              \n",
       "5  ~/link/res_publish/run/cache/healthy/MCA_Bone-Marrow                \n",
       "6  ~/link/res_publish/run/cache/healthy/MCA_Brain                      \n",
       "7  ~/link/res_publish/run/cache/healthy/MCA_Heart                      \n",
       "8  ~/link/res_publish/run/cache/healthy/MCA_Intestine                  \n",
       "9  ~/link/res_publish/run/cache/healthy/MCA_Kidney                     \n",
       "10 ~/link/res_publish/run/cache/healthy/MCA_Liver                      \n",
       "11 ~/link/res_publish/run/cache/healthy/MCA_Lung                       \n",
       "12 ~/link/res_publish/run/cache/healthy/MCA_Pancreas                   \n",
       "13 ~/link/res_publish/run/cache/healthy/MCA_Spleen                     \n",
       "14 ~/link/res_publish/run/cache/healthy/MCA_Stomach                    \n",
       "15 ~/link/res_publish/run/cache/healthy/HCL_Adrenal-Gland              \n",
       "16 ~/link/res_publish/run/cache/healthy/HCL_Bone-Marrow                \n",
       "17 ~/link/res_publish/run/cache/healthy/HCL_Brain                      \n",
       "18 ~/link/res_publish/run/cache/healthy/HCL_Heart                      \n",
       "19 ~/link/res_publish/run/cache/healthy/HCL_Intestine                  \n",
       "20 ~/link/res_publish/run/cache/healthy/HCL_Kidney                     \n",
       "21 ~/link/res_publish/run/cache/healthy/HCL_Liver                      \n",
       "22 ~/link/res_publish/run/cache/healthy/HCL_Lung                       \n",
       "23 ~/link/res_publish/run/cache/healthy/HCL_Pancreas                   \n",
       "24 ~/link/res_publish/run/cache/healthy/HCL_Spleen                     \n",
       "25 ~/link/res_publish/run/cache/healthy/HCL_Stomach                    \n",
       "26 ~/link/res_publish/run/cache/disease/LC_m_DendriticCells            \n",
       "27 ~/link/res_publish/run/cache/disease/LC_m_Neutrophils               \n",
       "28 ~/link/res_publish/run/cache/disease/LC_m_allImm                    \n",
       "29 ~/link/res_publish/run/cache/disease/LC_m_Monocyte                  \n",
       "30 ~/link/res_publish/run/cache/disease/Mtb_h_macrophage_GSE167232     \n",
       "31 ~/link/res_publish/run/cache/disease/Mtb_ma_macrophage_delCD8_SCP642\n",
       "32 ~/link/res_publish/run/cache/disease/COVID_ma_macrophage_GSE217483  \n",
       "33 ~/link/res_publish/run/cache/disease/RA_m_fibroblast_GSE145286      \n",
       "   name_que      sp_simple_que key_cell_type \n",
       "1  Macosko       m             cell_type     \n",
       "2  Menon_seqwell h             cell_type     \n",
       "3  chicken_P10   c             cell_type     \n",
       "4  m_adr         m             CL            \n",
       "5  m_bon         m             CL            \n",
       "6  m_bra         m             CL            \n",
       "7  m_hea         m             CL            \n",
       "8  m_int         m             CL            \n",
       "9  m_kid         m             CL            \n",
       "10 m_liv         m             CL            \n",
       "11 m_lun         m             CL            \n",
       "12 m_pan         m             CL            \n",
       "13 m_spl         m             CL            \n",
       "14 m_sto         m             CL            \n",
       "15 h_adr         h             CL            \n",
       "16 h_bon         h             CL            \n",
       "17 h_bra         h             CL            \n",
       "18 h_hea         h             CL            \n",
       "19 h_int         h             CL            \n",
       "20 h_kid         h             CL            \n",
       "21 h_liv         h             CL            \n",
       "22 h_lun         h             CL            \n",
       "23 h_pan         h             CL            \n",
       "24 h_spl         h             CL            \n",
       "25 h_sto         h             CL            \n",
       "26 LCmDCs        m             sub_cell_type \n",
       "27 LCmNeu        m             sub_cell_type \n",
       "28 LCmall        m             cell_type     \n",
       "29 LCmMono       m             sub_cell_type \n",
       "30 Mtb-h-macr    h             macr_cell_type\n",
       "31 Mtb-ma-macr   ma            macr_cell_type\n",
       "32 COVID-ma-macr ma            macr_cell_type\n",
       "33 RAmfib        m             sub_cell_type "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_para <- bind_rows(\n",
    "  read_csv(file.path(p_cache, \"parameter_healthy_Retina.csv\")),\n",
    "  read_csv(file.path(p_cache, \"parameter_healthy_Retina-parameters.csv\")),\n",
    "  read_csv(file.path(p_cache, \"parameter_healthy_HCL_MCA.csv\")),\n",
    "  read_csv(file.path(p_cache, \"parameter_LC.csv\")),\n",
    "  read_csv(file.path(p_cache, \"parameter_Mtb.csv\")),\n",
    "  read_csv(file.path(p_cache, \"parameter_RA.csv\")),\n",
    "\n",
    ")\n",
    "df_para <- df_para %>% mutate(\n",
    "  path_ref = file.path(p_cache, path_ref),\n",
    "  path_que = file.path(p_cache, path_que)\n",
    ")\n",
    "# cat(df_para %>% colnames(), sep = \"\\n\")\n",
    "df_para"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cec27447-1a34-45df-b243-d5a9f74b1c40",
   "metadata": {},
   "outputs": [],
   "source": [
    "run = function(tissue,\n",
    "    sp_ref,path_ref,name_ref,sp_simple_ref,\n",
    "    sp_que,path_que,name_que,sp_simple_que,\n",
    "    key_cell_type,aligned)\n",
    "{\n",
    "    is_1v1 = TRUE\n",
    "    path_varmap = get_path_varmap(sp_ref,sp_que)\n",
    "    cat(sprintf('\\n[path_varmap] %s\\n',path_varmap))\n",
    "    \n",
    "    dims = 1:20\n",
    "    resolution = 0.1\n",
    "    refdata = list(\".\" = key_cell_type)\n",
    "    \n",
    "    # Seurat;pancreas_human-map-pancreas_mouse\n",
    "    resdir_tag = sprintf('Seurat;%s-map-%s;is_1v1=True',\n",
    "                        name_ref,name_que)\n",
    "    resdir = p_res\n",
    "    # cat(sprintf('\\n[resdir_tag] %s\\n',resdir_tag))\n",
    "    # print(c(dir.exists(path_ref),dir.exists(path_que),file.exists(path_varmap)))\n",
    "    run_Seurat(\n",
    "        path_adata1 = path_ref,\n",
    "        path_adata2 = path_que,        \n",
    "        path_varmap=path_varmap,\n",
    "        key_class1=key_cell_type,\n",
    "        key_class2=key_cell_type,\n",
    "        sp1=sp_simple_ref,sp2=sp_simple_que,\n",
    "        tissue_name = tissue,\n",
    "        resdir=resdir,resdir_tag=resdir_tag,\n",
    "        dims=dims,\n",
    "        resolution=resolution,\n",
    "        refdata=refdata, aligned =aligned ,is_1v1=is_1v1\n",
    "    )\n",
    "}\n",
    "\n",
    "\n",
    "    \n",
    "\n",
    "# for healthy\n",
    "run_aligned = function(tissue,\n",
    "    sp_ref,path_ref,name_ref,sp_simple_ref,\n",
    "    sp_que,path_que,name_que,sp_simple_que,\n",
    "    key_cell_type)\n",
    "{\n",
    "    run(tissue,\n",
    "        sp_ref,path_ref,name_ref,sp_simple_ref,\n",
    "        sp_que,path_que,name_que,sp_simple_que,\n",
    "        key_cell_type,aligned=TRUE)\n",
    "}\n",
    "# for disease LC,Mtb,RA\n",
    "run_not_aligned = function(tissue,\n",
    "    sp_ref,path_ref,name_ref,sp_simple_ref,\n",
    "    sp_que,path_que,name_que,sp_simple_que,\n",
    "    key_cell_type)\n",
    "{\n",
    "    run(tissue,\n",
    "        sp_ref,path_ref,name_ref,sp_simple_ref,\n",
    "        sp_que,path_que,name_que,sp_simple_que,\n",
    "        key_cell_type,aligned=FALSE)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3dadf624-9780-4e56-a220-8f4cc04cbb99",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[start] healthy\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/zebrafish_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/retina_z-corss-m;Seurat;zebrafish_Adult-map-Macosko;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/zebrafish_to_human.txt\n",
      "[has finish] ~/link/res_publish/res/retina_z-corss-h;Seurat;zebrafish_Adult-map-Menon_seqwell;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/zebrafish_to_chicken.txt\n",
      "[has finish] ~/link/res_publish/res/retina_z-corss-c;Seurat;zebrafish_Adult-map-chicken_P10;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Adrenal-Gland_h-corss-m;Seurat;h_adr-map-m_adr;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Bone-Marrow_h-corss-m;Seurat;h_bon-map-m_bon;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Brain_h-corss-m;Seurat;h_bra-map-m_bra;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Heart_h-corss-m;Seurat;h_hea-map-m_hea;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Intestine_h-corss-m;Seurat;h_int-map-m_int;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Kidney_h-corss-m;Seurat;h_kid-map-m_kid;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/Liver_h-corss-m;Seurat;h_liv-map-m_liv;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[start] ~/link/res_publish/res/Lung_h-corss-m;Seurat;h_lun-map-m_lun;is_1v1=True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[1mRows: \u001b[22m\u001b[34m75046\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m3\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \",\"\n",
      "\u001b[31mchr\u001b[39m (3): Gene name, Mouse gene name, Mouse homology type\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[homology one2one]find 10592 genes\n",
      "[homology one2one]use 9009 genes\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Feature names cannot have underscores ('_'), replacing with dashes ('-')”\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "\u001b[1m\u001b[33mError\u001b[39m in `pmap()`:\u001b[22m\n\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In index: 11.\n\u001b[1mCaused by error in `group_by()`:\u001b[22m\n\u001b[1m\u001b[22m\u001b[33m!\u001b[39m Must group by variables found in `.data`.\n\u001b[31m✖\u001b[39m Column `cell_type` is not found.\n",
     "output_type": "error",
     "traceback": [
      "\u001b[1m\u001b[33mError\u001b[39m in `pmap()`:\u001b[22m\n\u001b[1m\u001b[22m\u001b[36mℹ\u001b[39m In index: 11.\n\u001b[1mCaused by error in `group_by()`:\u001b[22m\n\u001b[1m\u001b[22m\u001b[33m!\u001b[39m Must group by variables found in `.data`.\n\u001b[31m✖\u001b[39m Column `cell_type` is not found.\nTraceback:\n",
      "1. df_para %>% filter(!tissue %in% c(\"LC\", \"Mtb\", \"RA\")) %>% pwalk(run_aligned)",
      "2. pwalk(., run_aligned)",
      "3. pmap(.l, .f, ..., .progress = .progress)",
      "4. pmap_(\"list\", .l, .f, ..., .progress = .progress)",
      "5. with_indexed_errors(i = i, names = names, error_call = .purrr_error_call, \n .     call_with_cleanup(pmap_impl, environment(), .type, .progress, \n .         n, names, i, call_names, call_n))",
      "6. withCallingHandlers(expr, error = function(cnd) {\n .     if (i == 0L) {\n .     }\n .     else {\n .         message <- c(i = \"In index: {i}.\")\n .         if (!is.null(names) && !is.na(names[[i]]) && names[[i]] != \n .             \"\") {\n .             name <- names[[i]]\n .             message <- c(message, i = \"With name: {name}.\")\n .         }\n .         else {\n .             name <- NULL\n .         }\n .         cli::cli_abort(message, location = i, name = name, parent = cnd, \n .             call = error_call, class = \"purrr_error_indexed\")\n .     }\n . })",
      "7. call_with_cleanup(pmap_impl, environment(), .type, .progress, \n .     n, names, i, call_names, call_n)",
      "8. .f(tissue = .l[[1L]][[i]], sp_ref = .l[[2L]][[i]], path_ref = .l[[3L]][[i]], \n .     name_ref = .l[[4L]][[i]], sp_simple_ref = .l[[5L]][[i]], \n .     sp_que = .l[[6L]][[i]], path_que = .l[[7L]][[i]], name_que = .l[[8L]][[i]], \n .     sp_simple_que = .l[[9L]][[i]], key_cell_type = .l[[10L]][[i]], \n .     ...)",
      "9. run(tissue, sp_ref, path_ref, name_ref, sp_simple_ref, sp_que, \n .     path_que, name_que, sp_simple_que, key_cell_type, aligned = TRUE)   # at line 44-47 of file <text>",
      "10. run_Seurat(path_adata1 = path_ref, path_adata2 = path_que, path_varmap = path_varmap, \n  .     key_class1 = key_cell_type, key_class2 = key_cell_type, sp1 = sp_simple_ref, \n  .     sp2 = sp_simple_que, tissue_name = tissue, resdir = resdir, \n  .     resdir_tag = resdir_tag, dims = dims, resolution = resolution, \n  .     refdata = refdata, aligned = aligned, is_1v1 = is_1v1)   # at line 20-32 of file <text>",
      "11. bind_rows(tibble(dataset = paste(tissue_name, sp1, sep = \"_\"), \n  .     cell_type = adata_ref@meta.data[[key_celltype]]), tibble(dataset = paste(tissue_name, \n  .     sp2, sep = \"_\"), cell_type = adata_que@meta.data[[key_celltype]])) %>% \n  .     group_by(dataset, cell_type) %>% count() %>% pivot_wider(names_from = \"dataset\", \n  .     values_from = \"n\")",
      "12. pivot_wider(., names_from = \"dataset\", values_from = \"n\")",
      "13. count(.)",
      "14. group_by(., dataset, cell_type)",
      "15. group_by.data.frame(., dataset, cell_type)",
      "16. group_by_prepare(.data, ..., .add = .add, error_call = current_env())",
      "17. abort(bullets, call = error_call)",
      "18. signal_abort(cnd, .file)",
      "19. signalCondition(cnd)",
      "20. (function (cnd) \n  . {\n  .     if (i == 0L) {\n  .     }\n  .     else {\n  .         message <- c(i = \"In index: {i}.\")\n  .         if (!is.null(names) && !is.na(names[[i]]) && names[[i]] != \n  .             \"\") {\n  .             name <- names[[i]]\n  .             message <- c(message, i = \"With name: {name}.\")\n  .         }\n  .         else {\n  .             name <- NULL\n  .         }\n  .         cli::cli_abort(message, location = i, name = name, parent = cnd, \n  .             call = error_call, class = \"purrr_error_indexed\")\n  .     }\n  . })(structure(list(message = structure(\"Must group by variables found in `.data`.\", names = \"\"), \n  .     trace = structure(list(call = list(IRkernel::main(), kernel$run(), \n  .         handle_shell(), executor$execute(msg), tryCatch(evaluate(request$content$code, \n  .             envir = .GlobalEnv, output_handler = oh, stop_on_error = 1L), \n  .             interrupt = function(cond) {\n  .                 log_debug(\"Interrupt during execution\")\n  .                 interrupted <<- TRUE\n  .             }, error = .self$handle_error), tryCatchList(expr, \n  .             classes, parentenv, handlers), tryCatchOne(tryCatchList(expr, \n  .             names[-nh], parentenv, handlers[-nh]), names[nh], \n  .             parentenv, handlers[[nh]]), doTryCatch(return(expr), \n  .             name, parentenv, handler), tryCatchList(expr, names[-nh], \n  .             parentenv, handlers[-nh]), tryCatchOne(expr, names, \n  .             parentenv, handlers[[1L]]), doTryCatch(return(expr), \n  .             name, parentenv, handler), evaluate(request$content$code, \n  .             envir = .GlobalEnv, output_handler = oh, stop_on_error = 1L), \n  .         evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos, \n  .             debug = debug, last = i == length(out), use_try = stop_on_error != \n  .                 2L, keep_warning = keep_warning, keep_message = keep_message, \n  .             log_echo = log_echo, log_warning = log_warning, output_handler = output_handler, \n  .             include_timing = include_timing), timing_fn(handle(ev <- withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n  .             envir, enclos, user_handlers)), warning = wHandler, \n  .             error = eHandler, message = mHandler))), handle(ev <- withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n  .             envir, enclos, user_handlers)), warning = wHandler, \n  .             error = eHandler, message = mHandler)), try(f, silent = TRUE), \n  .         tryCatch(expr, error = function(e) {\n  .             call <- conditionCall(e)\n  .             if (!is.null(call)) {\n  .                 if (identical(call[[1L]], quote(doTryCatch))) \n  .                   call <- sys.call(-4L)\n  .                 dcall <- deparse(call, nlines = 1L)\n  .                 prefix <- paste(\"Error in\", dcall, \": \")\n  .                 LONG <- 75L\n  .                 sm <- strsplit(conditionMessage(e), \"\\n\")[[1L]]\n  .                 w <- 14L + nchar(dcall, type = \"w\") + nchar(sm[1L], \n  .                   type = \"w\")\n  .                 if (is.na(w)) \n  .                   w <- 14L + nchar(dcall, type = \"b\") + nchar(sm[1L], \n  .                     type = \"b\")\n  .                 if (w > LONG) \n  .                   prefix <- paste0(prefix, \"\\n  \")\n  .             }\n  .             else prefix <- \"Error : \"\n  .             msg <- paste0(prefix, conditionMessage(e), \"\\n\")\n  .             .Internal(seterrmessage(msg[1L]))\n  .             if (!silent && isTRUE(getOption(\"show.error.messages\"))) {\n  .                 cat(msg, file = outFile)\n  .                 .Internal(printDeferredWarnings())\n  .             }\n  .             invisible(structure(msg, class = \"try-error\", condition = e))\n  .         }), tryCatchList(expr, classes, parentenv, handlers), \n  .         tryCatchOne(expr, names, parentenv, handlers[[1L]]), \n  .         doTryCatch(return(expr), name, parentenv, handler), withCallingHandlers(withVisible(eval_with_user_handlers(expr, \n  .             envir, enclos, user_handlers)), warning = wHandler, \n  .             error = eHandler, message = mHandler), withVisible(eval_with_user_handlers(expr, \n  .             envir, enclos, user_handlers)), eval_with_user_handlers(expr, \n  .             envir, enclos, user_handlers), eval(expr, envir, \n  .             enclos), eval(expr, envir, enclos), df_para %>% filter(!tissue %in% \n  .             c(\"LC\", \"Mtb\", \"RA\")) %>% pwalk(run_aligned), pwalk(., \n  .             run_aligned), pmap(.l, .f, ..., .progress = .progress), \n  .         pmap_(\"list\", .l, .f, ..., .progress = .progress), with_indexed_errors(i = i, \n  .             names = names, error_call = .purrr_error_call, call_with_cleanup(pmap_impl, \n  .                 environment(), .type, .progress, n, names, i, \n  .                 call_names, call_n)), withCallingHandlers(expr, \n  .             error = function(cnd) {\n  .                 if (i == 0L) {\n  .                 }\n  .                 else {\n  .                   message <- c(i = \"In index: {i}.\")\n  .                   if (!is.null(names) && !is.na(names[[i]]) && \n  .                     names[[i]] != \"\") {\n  .                     name <- names[[i]]\n  .                     message <- c(message, i = \"With name: {name}.\")\n  .                   }\n  .                   else {\n  .                     name <- NULL\n  .                   }\n  .                   cli::cli_abort(message, location = i, name = name, \n  .                     parent = cnd, call = error_call, class = \"purrr_error_indexed\")\n  .                 }\n  .             }), call_with_cleanup(pmap_impl, environment(), .type, \n  .             .progress, n, names, i, call_names, call_n), .f(tissue = .l[[1L]][[i]], \n  .             sp_ref = .l[[2L]][[i]], path_ref = .l[[3L]][[i]], \n  .             name_ref = .l[[4L]][[i]], sp_simple_ref = .l[[5L]][[i]], \n  .             sp_que = .l[[6L]][[i]], path_que = .l[[7L]][[i]], \n  .             name_que = .l[[8L]][[i]], sp_simple_que = .l[[9L]][[i]], \n  .             key_cell_type = .l[[10L]][[i]], ...), run(tissue, \n  .             sp_ref, path_ref, name_ref, sp_simple_ref, sp_que, \n  .             path_que, name_que, sp_simple_que, key_cell_type, \n  .             aligned = TRUE), run_Seurat(path_adata1 = path_ref, \n  .             path_adata2 = path_que, path_varmap = path_varmap, \n  .             key_class1 = key_cell_type, key_class2 = key_cell_type, \n  .             sp1 = sp_simple_ref, sp2 = sp_simple_que, tissue_name = tissue, \n  .             resdir = resdir, resdir_tag = resdir_tag, dims = dims, \n  .             resolution = resolution, refdata = refdata, aligned = aligned, \n  .             is_1v1 = is_1v1), bind_rows(tibble(dataset = paste(tissue_name, \n  .             sp1, sep = \"_\"), cell_type = adata_ref@meta.data[[key_celltype]]), \n  .             tibble(dataset = paste(tissue_name, sp2, sep = \"_\"), \n  .                 cell_type = adata_que@meta.data[[key_celltype]])) %>% \n  .             group_by(dataset, cell_type) %>% count() %>% pivot_wider(names_from = \"dataset\", \n  .             values_from = \"n\"), pivot_wider(., names_from = \"dataset\", \n  .             values_from = \"n\"), count(.), group_by(., dataset, \n  .             cell_type), group_by.data.frame(., dataset, cell_type), \n  .         group_by_prepare(.data, ..., .add = .add, error_call = current_env()), \n  .         abort(bullets, call = error_call)), parent = c(0L, 1L, \n  .     2L, 3L, 4L, 5L, 6L, 7L, 6L, 9L, 10L, 4L, 12L, 13L, 13L, 15L, \n  .     16L, 17L, 18L, 19L, 13L, 13L, 13L, 23L, 24L, 0L, 0L, 27L, \n  .     28L, 29L, 30L, 29L, 29L, 33L, 34L, 35L, 0L, 0L, 0L, 0L, 40L, \n  .     41L), visible = c(TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  .     TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  .     TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  .     TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE, \n  .     TRUE, TRUE, TRUE, FALSE, FALSE), namespace = c(\"IRkernel\", \n  .     NA, \"IRkernel\", NA, \"base\", \"base\", \"base\", \"base\", \"base\", \n  .     \"base\", \"base\", \"evaluate\", \"evaluate\", \"evaluate\", \"evaluate\", \n  .     \"base\", \"base\", \"base\", \"base\", \"base\", \"base\", \"base\", \"evaluate\", \n  .     \"base\", \"base\", NA, \"purrr\", \"purrr\", \"purrr\", \"purrr\", \"base\", \n  .     \"purrr\", NA, NA, NA, NA, \"tidyr\", \"dplyr\", \"dplyr\", \"dplyr\", \n  .     \"dplyr\", \"rlang\"), scope = c(\"::\", NA, \"local\", NA, \"::\", \n  .     \"local\", \"local\", \"local\", \"local\", \"local\", \"local\", \"::\", \n  .     \":::\", \"local\", \"local\", \"::\", \"::\", \"local\", \"local\", \"local\", \n  .     \"::\", \"::\", \":::\", \"::\", \"::\", NA, \"::\", \"::\", \":::\", \":::\", \n  .     \"::\", \":::\", \"global\", \"global\", \"global\", NA, \"::\", \"::\", \n  .     \"::\", \":::\", \"::\", \"::\"), error_frame = c(FALSE, FALSE, FALSE, \n  .     FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  .     FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  .     FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  .     FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, FALSE, \n  .     TRUE, FALSE, FALSE)), row.names = c(NA, -42L), version = 2L, class = c(\"rlang_trace\", \n  .     \"rlib_trace\", \"tbl\", \"data.frame\")), parent = NULL, body = c(x = \"Column `cell_type` is not found.\"), \n  .     rlang = list(inherit = TRUE), call = group_by(., dataset, \n  .         cell_type), use_cli_format = TRUE), class = c(\"rlang_error\", \n  . \"error\", \"condition\")))",
      "21. cli::cli_abort(message, location = i, name = name, parent = cnd, \n  .     call = error_call, class = \"purrr_error_indexed\")",
      "22. rlang::abort(message, ..., call = call, use_cli_format = TRUE, \n  .     .frame = .frame)",
      "23. signal_abort(cnd, .file)"
     ]
    }
   ],
   "source": [
    "cat('\\n[start] healthy\\n')\n",
    "df_para %>% filter(!tissue %in% c('LC','Mtb','RA','retina-p')) %>% pwalk(run_aligned)\n",
    "cat('\\n[finish] healthy\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7161400c-0574-4f4a-a01c-c76b21048b84",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[start] disease\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/LC_h-corss-m;Seurat;LChDCs-map-LCmDCs;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/LC_h-corss-m;Seurat;LChNeu-map-LCmNeu;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/LC_h-corss-m;Seurat;LChall-map-LCmall;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/LC_h-corss-m;Seurat;LChMono-map-LCmMono;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[has finish] ~/link/res_publish/res/Mtb_m-corss-h;Seurat;Mtb-m-macr-map-Mtb-h-macr;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_macaque.txt\n",
      "[has finish] ~/link/res_publish/res/Mtb_m-corss-ma;Seurat;Mtb-m-macr-map-Mtb-ma-macr;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_macaque.txt\n",
      "[has finish] ~/link/res_publish/res/Mtb_m-corss-ma;Seurat;Mtb-m-macr-map-COVID-ma-macr;is_1v1=True\n",
      "\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[has finish] ~/link/res_publish/res/RA_h-corss-m;Seurat;RAhfib-map-RAmfib;is_1v1=True\n",
      "\n",
      "[finish] disease\n"
     ]
    }
   ],
   "source": [
    "cat(\"\\n[start] disease\\n\")\n",
    "df_para %>% filter(tissue %in% c(\"LC\", \"Mtb\", \"RA\",\"retina-p\")) %>% pwalk(run_not_aligned)\n",
    "cat(\"\\n[finish] disease\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb8e8dca-8b03-41c1-85f8-2132bda7b364",
   "metadata": {},
   "outputs": [],
   "source": [
    "cat('\\n[finish]\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R_publish",
   "language": "R",
   "name": "r_publish"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
