{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fd2db94f-a10b-4da7-9f70-1dc623630579",
   "metadata": {},
   "source": [
    "# disease_run\n",
    "\n",
    "对disease数据集运行`run_cross_species_models`\n",
    "\n",
    "\n",
    "```shell\n",
    "{\n",
    "conda activate\n",
    "cd ~/link/res_publish/run\n",
    "jupyter nbconvert disease_run.ipynb --to python\n",
    "\n",
    "conda activate publish\n",
    "nohup python disease_run.py > nohup_disease_run.txt &\n",
    "sleep 10  && rm disease_run.py\n",
    "jobs\n",
    "\n",
    "conda activate\n",
    "echo 'finish'\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4988b915-bdf0-4ff9-9be5-e800196a876e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "p_root = Path('~/link/res_publish').expanduser()\n",
    "None if str(p_root) in sys.path else sys.path.append(str(p_root))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4bc19322-0d07-4fa6-b692-97e2abc89980",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-09-07 11:35:00.074978: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[func help]----------------------------------------------------------------\n",
      "> parameter\n",
      "    p_root\t[name] res_publish\n",
      "        p_run, p_plot, p_res, p_cache, p_pdf\n",
      "    p_df_varmap\n",
      "    map_sp_reverse\n",
      "    rng\n",
      "[from func import * ]------------------------------------------------------\n",
      "| p_root                  | display              | get_res_obs               |\n",
      "| p_run                   | h5ad_to_mtx          | get_source_obs            |\n",
      "| p_plot                  | load_adata           | get_matrix_max_prob_median|\n",
      "| p_res                   | load_normalized_adata| get_matrix_count          |\n",
      "| p_cache                 | pdf2_merge           | get_path_varmap           |\n",
      "| p_pdf                   | show_umap            | get_res_stat              |\n",
      "| p_data_process          | time_tag_detect      | find_path_from_para       |\n",
      "| map_sp                  | time_tag_get         | func_help                 |\n",
      "| map_sp_reverse          | time_tag_toggle      |                           |\n",
      "| map_sp_SAMap            | get_test_result_df   |                           |\n",
      "| run_cross_species_models| get_adata_umap       |                           |\n",
      "\n",
      "[func all name ]-----------------------------------------------------------\n",
      "| BytesIO                                 | map_func_run_cross_species_models|\n",
      "| ImageReader                             | map_sp                           |\n",
      "| Iterable                                | map_sp_SAMap                     |\n",
      "| Path                                    | map_sp_reverse                   |\n",
      "| PyPDF2                                  | mmwrite                          |\n",
      "| SAM                                     | mpimg                            |\n",
      "| SAMAP                                   | mpl                              |\n",
      "| StringIO                                | namedtuple                       |\n",
      "| TTFont                                  | np                               |\n",
      "| ZipFile                                 | os                               |\n",
      "| aligned_type                            | p_cache                          |\n",
      "| archive_gzip                            | p_data_process                   |\n",
      "| calculate_F1Score_with_confusion_matrix | p_df_varmap                      |\n",
      "| calculate_accuracy_with_confusion_matrix| p_maps_SAMap                     |\n",
      "| calculate_confusion_matrix              | p_pdf                            |\n",
      "| calculate_more_with_confusion_matrix    | p_plot                           |\n",
      "| calculate_res_stat                      | p_res                            |\n",
      "| came                                    | p_root                           |\n",
      "| canvas                                  | p_run                            |\n",
      "| csMAHN                                  | palettes                         |\n",
      "| csr_matrix                              | pd                               |\n",
      "| customize__init__for_SAMAP              | pdf2_merge                       |\n",
      "| customize_calculate_blast_graph         | plt                              |\n",
      "| df_apply_merge_field                    | precess_after_came               |\n",
      "| df_varmap_query_exists                  | precess_after_csMAHN             |\n",
      "| display                                 | re                               |\n",
      "| dumps                                   | reportlab                        |\n",
      "| exchange_gn                             | rng                              |\n",
      "| find_path_from_para                     | run_cross_species_models         |\n",
      "| get_1v1_matches                         | samap                            |\n",
      "| get_adata_umap                          | sana                             |\n",
      "| get_alignment_score_for_each_cell       | sc                               |\n",
      "| get_homology_parameters                 | show_obj_attr                    |\n",
      "| get_matrix_count                        | show_str_arr                     |\n",
      "| get_matrix_count_ratio                  | show_umap                        |\n",
      "| get_matrix_max_prob_median              | sns                              |\n",
      "| get_path_varmap                         | sp                               |\n",
      "| get_res_obs                             | stats                            |\n",
      "| get_res_stat                            | subset_adata                     |\n",
      "| get_significance_marker                 | tarfile                          |\n",
      "| get_source_obs                          | time                             |\n",
      "| get_test_result_df                      | time_tag_detect                  |\n",
      "| get_type_counts_info                    | time_tag_get                     |\n",
      "| group_agg                               | time_tag_toggle                  |\n",
      "| h5ad_to_mtx                             | typing                           |\n",
      "| is_1v1                                  | unify_group_counts_index_name    |\n",
      "| load_adata                              | warnings                         |\n",
      "| load_normalized_adata                   | zip_longest                      |\n",
      "| loads                                   |                                  |\n",
      "\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "    if (document.getElementsByClassName(\"user-container-lcc\").length == 0) {\n",
       "        var ele_body = document.getElementsByTagName(\"body\")[0]\n",
       "        var ele_div = document.createElement(\"div\")\n",
       "        var ele_scr = document.createElement(\"script\")\n",
       "        ele_scr.src = \"https://code.jquery.com/jquery-3.6.2.min.js\"\n",
       "        ele_div.className = \"user-container-lcc\"\n",
       "        ele_div.insertBefore(ele_scr, ele_div.firstChild)\n",
       "        ele_body.insertBefore(ele_div, ele_body.firstChild)\n",
       "    }\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from func import *\n",
    "from utils.general import *\n",
    "import utils as ut\n",
    "from IPython.display import display"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "16b881e2-f011-4b07-bcf9-b4756a568ca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>sp_ref</th>\n",
       "      <th>path_ref</th>\n",
       "      <th>name_ref</th>\n",
       "      <th>sp_simple_ref</th>\n",
       "      <th>sp_que</th>\n",
       "      <th>path_que</th>\n",
       "      <th>name_que</th>\n",
       "      <th>sp_simple_que</th>\n",
       "      <th>key_cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChDCs</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmDCs</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChNeu</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmNeu</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChall</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmall</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LC</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LChMono</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>LCmMono</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RA</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>RAhfib</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>RAmfib</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-h-macr</td>\n",
       "      <td>h</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-ma-macr</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma-macr</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma-macr-time-choice</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Mtb</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Mtb-m-macr</td>\n",
       "      <td>m</td>\n",
       "      <td>macaque</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>BCG_ma_W25_macr</td>\n",
       "      <td>ma</td>\n",
       "      <td>macr_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_T</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_T</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_imm</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_imm</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_B</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_B</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_NK</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_NK</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_mar</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_mar</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>COVID</td>\n",
       "      <td>ma</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma</td>\n",
       "      <td>ma</td>\n",
       "      <td>h</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-hu</td>\n",
       "      <td>h</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COVID</td>\n",
       "      <td>ma</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma-mar</td>\n",
       "      <td>ma</td>\n",
       "      <td>h</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-hu-mar</td>\n",
       "      <td>h</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COVID</td>\n",
       "      <td>ma</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma-c</td>\n",
       "      <td>ma</td>\n",
       "      <td>h</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-hu</td>\n",
       "      <td>h</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COVID</td>\n",
       "      <td>ma</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-ma-mar-c</td>\n",
       "      <td>ma</td>\n",
       "      <td>h</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>COVID-hu-mar</td>\n",
       "      <td>h</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tissue sp_ref                                           path_ref  \\\n",
       "0     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "1     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "2     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "3     LC  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "0     RA  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "0    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "1    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "2    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "3    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "4    Mtb  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "0     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "1     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "2     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "3     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "4     LN  human  /public/workspace/licanchengup/link/res_publis...   \n",
       "0  COVID     ma  /public/workspace/licanchengup/link/res_publis...   \n",
       "1  COVID     ma  /public/workspace/licanchengup/link/res_publis...   \n",
       "2  COVID     ma  /public/workspace/licanchengup/link/res_publis...   \n",
       "3  COVID     ma  /public/workspace/licanchengup/link/res_publis...   \n",
       "\n",
       "         name_ref sp_simple_ref   sp_que  \\\n",
       "0          LChDCs             h    mouse   \n",
       "1          LChNeu             h    mouse   \n",
       "2          LChall             h    mouse   \n",
       "3         LChMono             h    mouse   \n",
       "0          RAhfib             h    mouse   \n",
       "0      Mtb-m-macr             m    human   \n",
       "1      Mtb-m-macr             m  macaque   \n",
       "2      Mtb-m-macr             m  macaque   \n",
       "3      Mtb-m-macr             m  macaque   \n",
       "4      Mtb-m-macr             m  macaque   \n",
       "0         SLE_h_T             h    mouse   \n",
       "1       SLE_h_imm             h    mouse   \n",
       "2         SLE_h_B             h    mouse   \n",
       "3        SLE_h_NK             h    mouse   \n",
       "4       SLE_h_mar             h    mouse   \n",
       "0        COVID-ma            ma        h   \n",
       "1    COVID-ma-mar            ma        h   \n",
       "2      COVID-ma-c            ma        h   \n",
       "3  COVID-ma-mar-c            ma        h   \n",
       "\n",
       "                                            path_que  \\\n",
       "0  /public/workspace/licanchengup/link/res_publis...   \n",
       "1  /public/workspace/licanchengup/link/res_publis...   \n",
       "2  /public/workspace/licanchengup/link/res_publis...   \n",
       "3  /public/workspace/licanchengup/link/res_publis...   \n",
       "0  /public/workspace/licanchengup/link/res_publis...   \n",
       "0  /public/workspace/licanchengup/link/res_publis...   \n",
       "1  /public/workspace/licanchengup/link/res_publis...   \n",
       "2  /public/workspace/licanchengup/link/res_publis...   \n",
       "3  /public/workspace/licanchengup/link/res_publis...   \n",
       "4  /public/workspace/licanchengup/link/res_publis...   \n",
       "0  /public/workspace/licanchengup/link/res_publis...   \n",
       "1  /public/workspace/licanchengup/link/res_publis...   \n",
       "2  /public/workspace/licanchengup/link/res_publis...   \n",
       "3  /public/workspace/licanchengup/link/res_publis...   \n",
       "4  /public/workspace/licanchengup/link/res_publis...   \n",
       "0  /public/workspace/licanchengup/link/res_publis...   \n",
       "1  /public/workspace/licanchengup/link/res_publis...   \n",
       "2  /public/workspace/licanchengup/link/res_publis...   \n",
       "3  /public/workspace/licanchengup/link/res_publis...   \n",
       "\n",
       "                    name_que sp_simple_que   key_cell_type  \n",
       "0                     LCmDCs             m   sub_cell_type  \n",
       "1                     LCmNeu             m   sub_cell_type  \n",
       "2                     LCmall             m       cell_type  \n",
       "3                    LCmMono             m   sub_cell_type  \n",
       "0                     RAmfib             m   sub_cell_type  \n",
       "0                 Mtb-h-macr             h  macr_cell_type  \n",
       "1                Mtb-ma-macr            ma  macr_cell_type  \n",
       "2              COVID-ma-macr            ma  macr_cell_type  \n",
       "3  COVID-ma-macr-time-choice            ma  macr_cell_type  \n",
       "4            BCG_ma_W25_macr            ma  macr_cell_type  \n",
       "0           merge_m_kidney_T             m   sub_cell_type  \n",
       "1         merge_m_kidney_imm             m       cell_type  \n",
       "2           merge_m_kidney_B             m   sub_cell_type  \n",
       "3          merge_m_kidney_NK             m   sub_cell_type  \n",
       "4         merge_m_kidney_mar             m   sub_cell_type  \n",
       "0                   COVID-hu             h       cell_type  \n",
       "1               COVID-hu-mar             h   sub_cell_type  \n",
       "2                   COVID-hu             h       cell_type  \n",
       "3               COVID-hu-mar             h   sub_cell_type  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "tissue                                                       COVID\n",
       "sp_ref                                                          ma\n",
       "path_ref         /public/workspace/licanchengup/link/res_publis...\n",
       "name_ref                                            COVID-ma-mar-c\n",
       "sp_simple_ref                                                   ma\n",
       "sp_que                                                           h\n",
       "path_que         /public/workspace/licanchengup/link/res_publis...\n",
       "name_que                                              COVID-hu-mar\n",
       "sp_simple_que                                                    h\n",
       "key_cell_type                                        sub_cell_type\n",
       "Name: 3, dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q_disease = 'LC,RA,Mtb,LN,COVID'.split(',')\n",
    "\n",
    "df_para = pd.concat([pd.read_csv(p_cache.joinpath(\n",
    "    'parameter_{}.csv'.format(i)\n",
    ")) for i in q_disease])\n",
    "df_para['path_ref'] = df_para['path_ref'].apply(lambda x:p_cache.joinpath(x))\n",
    "df_para['path_que'] = df_para['path_que'].apply(lambda x:p_cache.joinpath(x))\n",
    "display(df_para)\n",
    "\n",
    "for _i, _row in df_para.iterrows():\n",
    "    pass\n",
    "_row"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5e2b20-c197-443b-b2ba-04e897dbbccb",
   "metadata": {},
   "source": [
    "# RA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fbc132c1-5cbf-4955-bbc0-1da0d0d24ec9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>cell_count_ref</th>\n",
       "      <th>_batch_ref</th>\n",
       "      <th>cell_count_que</th>\n",
       "      <th>_batch_que</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAhfib-map-RAmfib</td>\n",
       "      <td>1844</td>\n",
       "      <td>17</td>\n",
       "      <td>11707</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tag  cell_count_ref  _batch_ref  cell_count_que  _batch_que\n",
       "0  RAhfib-map-RAmfib            1844          17           11707           6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>任务名</th>\n",
       "      <th>人类细胞数量</th>\n",
       "      <th>人类样本数量</th>\n",
       "      <th>小鼠细胞数量</th>\n",
       "      <th>小鼠样本数量</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RAhfib-map-RAmfib</td>\n",
       "      <td>1844</td>\n",
       "      <td>17</td>\n",
       "      <td>11707</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 任务名  人类细胞数量  人类样本数量  小鼠细胞数量  小鼠样本数量\n",
       "0  RAhfib-map-RAmfib    1844      17   11707       6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "disease = 'RA'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = ut.df.apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "display(df_para_disease_info)\n",
    "display(df_para_disease_info.rename(columns={\n",
    "    'tag':'任务名',\n",
    "    'cell_count_ref':'人类细胞数量',\n",
    "    '_batch_ref':'人类样本数量',\n",
    "    'cell_count_que':'小鼠细胞数量',\n",
    "    '_batch_que':'小鼠样本数量',\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "57ebd526-dfe2-4f8e-a893-0dde415ebdb7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[run start] RA\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 RA_h-corss-m;csMAHN;RAhfib-map-RAmfib;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 RA_h-corss-m;came;RAhfib-map-RAmfib;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240907-1135 RA_h-corss-m;SAMap;RAhfib-map-RAmfib;is_1v1=False\n",
      "\n",
      "[run finish] RA\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 100, 100]    \n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# SAMap\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='SAMap'.split(','),\n",
    "        **{\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c6d92ba-6a8e-47f8-840d-061e90ecb1d1",
   "metadata": {},
   "source": [
    "# LC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6454d5b5-502b-4a4a-a9c6-fb397aea422a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>cell_count_ref</th>\n",
       "      <th>_batch_ref</th>\n",
       "      <th>cell_count_que</th>\n",
       "      <th>_batch_que</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LChDCs-map-LCmDCs</td>\n",
       "      <td>2010</td>\n",
       "      <td>18</td>\n",
       "      <td>646</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LChNeu-map-LCmNeu</td>\n",
       "      <td>2728</td>\n",
       "      <td>18</td>\n",
       "      <td>3593</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LChall-map-LCmall</td>\n",
       "      <td>34558</td>\n",
       "      <td>18</td>\n",
       "      <td>9201</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LChMono-map-LCmMono</td>\n",
       "      <td>1176</td>\n",
       "      <td>18</td>\n",
       "      <td>771</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   tag  cell_count_ref  _batch_ref  cell_count_que  _batch_que\n",
       "0    LChDCs-map-LCmDCs            2010          18             646           9\n",
       "1    LChNeu-map-LCmNeu            2728          18            3593           9\n",
       "2    LChall-map-LCmall           34558          18            9201           9\n",
       "3  LChMono-map-LCmMono            1176          18             771           9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>任务名</th>\n",
       "      <th>人类细胞数量</th>\n",
       "      <th>人类样本数量</th>\n",
       "      <th>小鼠细胞数量</th>\n",
       "      <th>小鼠样本数量</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LChDCs-map-LCmDCs</td>\n",
       "      <td>2010</td>\n",
       "      <td>18</td>\n",
       "      <td>646</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LChMono-map-LCmMono</td>\n",
       "      <td>1176</td>\n",
       "      <td>18</td>\n",
       "      <td>771</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LChNeu-map-LCmNeu</td>\n",
       "      <td>2728</td>\n",
       "      <td>18</td>\n",
       "      <td>3593</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   任务名  人类细胞数量  人类样本数量  小鼠细胞数量  小鼠样本数量\n",
       "0    LChDCs-map-LCmDCs    2010      18     646       9\n",
       "3  LChMono-map-LCmMono    1176      18     771       9\n",
       "1    LChNeu-map-LCmNeu    2728      18    3593       9"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "disease = 'LC'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = ut.df.apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "display(df_para_disease_info)\n",
    "display(df_para_disease_info.loc[[0,3,1],:].rename(columns={\n",
    "    'tag':'任务名',\n",
    "    'cell_count_ref':'人类细胞数量',\n",
    "    '_batch_ref':'人类样本数量',\n",
    "    'cell_count_que':'小鼠细胞数量',\n",
    "    '_batch_que':'小鼠样本数量',\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "61e5543d-6fa6-43dd-986a-6e7d2de0f1a4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[run start] LC\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;csMAHN;LChDCs-map-LCmDCs;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;came;LChDCs-map-LCmDCs;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;csMAHN;LChNeu-map-LCmNeu;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;came;LChNeu-map-LCmNeu;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;csMAHN;LChall-map-LCmall;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;came;LChall-map-LCmall;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;csMAHN;LChMono-map-LCmMono;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;came;LChMono-map-LCmMono;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;csMAHN;LChall-map-LCmall;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LC_h-corss-m;came;LChall-map-LCmall;epochs=300;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240907-1135 LC_h-corss-m;SAMap;LChDCs-map-LCmDCs;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240907-1135 LC_h-corss-m;SAMap;LChNeu-map-LCmNeu;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240907-1135 LC_h-corss-m;SAMap;LChall-map-LCmall;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/SAMap/maps_gene_name\n",
      "[path_varmap] SAMap\tmaps_gene_name\n",
      "[has finish]240907-1135 LC_h-corss-m;SAMap;LChMono-map-LCmMono;is_1v1=False\n",
      "\n",
      "[run finish] LC\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "n_epochs = [100, 100, 100]\n",
    "for _i, _row in df_para_disease.query(\"key_cell_type == 'cell_type'\").iterrows():\n",
    "    \n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# SAMap\n",
    "for _i, _row in df_para_disease.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='SAMap'.split(','),\n",
    "        **{\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5b1c5a7-abd4-4fe4-a7f7-cc8f3d9c9a34",
   "metadata": {},
   "source": [
    "# Mtb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "55575306-3cb5-4d85-b88b-8e2b2a3c21fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# disease = 'Mtb'\n",
    "# df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "# df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "# df_para_disease_info['tag'] = ut.df.apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "# df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "# df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "#     lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "# df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "#     lambda x:x.shape[0]\n",
    "# )\n",
    "# df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "#     lambda x:x['_batch'].unique().size\n",
    "# )\n",
    "# df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# # que\n",
    "# df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "#     lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "# df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "#     lambda x:x.shape[0]\n",
    "# )\n",
    "# df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "#     lambda x:x['_batch'].unique().size\n",
    "# )\n",
    "\n",
    "# df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "# df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "# df_para_disease_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dbb00256-ceb8-48b7-b63a-a726faeab1b0",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "# print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "# aligned=False\n",
    "# is_1v1 = False\n",
    "\n",
    "# n_epochs = [100, 200, 200]\n",
    "# for _i, _row in df_para_disease.iterrows():\n",
    "#     run_cross_species_models(\n",
    "#         path_adata1=_row['path_ref'],\n",
    "#         path_adata2=_row['path_que'],\n",
    "#         key_class1=_row['key_cell_type'],\n",
    "#         key_class2=_row['key_cell_type'],\n",
    "#         sp1=_row['sp_simple_ref'],\n",
    "#         sp2=_row['sp_simple_que'],\n",
    "#         tissue_name=_row['tissue'],\n",
    "#         aligned=aligned,\n",
    "#         resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "#             _row['name_ref'],\n",
    "#             _row['name_que'],\n",
    "#             sum(n_epochs),\n",
    "#             is_1v1),\n",
    "#         resdir=p_res,\n",
    "#         models='csMAHN,came'.split(','),\n",
    "#         **{'n_epochs': n_epochs,\n",
    "#            'is_1v1': is_1v1})\n",
    "\n",
    "# # SAMap\n",
    "# for _i, _row in df_para_disease.iterrows():\n",
    "#     run_cross_species_models(\n",
    "#         path_adata1=_row['path_ref'],\n",
    "#         path_adata2=_row['path_que'],\n",
    "#         key_class1=_row['key_cell_type'],\n",
    "#         key_class2=_row['key_cell_type'],\n",
    "#         sp1=_row['sp_simple_ref'],\n",
    "#         sp2=_row['sp_simple_que'],\n",
    "#         tissue_name=_row['tissue'],\n",
    "#         aligned=aligned,\n",
    "#         resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "#             _row['name_ref'],\n",
    "#             _row['name_que'],\n",
    "#             is_1v1),\n",
    "#         resdir=p_res,\n",
    "#         models='SAMap'.split(','),\n",
    "#         **{\n",
    "#            'is_1v1': is_1v1})\n",
    "\n",
    "\n",
    "# print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2d67ab75-48f0-4330-b9f4-0a2fc56e3eaf",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# for _ in range(3):\n",
    "#     n_epochs = [100, 200, 200]\n",
    "#     for _i, _row in df_para_disease[df_para_disease['name_que'].isin('Mtb-h-macr,merge-ma-macr'.split(','))].iterrows():\n",
    "#         run_cross_species_models(\n",
    "#             path_adata1=_row['path_ref'],\n",
    "#             path_adata2=_row['path_que'],\n",
    "#             key_class1=_row['key_cell_type'],\n",
    "#             key_class2=_row['key_cell_type'],\n",
    "#             sp1=_row['sp_simple_ref'],\n",
    "#             sp2=_row['sp_simple_que'],\n",
    "#             tissue_name=_row['tissue'],\n",
    "#             aligned=aligned,\n",
    "#             resdir_tag='{}-map-{};epochs={};is_1v1={};{}'.format(\n",
    "#                 _row['name_ref'],\n",
    "#                 _row['name_que'],\n",
    "#                 sum(n_epochs),\n",
    "#                 is_1v1,time_tag_get()),\n",
    "#             resdir=p_res,\n",
    "#             models='csMAHN'.split(','),\n",
    "#             **{'n_epochs': n_epochs,\n",
    "#                'is_1v1': is_1v1})"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b64345ea-c16f-43d3-8640-8a6967bfcecd",
   "metadata": {},
   "source": [
    "# LN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "49f5557d-840f-45b2-9125-b1f0d66eeb9f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tag</th>\n",
       "      <th>cell_count_ref</th>\n",
       "      <th>_batch_ref</th>\n",
       "      <th>cell_count_que</th>\n",
       "      <th>_batch_que</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLE_h_T-map-merge_m_kidney_T</td>\n",
       "      <td>1340</td>\n",
       "      <td>31</td>\n",
       "      <td>3087</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SLE_h_imm-map-merge_m_kidney_imm</td>\n",
       "      <td>2665</td>\n",
       "      <td>34</td>\n",
       "      <td>9288</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SLE_h_B-map-merge_m_kidney_B</td>\n",
       "      <td>418</td>\n",
       "      <td>18</td>\n",
       "      <td>2507</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SLE_h_NK-map-merge_m_kidney_NK</td>\n",
       "      <td>424</td>\n",
       "      <td>28</td>\n",
       "      <td>371</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLE_h_mar-map-merge_m_kidney_mar</td>\n",
       "      <td>384</td>\n",
       "      <td>25</td>\n",
       "      <td>1728</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                tag  cell_count_ref  _batch_ref  \\\n",
       "0      SLE_h_T-map-merge_m_kidney_T            1340          31   \n",
       "1  SLE_h_imm-map-merge_m_kidney_imm            2665          34   \n",
       "2      SLE_h_B-map-merge_m_kidney_B             418          18   \n",
       "3    SLE_h_NK-map-merge_m_kidney_NK             424          28   \n",
       "4  SLE_h_mar-map-merge_m_kidney_mar             384          25   \n",
       "\n",
       "   cell_count_que  _batch_que  \n",
       "0            3087           6  \n",
       "1            9288           6  \n",
       "2            2507           6  \n",
       "3             371           6  \n",
       "4            1728           6  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>任务名</th>\n",
       "      <th>人类细胞数量</th>\n",
       "      <th>人类样本数量</th>\n",
       "      <th>小鼠细胞数量</th>\n",
       "      <th>小鼠样本数量</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SLE_h_imm-map-merge_m_kidney_imm</td>\n",
       "      <td>2665</td>\n",
       "      <td>34</td>\n",
       "      <td>9288</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SLE_h_T-map-merge_m_kidney_T</td>\n",
       "      <td>1340</td>\n",
       "      <td>31</td>\n",
       "      <td>3087</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SLE_h_mar-map-merge_m_kidney_mar</td>\n",
       "      <td>384</td>\n",
       "      <td>25</td>\n",
       "      <td>1728</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                任务名  人类细胞数量  人类样本数量  小鼠细胞数量  小鼠样本数量\n",
       "1  SLE_h_imm-map-merge_m_kidney_imm    2665      34    9288       6\n",
       "0      SLE_h_T-map-merge_m_kidney_T    1340      31    3087       6\n",
       "4  SLE_h_mar-map-merge_m_kidney_mar     384      25    1728       6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "disease = 'LN'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = ut.df.apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "display(df_para_disease_info)\n",
    "\n",
    "display(df_para_disease_info.loc[[1,0,4],:].rename(columns={\n",
    "    'tag':'任务名',\n",
    "    'cell_count_ref':'人类细胞数量',\n",
    "    '_batch_ref':'人类样本数量',\n",
    "    'cell_count_que':'小鼠细胞数量',\n",
    "    '_batch_que':'小鼠样本数量',\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7878ace2-cff0-4d10-84fd-17e66d5aba4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>sp_ref</th>\n",
       "      <th>path_ref</th>\n",
       "      <th>name_ref</th>\n",
       "      <th>sp_simple_ref</th>\n",
       "      <th>sp_que</th>\n",
       "      <th>path_que</th>\n",
       "      <th>name_que</th>\n",
       "      <th>sp_simple_que</th>\n",
       "      <th>key_cell_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_T</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_T</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_imm</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_imm</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_B</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_B</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_NK</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_NK</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LN</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>SLE_h_mar</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>merge_m_kidney_mar</td>\n",
       "      <td>m</td>\n",
       "      <td>sub_cell_type</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  tissue sp_ref                                           path_ref   name_ref  \\\n",
       "0     LN  human  /public/workspace/licanchengup/link/res_publis...    SLE_h_T   \n",
       "1     LN  human  /public/workspace/licanchengup/link/res_publis...  SLE_h_imm   \n",
       "2     LN  human  /public/workspace/licanchengup/link/res_publis...    SLE_h_B   \n",
       "3     LN  human  /public/workspace/licanchengup/link/res_publis...   SLE_h_NK   \n",
       "4     LN  human  /public/workspace/licanchengup/link/res_publis...  SLE_h_mar   \n",
       "\n",
       "  sp_simple_ref sp_que                                           path_que  \\\n",
       "0             h  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "1             h  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "2             h  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "3             h  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "4             h  mouse  /public/workspace/licanchengup/link/res_publis...   \n",
       "\n",
       "             name_que sp_simple_que  key_cell_type  \n",
       "0    merge_m_kidney_T             m  sub_cell_type  \n",
       "1  merge_m_kidney_imm             m      cell_type  \n",
       "2    merge_m_kidney_B             m  sub_cell_type  \n",
       "3   merge_m_kidney_NK             m  sub_cell_type  \n",
       "4  merge_m_kidney_mar             m  sub_cell_type  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_para_disease = df_para_disease.query(\"sp_ref == 'human'\")\n",
    "df_para_disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "705fff41-569f-4904-811e-f93817161ed8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[run start] LN\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LN_h-corss-m;csMAHN;SLE_h_T-map-merge_m_kidney_T;epochs=600;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LN_h-corss-m;csMAHN;SLE_h_imm-map-merge_m_kidney_imm;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LN_h-corss-m;csMAHN;SLE_h_B-map-merge_m_kidney_B;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LN_h-corss-m;csMAHN;SLE_h_NK-map-merge_m_kidney_NK;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240907-1135 LN_h-corss-m;csMAHN;SLE_h_mar-map-merge_m_kidney_mar;epochs=500;is_1v1=False\n"
     ]
    }
   ],
   "source": [
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "for i, row in df_para_disease.iterrows():\n",
    "    \n",
    "    n_epochs = [200, 200, 200] if row['name_ref'] == 'SLE_h_T'   else [100, 200, 200]\n",
    "    run_cross_species_models(\n",
    "        path_adata1=row['path_ref'],\n",
    "        path_adata2=row['path_que'],\n",
    "        key_class1=row['key_cell_type'],\n",
    "        key_class2=row['key_cell_type'],\n",
    "        sp1=row['sp_simple_ref'],\n",
    "        sp2=row['sp_simple_que'],\n",
    "        tissue_name=row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            row['name_ref'],\n",
    "            row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        # models='csMAHN,came'.split(','),\n",
    "        models='csMAHN'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1053f594-18e0-4286-ae4f-2ab635c88410",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[start]240907-1135 LN_h-corss-m;csMAHN;SLE_h_T-map-merge_m_kidney_T;epochs=400;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "                               LN_h   LN_m\n",
      "sub_cell_type                             \n",
      "CTLs                          348.0  126.0\n",
      "Effector memory CD4+ T cells  220.0    NaN\n",
      "GZMK+ CD8+ Tcells             195.0    NaN\n",
      "Central memory CD4+ T cells   181.0    NaN\n",
      "Treg cells                    179.0    NaN\n",
      "Resident memory CD8+ Tcells   106.0    NaN\n",
      "TFH-like cells                 58.0    NaN\n",
      "ISG-high CD4+ T cells          53.0    NaN\n",
      "CD4+ naive                      NaN  705.0\n",
      "CD4+ activated                  NaN  492.0\n",
      "CD4+ Tem                        NaN  441.0\n",
      "CD8+ naive                      NaN  437.0\n",
      "CD8+ Tem                        NaN  363.0\n",
      "CD8+ activated                  NaN  206.0\n",
      "Treg                            NaN  149.0\n",
      "Th2-like                        NaN   74.0\n",
      "Th17-like                       NaN   55.0\n",
      "Th1-like                        NaN   39.0\n",
      "Task: refernece:LN_h 1340 cells x 30155 gene -> query:LN_m 3087 cells x 16261 gene in LN\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      16612\n",
      "ortholog_many2many     5356\n",
      "ortholog_one2many      3174\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 5.222360134124756\n",
      "Leiden results:\n",
      "leiden\n",
      "0    869\n",
      "1    582\n",
      "2    525\n",
      "3    511\n",
      "4    422\n",
      "5     97\n",
      "6     50\n",
      "7     31\n",
      "Name: count, dtype: int64\n",
      "the time of leiden is 1.2928466796875\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 8\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m i, row \u001b[38;5;129;01min\u001b[39;00m df_para_disease\\\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;241m.\u001b[39mquery(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mname_ref == \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSLE_h_B\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m | name_ref == \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSLE_h_T\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m n_epochs \u001b[38;5;129;01min\u001b[39;00m [\n\u001b[1;32m      4\u001b[0m         [\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m200\u001b[39m],\n\u001b[1;32m      5\u001b[0m         [\u001b[38;5;241m100\u001b[39m, \u001b[38;5;241m200\u001b[39m, \u001b[38;5;241m200\u001b[39m],\n\u001b[1;32m      6\u001b[0m         [\u001b[38;5;241m200\u001b[39m, \u001b[38;5;241m200\u001b[39m, \u001b[38;5;241m200\u001b[39m],\n\u001b[1;32m      7\u001b[0m     ]:\n\u001b[0;32m----> 8\u001b[0m         \u001b[43mrun_cross_species_models\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpath_adata1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpath_ref\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpath_adata2\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpath_que\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[43m            \u001b[49m\u001b[43mkey_class1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mkey_cell_type\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     12\u001b[0m \u001b[43m            \u001b[49m\u001b[43mkey_class2\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mkey_cell_type\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     13\u001b[0m \u001b[43m            \u001b[49m\u001b[43msp1\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msp_simple_ref\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     14\u001b[0m \u001b[43m            \u001b[49m\u001b[43msp2\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43msp_simple_que\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     15\u001b[0m \u001b[43m            \u001b[49m\u001b[43mtissue_name\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mtissue\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[43m            \u001b[49m\u001b[43maligned\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maligned\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     17\u001b[0m \u001b[43m            \u001b[49m\u001b[43mresdir_tag\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;132;43;01m{}\u001b[39;49;00m\u001b[38;5;124;43m-map-\u001b[39;49m\u001b[38;5;132;43;01m{}\u001b[39;49;00m\u001b[38;5;124;43m;epochs=\u001b[39;49m\u001b[38;5;132;43;01m{}\u001b[39;49;00m\u001b[38;5;124;43m;is_1v1=\u001b[39;49m\u001b[38;5;132;43;01m{}\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mformat\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     18\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mname_ref\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m                \u001b[49m\u001b[43mrow\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mname_que\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m                \u001b[49m\u001b[38;5;28;43msum\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43mn_epochs\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     21\u001b[0m \u001b[43m                \u001b[49m\u001b[43mis_1v1\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[43m            \u001b[49m\u001b[43mresdir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mp_res\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;66;43;03m# models='csMAHN,came'.split(','),\u001b[39;49;00m\n\u001b[1;32m     24\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmodels\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mcsMAHN\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msplit\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m,\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     25\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43m{\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mn_epochs\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_epochs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     26\u001b[0m \u001b[43m               \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mis_1v1\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mis_1v1\u001b[49m\u001b[43m}\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     27\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m[run finish] \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(disease))\n",
      "File \u001b[0;32m~/link/res_publish/func.py:2634\u001b[0m, in \u001b[0;36mrun_cross_species_models\u001b[0;34m(path_adata1, path_adata2, key_class1, key_class2, sp1, sp2, tissue_name, resdir, resdir_tag, aligned, limite_func, models, **kvargs)\u001b[0m\n\u001b[1;32m   2632\u001b[0m     \u001b[38;5;28mprint\u001b[39m(path_varmap)\n\u001b[1;32m   2633\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m[path_varmap] \u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;241m.\u001b[39mformat(model, Path(path_varmap)\u001b[38;5;241m.\u001b[39mname))\n\u001b[0;32m-> 2634\u001b[0m     _res \u001b[38;5;241m=\u001b[39m \u001b[43mmap_func_run_cross_species_models\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2635\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpath_adata1\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2636\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpath_adata2\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2637\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkey_class1\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2638\u001b[0m \u001b[43m        \u001b[49m\u001b[43mkey_class2\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2639\u001b[0m \u001b[43m        \u001b[49m\u001b[43msp1\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2640\u001b[0m \u001b[43m        \u001b[49m\u001b[43msp2\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2641\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtissue_name\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2642\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpath_varmap\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2643\u001b[0m \u001b[43m        \u001b[49m\u001b[43maligned\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maligned\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2644\u001b[0m \u001b[43m        \u001b[49m\u001b[43mresdir_tag\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m;\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mjoin\u001b[49m\u001b[43m(\u001b[49m\u001b[43m[\u001b[49m\u001b[43mmodel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mresdir_tag\u001b[49m\u001b[43m]\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2645\u001b[0m \u001b[43m        \u001b[49m\u001b[43mresdir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mresdir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2646\u001b[0m \u001b[43m        \u001b[49m\u001b[43mlimite_func\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlimite_func\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2647\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkvargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   2648\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2649\u001b[0m     res\u001b[38;5;241m.\u001b[39mupdate({model: _res})\n\u001b[1;32m   2650\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(res\u001b[38;5;241m.\u001b[39mkeys()) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n",
      "File \u001b[0;32m~/link/res_publish/func.py:2115\u001b[0m, in \u001b[0;36mrun_csMAHN\u001b[0;34m(path_adata1, path_adata2, key_class1, key_class2, sp1, sp2, tissue_name, path_varmap, aligned, resdir_tag, resdir, limite_func, **kvargs)\u001b[0m\n\u001b[1;32m   2113\u001b[0m \u001b[38;5;66;03m# knn时间较长\u001b[39;00m\n\u001b[1;32m   2114\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m[process_for_graph]\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mcenter(\u001b[38;5;241m100\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[0;32m-> 2115\u001b[0m adatas, features_genes, nodes_genes, scnets, one2one, n2n \u001b[38;5;241m=\u001b[39m \u001b[43mcsMAHN\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpp\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mprocess_for_graph\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   2116\u001b[0m \u001b[43m    \u001b[49m\u001b[43m[\u001b[49m\u001b[43madata_raw1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madata_raw2\u001b[49m\u001b[43m]\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mhomo\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey_class\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mleiden\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_hvgs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_hvgs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_degs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_degs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   2117\u001b[0m g, inter_net, one2one_gene_nodes_net, cell_label, n_classes, list_idx \u001b[38;5;241m=\u001b[39m csMAHN\u001b[38;5;241m.\u001b[39mpp\u001b[38;5;241m.\u001b[39mmake_graph(\n\u001b[1;32m   2118\u001b[0m     adatas, aligned, key_class, features_genes, nodes_genes, scnets, one2one, n2n, has_mnn\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, seed\u001b[38;5;241m=\u001b[39mseed)\n\u001b[1;32m   2119\u001b[0m end \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n",
      "File \u001b[0;32m~/link/res_publish/csMAHN/utils/preprocess.py:88\u001b[0m, in \u001b[0;36mprocess_for_graph\u001b[0;34m(adatas, homo, key_class, key_clust, relation, add_no1v1, n_pcs, n_hvgs, n_degs, use_scnets, n_neighbors_scnet, n_neighbors_clust, reso, degs_method)\u001b[0m\n\u001b[1;32m     83\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mthe time of leiden is \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mtime2\u001b[38;5;250m \u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;250m \u001b[39mtime1\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     86\u001b[0m adatas[\u001b[38;5;241m1\u001b[39m]\u001b[38;5;241m.\u001b[39mobs[key_clust] \u001b[38;5;241m=\u001b[39m clust_lbs2\n\u001b[0;32m---> 88\u001b[0m reference_degs \u001b[38;5;241m=\u001b[39m \u001b[43mget_degs\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata1\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroupby\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkey_class\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_degs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mn_degs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdegs_method\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     89\u001b[0m query_degs \u001b[38;5;241m=\u001b[39m get_degs(adata2, groupby\u001b[38;5;241m=\u001b[39mkey_clust, n_degs\u001b[38;5;241m=\u001b[39mn_degs, method\u001b[38;5;241m=\u001b[39mdegs_method)\n\u001b[1;32m     90\u001b[0m time3 \u001b[38;5;241m=\u001b[39m time\u001b[38;5;241m.\u001b[39mtime()\n",
      "File \u001b[0;32m~/link/res_publish/csMAHN/utils/preprocess.py:673\u001b[0m, in \u001b[0;36mget_degs\u001b[0;34m(adata, groupby, key_added, method, n_degs)\u001b[0m\n\u001b[1;32m    661\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_degs\u001b[39m(adata,\n\u001b[1;32m    662\u001b[0m              groupby,\n\u001b[1;32m    663\u001b[0m              key_added\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrank_genes_groups\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    664\u001b[0m              method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mwilcoxon\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    665\u001b[0m              n_degs\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m50\u001b[39m):\n\u001b[1;32m    666\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m    667\u001b[0m \u001b[38;5;124;03m    get different expression genes\u001b[39;00m\n\u001b[1;32m    668\u001b[0m \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    671\u001b[0m \u001b[38;5;124;03m    :return: a list of different expression genes\u001b[39;00m\n\u001b[1;32m    672\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m--> 673\u001b[0m     \u001b[43msc\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtl\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank_genes_groups\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mgroupby\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mgroupby\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey_added\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mkey_added\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    675\u001b[0m     degs_all \u001b[38;5;241m=\u001b[39m adata\u001b[38;5;241m.\u001b[39muns[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrank_genes_groups\u001b[39m\u001b[38;5;124m'\u001b[39m][\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m'\u001b[39m]\n\u001b[1;32m    676\u001b[0m     degs_index \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39marray(degs_all\u001b[38;5;241m.\u001b[39mtolist())[\u001b[38;5;241m0\u001b[39m:n_degs, :]\u001b[38;5;241m.\u001b[39mflatten()\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:608\u001b[0m, in \u001b[0;36mrank_genes_groups\u001b[0;34m(adata, groupby, use_raw, groups, reference, n_genes, rankby_abs, pts, key_added, copy, method, corr_method, tie_correct, layer, **kwds)\u001b[0m\n\u001b[1;32m    605\u001b[0m logg\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mconsider \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mgroupby\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m groups:\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    606\u001b[0m logg\u001b[38;5;241m.\u001b[39mdebug(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwith sizes: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnp\u001b[38;5;241m.\u001b[39mcount_nonzero(test_obj\u001b[38;5;241m.\u001b[39mgroups_masks,\u001b[38;5;250m \u001b[39maxis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m--> 608\u001b[0m \u001b[43mtest_obj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcompute_statistics\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    609\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcorr_method\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_genes_user\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrankby_abs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtie_correct\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\n\u001b[1;32m    610\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    612\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m test_obj\u001b[38;5;241m.\u001b[39mpts \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    613\u001b[0m     groups_names \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mstr\u001b[39m(name) \u001b[38;5;28;01mfor\u001b[39;00m name \u001b[38;5;129;01min\u001b[39;00m test_obj\u001b[38;5;241m.\u001b[39mgroups_order]\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:378\u001b[0m, in \u001b[0;36m_RankGenes.compute_statistics\u001b[0;34m(self, method, corr_method, n_genes_user, rankby_abs, tie_correct, **kwds)\u001b[0m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstats \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[1;32m    376\u001b[0m n_genes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mX\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m]\n\u001b[0;32m--> 378\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m group_index, scores, pvals \u001b[38;5;129;01min\u001b[39;00m generate_test_results:\n\u001b[1;32m    379\u001b[0m     group_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mstr\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgroups_order[group_index])\n\u001b[1;32m    381\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m n_genes_user \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:305\u001b[0m, in \u001b[0;36m_RankGenes.wilcoxon\u001b[0;34m(self, tie_correct)\u001b[0m\n\u001b[1;32m    302\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m tie_correct:\n\u001b[1;32m    303\u001b[0m     T \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39mzeros((n_groups, n_genes))\n\u001b[0;32m--> 305\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m ranks, left, right \u001b[38;5;129;01min\u001b[39;00m _ranks(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mX):\n\u001b[1;32m    306\u001b[0m     \u001b[38;5;66;03m# sum up adjusted_ranks to calculate W_m,n\u001b[39;00m\n\u001b[1;32m    307\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m imask, mask \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mgroups_masks):\n\u001b[1;32m    308\u001b[0m         scores[imask, left:right] \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39msum(ranks\u001b[38;5;241m.\u001b[39miloc[mask, :])\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/tools/_rank_genes_groups.py:63\u001b[0m, in \u001b[0;36m_ranks\u001b[0;34m(X, mask, mask_rest)\u001b[0m\n\u001b[1;32m     60\u001b[0m right \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mmin\u001b[39m(left \u001b[38;5;241m+\u001b[39m max_chunk, n_genes)\n\u001b[1;32m     62\u001b[0m df \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame(data\u001b[38;5;241m=\u001b[39mget_chunk(X, left, right))\n\u001b[0;32m---> 63\u001b[0m ranks \u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;28;01myield\u001b[39;00m ranks, left, right\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/pandas/core/generic.py:10101\u001b[0m, in \u001b[0;36mNDFrame.rank\u001b[0;34m(self, axis, method, numeric_only, na_option, ascending, pct)\u001b[0m\n\u001b[1;32m  10098\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m  10099\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\n\u001b[0;32m> 10101\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mranker\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdata\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/pandas/core/generic.py:10078\u001b[0m, in \u001b[0;36mNDFrame.rank.<locals>.ranker\u001b[0;34m(data)\u001b[0m\n\u001b[1;32m  10070\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m values\u001b[38;5;241m.\u001b[39m_rank(\n\u001b[1;32m  10071\u001b[0m         axis\u001b[38;5;241m=\u001b[39maxis_int,\n\u001b[1;32m  10072\u001b[0m         method\u001b[38;5;241m=\u001b[39mmethod,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m  10075\u001b[0m         pct\u001b[38;5;241m=\u001b[39mpct,\n\u001b[1;32m  10076\u001b[0m     )\n\u001b[1;32m  10077\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m> 10078\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m \u001b[43malgos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m  10079\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  10080\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis_int\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  10081\u001b[0m \u001b[43m        \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  10082\u001b[0m \u001b[43m        \u001b[49m\u001b[43mascending\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mascending\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  10083\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_option\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_option\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  10084\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpct\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpct\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m  10085\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m  10087\u001b[0m ranks_obj \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_constructor(ranks, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mdata\u001b[38;5;241m.\u001b[39m_construct_axes_dict())\n\u001b[1;32m  10088\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m ranks_obj\u001b[38;5;241m.\u001b[39m__finalize__(\u001b[38;5;28mself\u001b[39m, method\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mrank\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "File \u001b[0;32m~/apps/miniconda3/envs/publish/lib/python3.9/site-packages/pandas/core/algorithms.py:1111\u001b[0m, in \u001b[0;36mrank\u001b[0;34m(values, axis, method, na_option, ascending, pct)\u001b[0m\n\u001b[1;32m   1102\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m algos\u001b[38;5;241m.\u001b[39mrank_1d(\n\u001b[1;32m   1103\u001b[0m         values,\n\u001b[1;32m   1104\u001b[0m         is_datetimelike\u001b[38;5;241m=\u001b[39mis_datetimelike,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1108\u001b[0m         pct\u001b[38;5;241m=\u001b[39mpct,\n\u001b[1;32m   1109\u001b[0m     )\n\u001b[1;32m   1110\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m values\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m2\u001b[39m:\n\u001b[0;32m-> 1111\u001b[0m     ranks \u001b[38;5;241m=\u001b[39m \u001b[43malgos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrank_2d\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   1112\u001b[0m \u001b[43m        \u001b[49m\u001b[43mvalues\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1113\u001b[0m \u001b[43m        \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1114\u001b[0m \u001b[43m        \u001b[49m\u001b[43mis_datetimelike\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mis_datetimelike\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1115\u001b[0m \u001b[43m        \u001b[49m\u001b[43mties_method\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1116\u001b[0m \u001b[43m        \u001b[49m\u001b[43mascending\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mascending\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1117\u001b[0m \u001b[43m        \u001b[49m\u001b[43mna_option\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mna_option\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1118\u001b[0m \u001b[43m        \u001b[49m\u001b[43mpct\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mpct\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   1119\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1120\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1121\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mArray with ndim > 2 are not supported.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for i, row in df_para_disease\\\n",
    "    .query(\"name_ref == 'SLE_h_B' | name_ref == 'SLE_h_T' \").iterrows():\n",
    "    for n_epochs in [\n",
    "        [100, 100, 100],\n",
    "        [100, 100, 200],\n",
    "        [100, 200, 200],\n",
    "        [200, 200, 200],\n",
    "    ]:\n",
    "        run_cross_species_models(\n",
    "            path_adata1=row['path_ref'],\n",
    "            path_adata2=row['path_que'],\n",
    "            key_class1=row['key_cell_type'],\n",
    "            key_class2=row['key_cell_type'],\n",
    "            sp1=row['sp_simple_ref'],\n",
    "            sp2=row['sp_simple_que'],\n",
    "            tissue_name=row['tissue'],\n",
    "            aligned=aligned,\n",
    "            resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "                row['name_ref'],\n",
    "                row['name_que'],\n",
    "                sum(n_epochs),\n",
    "                is_1v1),\n",
    "            resdir=p_res,\n",
    "            # models='csMAHN,came'.split(','),\n",
    "            models='csMAHN'.split(','),\n",
    "            **{'n_epochs': n_epochs,\n",
    "               'is_1v1': is_1v1})\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53c975b5-04f7-4f1c-9f3a-7a05716ab092",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, row in df_para_disease\\\n",
    "    .query(\"name_ref == 'SLE_h_NK' | name_ref == 'SLE_h_mar' \").iterrows():\n",
    "    n_epochs = [100, 200, 200]\n",
    "    run_cross_species_models(\n",
    "        path_adata1=row['path_ref'],\n",
    "        path_adata2=row['path_que'],\n",
    "        key_class1=row['key_cell_type'],\n",
    "        key_class2=row['key_cell_type'],\n",
    "        sp1=row['sp_simple_ref'],\n",
    "        sp2=row['sp_simple_que'],\n",
    "        tissue_name=row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            row['name_ref'],\n",
    "            row['name_que'],\n",
    "            sum(n_epochs),\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        # models='csMAHN,came'.split(','),\n",
    "        models='csMAHN'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a2a4a96-d3d0-4f1b-9819-6df90e8e944c",
   "metadata": {},
   "source": [
    "# COVID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24b3fe17-9068-484b-9a39-cbd3cbc2261b",
   "metadata": {},
   "outputs": [],
   "source": [
    "disease = 'COVID'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "display(df_para_disease)\n",
    "\n",
    "df_para_disease.loc[:,'path_ref,path_que'.split(',')].map(lambda x:Path(x).joinpath('matrix.mtx.gz').exists()).join(\n",
    "    df_para_disease.loc[:,'name_ref,name_que'.split(',')] \n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fa62be3-0fbf-4200-bcfd-76360badf634",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "disease = 'COVID'\n",
    "df_para_disease = df_para.query(\"tissue == '{}'\".format(disease))\n",
    "display(df_para_disease)\n",
    "df_para_disease_info = df_para_disease.loc[:,'path_ref,path_que,name_ref,name_que'.split(',')]\n",
    "df_para_disease_info['tag'] = ut.df.apply_merge_field(df_para_disease_info,'{name_ref}-map-{name_que}')\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='name_ref,name_que'.split(','))\n",
    "\n",
    "df_para_disease_info['obs_ref'] = df_para_disease_info['path_ref'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0)\\\n",
    "    .rename(columns={'X_batch':'_batch'})\n",
    ")\n",
    "df_para_disease_info['cell_count_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_ref'] = df_para_disease_info['obs_ref'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_ref'.split(','))\n",
    "# que\n",
    "df_para_disease_info['obs_que'] = df_para_disease_info['path_que'].apply(\n",
    "    lambda x:pd.read_csv(x.joinpath('obs.csv'),index_col=0)\\\n",
    "    .rename(columns={'X_batch':'_batch'}))\n",
    "df_para_disease_info['cell_count_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x.shape[0]\n",
    ")\n",
    "df_para_disease_info['_batch_que'] = df_para_disease_info['obs_que'].apply(\n",
    "    lambda x:x['_batch'].unique().size\n",
    ")\n",
    "\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='obs_que'.split(','))\n",
    "df_para_disease_info = df_para_disease_info.drop(columns='path_ref,path_que'.split(','))\n",
    "display(df_para_disease_info)\n",
    "\n",
    "# display(df_para_disease_info.loc[[1,0,4],:].rename(columns={\n",
    "#     'tag':'任务名',\n",
    "#     'cell_count_ref':'人类细胞数量',\n",
    "#     '_batch_ref':'人类样本数量',\n",
    "#     'cell_count_que':'小鼠细胞数量',\n",
    "#     '_batch_que':'小鼠样本数量',\n",
    "# }))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3168bc05-410a-4416-911d-b68bf2fa757c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_para_disease = df_para_disease[df_para_disease['name_ref'].str.endswith('-c')]\n",
    "print(df_para_disease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a499faa2-e98f-4e6c-9046-c846c21e3455",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "print('\\n[run start] {}\\n'.format(disease))\n",
    "\n",
    "aligned=False\n",
    "is_1v1 = False\n",
    "\n",
    "for n_epochs in [\n",
    "    [100, 100, 100],\n",
    "    [100, 100, 200],\n",
    "    [100, 200, 200],\n",
    "    [200, 200, 200],\n",
    "]:\n",
    "    for i, row in df_para_disease.iterrows():\n",
    "        run_cross_species_models(\n",
    "            path_adata1=row['path_ref'],\n",
    "            path_adata2=row['path_que'],\n",
    "            key_class1=row['key_cell_type'],\n",
    "            key_class2=row['key_cell_type'],\n",
    "            sp1=row['sp_simple_ref'],\n",
    "            sp2=row['sp_simple_que'],\n",
    "            tissue_name=row['tissue'],\n",
    "            aligned=aligned,\n",
    "            resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "                row['name_ref'],\n",
    "                row['name_que'],\n",
    "                sum(n_epochs),\n",
    "                is_1v1),\n",
    "            resdir=p_res,\n",
    "            models='csMAHN'.split(','),\n",
    "            **{'n_epochs': n_epochs,\n",
    "               'is_1v1': is_1v1})\n",
    "\n",
    "print('\\n[run finish] {}\\n'.format(disease))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d377dfe-78bc-4080-9d50-5504d320e4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n[finish]\\n\".center(100, \"-\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "publish",
   "language": "python",
   "name": "publish"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
