{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "943fbf94-54a5-4daf-b28e-47eac8afb1c3",
   "metadata": {},
   "source": [
    "# healthy_run\n",
    "\n",
    "对healthy数据集运行`run_cross_species_models`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd2db94f-a10b-4da7-9f70-1dc623630579",
   "metadata": {},
   "source": [
    "```shell\n",
    "conda activate\n",
    "cd ~/link/res_publish/run\n",
    "jupyter nbconvert healthy_run.ipynb --to python\n",
    "\n",
    "conda activate publish\n",
    "nohup python healthy_run.py > nohup_healthy_run_Retina_parameters.txt &\n",
    "# nohup python healthy_run.py > nohup_healthy_run.txt &\n",
    "jobs\n",
    "\n",
    "conda activate\n",
    "echo 'finish'\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4988b915-bdf0-4ff9-9be5-e800196a876e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "from pathlib import Path\n",
    "p_root = Path('~/link/res_publish').expanduser()\n",
    "None if str(p_root) in sys.path else sys.path.append(str(p_root))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4bc19322-0d07-4fa6-b692-97e2abc89980",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "2024-04-28 12:46:56.328232: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "-------------------------help-------------------------\n",
      "> parameter\n",
      "    p_root\t[name] res_publish\n",
      "        p_run, p_plot, p_res, p_cache, p_pdf\n",
      "    p_df_varmap\n",
      "    map_sp_reverse\n",
      "    rng\n",
      "> run\n",
      "    run_cross_species_models\n",
      "    h5ad_to_mtx\n",
      "    load_adata\n",
      "    get_path_varmap\n",
      "    find_path_from_para\n",
      "    load_normalized_adata\n",
      "\n",
      "> res\n",
      "    get_test_result_df\n",
      "    get_res_obs\n",
      "    get_adata_umap\n",
      "    show_umap\n",
      "\n",
      "> plot\n",
      "    get_color_map\n",
      "    show_color_map\n",
      "    show_color\n",
      "    plot_umap\n",
      "    savefig\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from func import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "16b881e2-f011-4b07-bcf9-b4756a568ca9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tissue</th>\n",
       "      <th>sp_ref</th>\n",
       "      <th>path_ref</th>\n",
       "      <th>name_ref</th>\n",
       "      <th>sp_simple_ref</th>\n",
       "      <th>sp_que</th>\n",
       "      <th>path_que</th>\n",
       "      <th>name_que</th>\n",
       "      <th>sp_simple_que</th>\n",
       "      <th>key_cell_type</th>\n",
       "      <th>mask</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Adrenal-Gland</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_adr</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_adr</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Bone-Marrow</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_bon</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_bon</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Brain</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_bra</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_bra</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Heart</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_hea</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_hea</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Intestine</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_int</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_int</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_kid</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_kid</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Liver</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_liv</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_liv</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Lung</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_lun</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_lun</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Spleen</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_spl</td>\n",
       "      <td>h</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_spl</td>\n",
       "      <td>m</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Adrenal-Gland</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_adr</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_adr</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Bone-Marrow</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_bon</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_bon</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Brain</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_bra</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_bra</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Heart</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_hea</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_hea</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Intestine</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_int</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_int</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_kid</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_kid</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Liver</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_liv</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_liv</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Lung</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_lun</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_lun</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Spleen</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>m_spl</td>\n",
       "      <td>m</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>h_spl</td>\n",
       "      <td>h</td>\n",
       "      <td>CL</td>\n",
       "      <td>HCL_MCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>retina</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Macosko</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>retina</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Menon_seqwell</td>\n",
       "      <td>h</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>retina</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>chicken</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>chicken_P10</td>\n",
       "      <td>c</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>retina-p</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Macosko</td>\n",
       "      <td>m</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina-parameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>retina-p</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Menon_seqwell</td>\n",
       "      <td>h</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina-parameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>retina-p</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>chicken</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>chicken_P10</td>\n",
       "      <td>c</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina-parameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>retina-p</td>\n",
       "      <td>mouse</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Macosko</td>\n",
       "      <td>m</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina-parameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>retina-p</td>\n",
       "      <td>human</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>Menon_seqwell</td>\n",
       "      <td>h</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina-parameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>retina-p</td>\n",
       "      <td>chicken</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>chicken_P10</td>\n",
       "      <td>c</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>zebrafish_Adult</td>\n",
       "      <td>z</td>\n",
       "      <td>cell_type</td>\n",
       "      <td>Retina-parameters</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>embryo</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>emb_z_0</td>\n",
       "      <td>z</td>\n",
       "      <td>xenopus</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>emb_x_0</td>\n",
       "      <td>x</td>\n",
       "      <td>SAMap_cell_type</td>\n",
       "      <td>embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>embryo</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>emb_z_1</td>\n",
       "      <td>z</td>\n",
       "      <td>xenopus</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>emb_x_1</td>\n",
       "      <td>x</td>\n",
       "      <td>SAMap_cell_type</td>\n",
       "      <td>embryo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>embryo</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>emb_z_2</td>\n",
       "      <td>z</td>\n",
       "      <td>xenopus</td>\n",
       "      <td>/public/workspace/licanchengup/link/res_publis...</td>\n",
       "      <td>emb_x_2</td>\n",
       "      <td>x</td>\n",
       "      <td>SAMap_cell_type</td>\n",
       "      <td>embryo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           tissue     sp_ref  \\\n",
       "0   Adrenal-Gland      human   \n",
       "1     Bone-Marrow      human   \n",
       "2           Brain      human   \n",
       "3           Heart      human   \n",
       "4       Intestine      human   \n",
       "5          Kidney      human   \n",
       "6           Liver      human   \n",
       "7            Lung      human   \n",
       "8          Spleen      human   \n",
       "9   Adrenal-Gland      mouse   \n",
       "10    Bone-Marrow      mouse   \n",
       "11          Brain      mouse   \n",
       "12          Heart      mouse   \n",
       "13      Intestine      mouse   \n",
       "14         Kidney      mouse   \n",
       "15          Liver      mouse   \n",
       "16           Lung      mouse   \n",
       "17         Spleen      mouse   \n",
       "0          retina  zebrafish   \n",
       "1          retina  zebrafish   \n",
       "2          retina  zebrafish   \n",
       "0        retina-p  zebrafish   \n",
       "1        retina-p  zebrafish   \n",
       "2        retina-p  zebrafish   \n",
       "3        retina-p      mouse   \n",
       "4        retina-p      human   \n",
       "5        retina-p    chicken   \n",
       "0          embryo  zebrafish   \n",
       "1          embryo  zebrafish   \n",
       "2          embryo  zebrafish   \n",
       "\n",
       "                                             path_ref         name_ref  \\\n",
       "0   /public/workspace/licanchengup/link/res_publis...            h_adr   \n",
       "1   /public/workspace/licanchengup/link/res_publis...            h_bon   \n",
       "2   /public/workspace/licanchengup/link/res_publis...            h_bra   \n",
       "3   /public/workspace/licanchengup/link/res_publis...            h_hea   \n",
       "4   /public/workspace/licanchengup/link/res_publis...            h_int   \n",
       "5   /public/workspace/licanchengup/link/res_publis...            h_kid   \n",
       "6   /public/workspace/licanchengup/link/res_publis...            h_liv   \n",
       "7   /public/workspace/licanchengup/link/res_publis...            h_lun   \n",
       "8   /public/workspace/licanchengup/link/res_publis...            h_spl   \n",
       "9   /public/workspace/licanchengup/link/res_publis...            m_adr   \n",
       "10  /public/workspace/licanchengup/link/res_publis...            m_bon   \n",
       "11  /public/workspace/licanchengup/link/res_publis...            m_bra   \n",
       "12  /public/workspace/licanchengup/link/res_publis...            m_hea   \n",
       "13  /public/workspace/licanchengup/link/res_publis...            m_int   \n",
       "14  /public/workspace/licanchengup/link/res_publis...            m_kid   \n",
       "15  /public/workspace/licanchengup/link/res_publis...            m_liv   \n",
       "16  /public/workspace/licanchengup/link/res_publis...            m_lun   \n",
       "17  /public/workspace/licanchengup/link/res_publis...            m_spl   \n",
       "0   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "1   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "2   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "0   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "1   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "2   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "3   /public/workspace/licanchengup/link/res_publis...          Macosko   \n",
       "4   /public/workspace/licanchengup/link/res_publis...    Menon_seqwell   \n",
       "5   /public/workspace/licanchengup/link/res_publis...      chicken_P10   \n",
       "0   /public/workspace/licanchengup/link/res_publis...          emb_z_0   \n",
       "1   /public/workspace/licanchengup/link/res_publis...          emb_z_1   \n",
       "2   /public/workspace/licanchengup/link/res_publis...          emb_z_2   \n",
       "\n",
       "   sp_simple_ref     sp_que  \\\n",
       "0              h      mouse   \n",
       "1              h      mouse   \n",
       "2              h      mouse   \n",
       "3              h      mouse   \n",
       "4              h      mouse   \n",
       "5              h      mouse   \n",
       "6              h      mouse   \n",
       "7              h      mouse   \n",
       "8              h      mouse   \n",
       "9              m      human   \n",
       "10             m      human   \n",
       "11             m      human   \n",
       "12             m      human   \n",
       "13             m      human   \n",
       "14             m      human   \n",
       "15             m      human   \n",
       "16             m      human   \n",
       "17             m      human   \n",
       "0              z      mouse   \n",
       "1              z      human   \n",
       "2              z    chicken   \n",
       "0              z      mouse   \n",
       "1              z      human   \n",
       "2              z    chicken   \n",
       "3              m  zebrafish   \n",
       "4              h  zebrafish   \n",
       "5              c  zebrafish   \n",
       "0              z    xenopus   \n",
       "1              z    xenopus   \n",
       "2              z    xenopus   \n",
       "\n",
       "                                             path_que         name_que  \\\n",
       "0   /public/workspace/licanchengup/link/res_publis...            m_adr   \n",
       "1   /public/workspace/licanchengup/link/res_publis...            m_bon   \n",
       "2   /public/workspace/licanchengup/link/res_publis...            m_bra   \n",
       "3   /public/workspace/licanchengup/link/res_publis...            m_hea   \n",
       "4   /public/workspace/licanchengup/link/res_publis...            m_int   \n",
       "5   /public/workspace/licanchengup/link/res_publis...            m_kid   \n",
       "6   /public/workspace/licanchengup/link/res_publis...            m_liv   \n",
       "7   /public/workspace/licanchengup/link/res_publis...            m_lun   \n",
       "8   /public/workspace/licanchengup/link/res_publis...            m_spl   \n",
       "9   /public/workspace/licanchengup/link/res_publis...            h_adr   \n",
       "10  /public/workspace/licanchengup/link/res_publis...            h_bon   \n",
       "11  /public/workspace/licanchengup/link/res_publis...            h_bra   \n",
       "12  /public/workspace/licanchengup/link/res_publis...            h_hea   \n",
       "13  /public/workspace/licanchengup/link/res_publis...            h_int   \n",
       "14  /public/workspace/licanchengup/link/res_publis...            h_kid   \n",
       "15  /public/workspace/licanchengup/link/res_publis...            h_liv   \n",
       "16  /public/workspace/licanchengup/link/res_publis...            h_lun   \n",
       "17  /public/workspace/licanchengup/link/res_publis...            h_spl   \n",
       "0   /public/workspace/licanchengup/link/res_publis...          Macosko   \n",
       "1   /public/workspace/licanchengup/link/res_publis...    Menon_seqwell   \n",
       "2   /public/workspace/licanchengup/link/res_publis...      chicken_P10   \n",
       "0   /public/workspace/licanchengup/link/res_publis...          Macosko   \n",
       "1   /public/workspace/licanchengup/link/res_publis...    Menon_seqwell   \n",
       "2   /public/workspace/licanchengup/link/res_publis...      chicken_P10   \n",
       "3   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "4   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "5   /public/workspace/licanchengup/link/res_publis...  zebrafish_Adult   \n",
       "0   /public/workspace/licanchengup/link/res_publis...          emb_x_0   \n",
       "1   /public/workspace/licanchengup/link/res_publis...          emb_x_1   \n",
       "2   /public/workspace/licanchengup/link/res_publis...          emb_x_2   \n",
       "\n",
       "   sp_simple_que    key_cell_type               mask  \n",
       "0              m               CL            HCL_MCA  \n",
       "1              m               CL            HCL_MCA  \n",
       "2              m               CL            HCL_MCA  \n",
       "3              m               CL            HCL_MCA  \n",
       "4              m               CL            HCL_MCA  \n",
       "5              m               CL            HCL_MCA  \n",
       "6              m               CL            HCL_MCA  \n",
       "7              m               CL            HCL_MCA  \n",
       "8              m               CL            HCL_MCA  \n",
       "9              h               CL            HCL_MCA  \n",
       "10             h               CL            HCL_MCA  \n",
       "11             h               CL            HCL_MCA  \n",
       "12             h               CL            HCL_MCA  \n",
       "13             h               CL            HCL_MCA  \n",
       "14             h               CL            HCL_MCA  \n",
       "15             h               CL            HCL_MCA  \n",
       "16             h               CL            HCL_MCA  \n",
       "17             h               CL            HCL_MCA  \n",
       "0              m        cell_type             Retina  \n",
       "1              h        cell_type             Retina  \n",
       "2              c        cell_type             Retina  \n",
       "0              m        cell_type  Retina-parameters  \n",
       "1              h        cell_type  Retina-parameters  \n",
       "2              c        cell_type  Retina-parameters  \n",
       "3              z        cell_type  Retina-parameters  \n",
       "4              z        cell_type  Retina-parameters  \n",
       "5              z        cell_type  Retina-parameters  \n",
       "0              x  SAMap_cell_type             embryo  \n",
       "1              x  SAMap_cell_type             embryo  \n",
       "2              x  SAMap_cell_type             embryo  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCL_MCA,Retina,Retina-parameters,embryo\n"
     ]
    }
   ],
   "source": [
    "q_item = 'HCL_MCA,Retina,Retina-parameters,embryo'.split(',')\n",
    "\n",
    "df_para = pd.concat([pd.read_csv(p_cache.joinpath(\n",
    "    'parameter_healthy_{}.csv'.format(i))).assign(mask=i)\n",
    "    for i in q_item])\n",
    "df_para['path_ref'] = df_para['path_ref'].apply(\n",
    "    lambda x: p_cache.joinpath(x))\n",
    "df_para['path_que'] = df_para['path_que'].apply(\n",
    "    lambda x: p_cache.joinpath(x))\n",
    "display(df_para)\n",
    "\n",
    "for _i, _row in df_para.iterrows():\n",
    "    pass\n",
    "_row\n",
    "print(','.join(q_item))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d5e2b20-c197-443b-b2ba-04e897dbbccb",
   "metadata": {},
   "source": [
    "# HCL_MCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57ebd526-dfe2-4f8e-a893-0dde415ebdb7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[runstart] HCL_MCA\n",
      "\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Adrenal-Gland_h-corss-m;csMAHN;h_adr-map-m_adr;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Adrenal-Gland_h-corss-m;came;h_adr-map-m_adr;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Bone-Marrow_h-corss-m;csMAHN;h_bon-map-m_bon;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Bone-Marrow_h-corss-m;came;h_bon-map-m_bon;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Brain_h-corss-m;csMAHN;h_bra-map-m_bra;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Brain_h-corss-m;came;h_bra-map-m_bra;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Heart_h-corss-m;csMAHN;h_hea-map-m_hea;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Heart_h-corss-m;came;h_hea-map-m_hea;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Intestine_h-corss-m;csMAHN;h_int-map-m_int;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Intestine_h-corss-m;came;h_int-map-m_int;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Kidney_h-corss-m;csMAHN;h_kid-map-m_kid;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Kidney_h-corss-m;came;h_kid-map-m_kid;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Liver_h-corss-m;csMAHN;h_liv-map-m_liv;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Liver_h-corss-m;came;h_liv-map-m_liv;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Lung_h-corss-m;csMAHN;h_lun-map-m_lun;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Lung_h-corss-m;came;h_lun-map-m_lun;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] csMAHN\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Spleen_h-corss-m;csMAHN;h_spl-map-m_spl;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/human_to_mouse.txt\n",
      "[path_varmap] came\thuman_to_mouse.txt\n",
      "[has finish]240428-1247 Spleen_h-corss-m;came;h_spl-map-m_spl;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Adrenal-Gland_m-corss-h;csMAHN;m_adr-map-h_adr;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Adrenal-Gland_m-corss-h;came;m_adr-map-h_adr;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Bone-Marrow_m-corss-h;csMAHN;m_bon-map-h_bon;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Bone-Marrow_m-corss-h;came;m_bon-map-h_bon;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Brain_m-corss-h;csMAHN;m_bra-map-h_bra;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Brain_m-corss-h;came;m_bra-map-h_bra;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Heart_m-corss-h;csMAHN;m_hea-map-h_hea;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Heart_m-corss-h;came;m_hea-map-h_hea;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Intestine_m-corss-h;csMAHN;m_int-map-h_int;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Intestine_m-corss-h;came;m_int-map-h_int;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Kidney_m-corss-h;csMAHN;m_kid-map-h_kid;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Kidney_m-corss-h;came;m_kid-map-h_kid;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Liver_m-corss-h;csMAHN;m_liv-map-h_liv;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] came\tmouse_to_human.txt\n",
      "[has finish]240428-1247 Liver_m-corss-h;came;m_liv-map-h_liv;epochs=500;is_1v1=False\n",
      "/public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n",
      "[path_varmap] csMAHN\tmouse_to_human.txt\n",
      "[start]240428-1247 Lung_m-corss-h;csMAHN;m_lun-map-h_lun;epochs=500;is_1v1=False\n",
      "[path_varmap] /public/workspace/licanchengup/link/res_publish/homo/mouse_to_human.txt\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n",
      "/public/workspace/licanchengup/link/res_publish/csMAHN/utils/preprocess.py:413: FutureWarning: pandas.value_counts is deprecated and will be removed in a future version. Use pd.Series(obj).value_counts() instead.\n",
      "  type_counts_list.append(pd.value_counts(adatas[i].obs[key_class]))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "----raw----\n",
      "                                            reference  query\n",
      "CL                                                          \n",
      "neutrophil                                    23726.0   2787\n",
      "alveolar capillary type 1 endothelial cell    15032.0   2192\n",
      "mesenchymal cell                               8764.0   3293\n",
      "fibroblast                                     8233.0   7573\n",
      "macrophage                                     5659.0   5266\n",
      "alveolar capillary type 2 endothelial cell     2972.0   6473\n",
      "endothelial cell                                  NaN   3981\n",
      "T cell                                            NaN   1960\n",
      "----new----\n",
      "                                            reference  query\n",
      "CL                                                          \n",
      "neutrophil                                    23726.0   2787\n",
      "alveolar capillary type 1 endothelial cell    15032.0   2192\n",
      "mesenchymal cell                               8764.0   3293\n",
      "fibroblast                                     8233.0   7573\n",
      "macrophage                                     5659.0   5266\n",
      "alveolar capillary type 2 endothelial cell     2972.0   6473\n",
      "                                            Lung_m  Lung_h\n",
      "CL                                                        \n",
      "neutrophil                                   23726    2787\n",
      "alveolar capillary type 1 endothelial cell   15032    2192\n",
      "mesenchymal cell                              8764    3293\n",
      "fibroblast                                    8233    7573\n",
      "macrophage                                    5659    5266\n",
      "alveolar capillary type 2 endothelial cell    2972    6473\n",
      "Task: refernece:Lung_m 64386 cells x 11102 gene -> query:Lung_h 27584 cells x 13874 gene in Lung\n",
      "---------------------------------------\n",
      "[process_for_graph]\n",
      "----------------------------------------\n",
      "Homolog information follows\n",
      "homology_type\n",
      "ortholog_one2one      17068\n",
      "ortholog_many2many     4843\n",
      "ortholog_one2many      2673\n",
      "Name: count, dtype: int64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n",
      "/public/workspace/licanchengup/apps/miniconda3/envs/publish/lib/python3.9/site-packages/scanpy/preprocessing/_highly_variable_genes.py:216: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  disp_grouped = df.groupby('mean_bin')['dispersions']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the time2 of processing adatas is 26.809433937072754\n"
     ]
    }
   ],
   "source": [
    "# item = 'HCL_MCA'\n",
    "# df_para_item = df_para.query(\"mask == '{}'\".format(item))\n",
    "# print('\\n[runstart] {}\\n'.format(item))\n",
    "\n",
    "# aligned=True\n",
    "# is_1v1 = False\n",
    "\n",
    "# n_epochs = [100, 200, 200]\n",
    "# for _i, _row in df_para_item.iterrows():\n",
    "#     run_cross_species_models(\n",
    "#         path_adata1=_row['path_ref'],\n",
    "#         path_adata2=_row['path_que'],\n",
    "#         key_class1=_row['key_cell_type'],\n",
    "#         key_class2=_row['key_cell_type'],\n",
    "#         sp1=_row['sp_simple_ref'],\n",
    "#         sp2=_row['sp_simple_que'],\n",
    "#         tissue_name=_row['tissue'],\n",
    "#         aligned=aligned,\n",
    "#         resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "#             _row['name_ref'],\n",
    "#             _row['name_que'],\n",
    "#             sum(n_epochs),\n",
    "#             is_1v1),\n",
    "#         resdir=p_res,\n",
    "#         models='csMAHN,came'.split(','),\n",
    "#         **{'n_epochs': n_epochs,\n",
    "#            'is_1v1': is_1v1})\n",
    "\n",
    "# # SAMap\n",
    "# for _i, _row in df_para_item.iterrows():\n",
    "#     run_cross_species_models(\n",
    "#         path_adata1=_row['path_ref'],\n",
    "#         path_adata2=_row['path_que'],\n",
    "#         key_class1=_row['key_cell_type'],\n",
    "#         key_class2=_row['key_cell_type'],\n",
    "#         sp1=_row['sp_simple_ref'],\n",
    "#         sp2=_row['sp_simple_que'],\n",
    "#         tissue_name=_row['tissue'],\n",
    "#         aligned=aligned,\n",
    "#         resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "#             _row['name_ref'],\n",
    "#             _row['name_que'],\n",
    "#             is_1v1),\n",
    "#         resdir=p_res,\n",
    "#         models='SAMap'.split(','),\n",
    "#         **{\n",
    "#            'is_1v1': is_1v1})\n",
    "\n",
    "# print('\\n[run finish] {}\\n'.format(item))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffcac2a0-d2d7-4c67-be32-a23ab50a4ce0",
   "metadata": {},
   "source": [
    "# Retina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "676e8abf-752a-4497-84f5-29aa22f86e6b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "item = 'Retina'\n",
    "df_para_item = df_para.query(\"mask == '{}'\".format(item))\n",
    "print('\\n[run start] {}\\n'.format(item))\n",
    "\n",
    "aligned=True\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "\n",
    "for _i, _row in df_para_item.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};epochs={};is_1v1={}'.format(\n",
    "            _row['name_ref'],_row['name_que'],\n",
    "            sum(n_epochs),is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='csMAHN,came'.split(','),\n",
    "        **{'n_epochs': n_epochs,\n",
    "           'is_1v1': is_1v1})\n",
    "\n",
    "# SAMap\n",
    "for _i, _row in df_para_item.iterrows():\n",
    "    run_cross_species_models(\n",
    "        path_adata1=_row['path_ref'],\n",
    "        path_adata2=_row['path_que'],\n",
    "        key_class1=_row['key_cell_type'],\n",
    "        key_class2=_row['key_cell_type'],\n",
    "        sp1=_row['sp_simple_ref'],\n",
    "        sp2=_row['sp_simple_que'],\n",
    "        tissue_name=_row['tissue'],\n",
    "        aligned=aligned,\n",
    "        resdir_tag='{}-map-{};is_1v1={}'.format(\n",
    "            _row['name_ref'],\n",
    "            _row['name_que'],\n",
    "            is_1v1),\n",
    "        resdir=p_res,\n",
    "        models='SAMap'.split(','),\n",
    "        **{\n",
    "           'is_1v1': is_1v1})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d44406-8dbf-4130-8313-ca22ec372fd3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "item = 'Retina-parameters'\n",
    "df_para_item = df_para.query(\"mask == '{}'\".format(item))\n",
    "print('\\n[run start] {}\\n'.format(item))\n",
    "df_para_item\n",
    "aligned = True\n",
    "is_1v1 = False\n",
    "\n",
    "n_epochs = [100, 200, 200]\n",
    "\n",
    "for n_hvgs, n_degs in [\n",
    "    [1000, 50], [2000, 50], [3000, 50],\n",
    "        [1000, 100], [2000, 100], [3000, 100]]:\n",
    "    print('{} {}'.format(n_hvgs, n_degs))\n",
    "    for _i, _row in df_para_item.iterrows():\n",
    "        run_cross_species_models(\n",
    "            path_adata1=_row['path_ref'],\n",
    "            path_adata2=_row['path_que'],\n",
    "            key_class1=_row['key_cell_type'],\n",
    "            key_class2=_row['key_cell_type'],\n",
    "            sp1=_row['sp_simple_ref'],\n",
    "            sp2=_row['sp_simple_que'],\n",
    "            tissue_name=_row['tissue'],\n",
    "            aligned=aligned,\n",
    "            resdir_tag='{}-map-{};epochs={};is_1v1={};n_hvgs={};n_degs={}'.format(\n",
    "                _row['name_ref'], _row['name_que'],\n",
    "                sum(n_epochs), is_1v1,\n",
    "                n_hvgs, n_degs),\n",
    "            resdir=p_res,\n",
    "            models='csMAHN'.split(','),\n",
    "            **{'n_epochs': n_epochs,\n",
    "               'is_1v1': is_1v1,\n",
    "               'n_hvgs': n_hvgs,\n",
    "               'n_degs': n_degs\n",
    "               })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d50bd3d-b6e3-4d68-8cf8-5f55f9542a08",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "\n",
    "for nfeats, hidden in [\n",
    "    [32, 64],\n",
    "    [64, 64],  # default\n",
    "    [128, 64],\n",
    "    [64, 32],\n",
    "    [64, 128]\n",
    "\n",
    "]:\n",
    "    print('{} {}'.format(nfeats, hidden))\n",
    "    for _i, _row in df_para_item.iterrows():\n",
    "        run_cross_species_models(\n",
    "            path_adata1=_row['path_ref'],\n",
    "            path_adata2=_row['path_que'],\n",
    "            key_class1=_row['key_cell_type'],\n",
    "            key_class2=_row['key_cell_type'],\n",
    "            sp1=_row['sp_simple_ref'],\n",
    "            sp2=_row['sp_simple_que'],\n",
    "            tissue_name=_row['tissue'],\n",
    "            aligned=aligned,\n",
    "            resdir_tag='{}-map-{};epochs={};is_1v1={};nfeats={};hidden={}'.format(\n",
    "                _row['name_ref'], _row['name_que'],\n",
    "                sum(n_epochs), is_1v1,\n",
    "                nfeats, hidden),\n",
    "            resdir=p_res,\n",
    "            models='csMAHN'.split(','),\n",
    "            **{'n_epochs': n_epochs,\n",
    "               'is_1v1': is_1v1,\n",
    "              'nfeats':nfeats,\n",
    "               'hidden':hidden})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "554cf383-008e-4870-8e67-8634e723b39e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('\\n[run finish] {}\\n'.format(item))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3d377dfe-78bc-4080-9d50-5504d320e4ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"\\n[finish]\\n\".center(100, \"-\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "publish",
   "language": "python",
   "name": "publish"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
